

<!DOCTYPE html>
<html class="writer-html4" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>knowledgebases &mdash; Comprior  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript">
          var DOCUMENTATION_OPTIONS = {
              URL_ROOT:'../',
              VERSION:'',
              LANGUAGE:'None',
              COLLAPSE_INDEX:false,
              FILE_SUFFIX:'.html',
              HAS_SOURCE:  true,
              SOURCELINK_SUFFIX: '.txt'
          };
      </script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> Comprior
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation and Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../inputdata.html">Input Data Sets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../configparams.html">Configuration Parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../outputstructure.html">Output Folder Structure  - Where to find what Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../knowledgebases.html">Knowledge Bases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../priorknowledge.html">Prior Knowledge Approaches</a></li>
<li class="toctree-l1"><a class="reference internal" href="../howtos.html">How-Tos</a></li>
<li class="toctree-l1"><a class="reference internal" href="../comprior.html">Python Code Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../r-doc.html">R Code Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../java-doc.html">Java Code Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../architecture.html">System Architecture</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Comprior</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="index.html">Module code</a> &raquo;</li>
        
      <li>knowledgebases</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for knowledgebases</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">abc</span> <span class="k">import</span> <span class="n">abstractmethod</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">from</span> <span class="nn">opentargets</span> <span class="k">import</span> <span class="n">OpenTargetsClient</span>
<span class="kn">from</span> <span class="nn">bioservices</span> <span class="k">import</span> <span class="n">KEGG</span><span class="p">,</span> <span class="n">REST</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">benchutils</span> <span class="k">as</span> <span class="nn">util</span>
<span class="kn">import</span> <span class="nn">time</span><span class="o">,</span> <span class="nn">requests</span><span class="o">,</span> <span class="nn">json</span>
<span class="kn">from</span> <span class="nn">datetime</span> <span class="k">import</span> <span class="n">datetime</span>
<span class="kn">from</span> <span class="nn">lxml.html</span> <span class="k">import</span> <span class="n">fromstring</span>
<span class="kn">import</span> <span class="nn">pypath.core.network</span> <span class="k">as</span> <span class="nn">network</span>
<span class="kn">import</span> <span class="nn">pypath.core.entity</span> <span class="k">as</span> <span class="nn">entity</span>
<span class="kn">import</span> <span class="nn">pypath.internals.input_formats</span> <span class="k">as</span> <span class="nn">input_formats</span>


<span class="c1">############################### WEB SERVICES ###############################</span>

<div class="viewcode-block" id="ENRICHR"><a class="viewcode-back" href="../comprior.html#knowledgebases.ENRICHR">[docs]</a><span class="k">class</span> <span class="nc">ENRICHR</span><span class="p">(</span><span class="n">REST</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Queries some of the API endpoints of the EnrichR web service (https://maayanlab.cloud/Enrichr/help#api).</span>

<span class="sd">       :param config: configuration parameters for EnrichR web service (as specified in config file).</span>
<span class="sd">       :type config: dict</span>
<span class="sd">       &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">config</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfig</span><span class="p">(</span><span class="s2">&quot;Enrichr&quot;</span><span class="p">)</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="s2">&quot;ENRICHR&quot;</span><span class="p">,</span> <span class="n">url</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;webservice_uri&quot;</span><span class="p">])</span>

<div class="viewcode-block" id="ENRICHR.addlist"><a class="viewcode-back" href="../comprior.html#knowledgebases.ENRICHR.addlist">[docs]</a>    <span class="k">def</span> <span class="nf">addlist</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">geneList</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Queries EnrichR to annotate a given list of genes.</span>
<span class="sd">           Returns a userListID, which can be used to retrieve the actual results in a second query.</span>

<span class="sd">           :param geneList: list of genes to annotate</span>
<span class="sd">           :type genes: list of str</span>
<span class="sd">           :return: json response containing a userListID.</span>
<span class="sd">           :rtype: dict of str</span>
<span class="sd">           &quot;&quot;&quot;</span>

        <span class="n">genesStr</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">geneList</span><span class="p">:</span>
            <span class="n">genesStr</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="n">payload</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;list&#39;</span><span class="p">:</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="n">genesStr</span><span class="p">)</span>
        <span class="p">}</span>

        <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">http_post</span><span class="p">(</span><span class="s1">&#39;addList&#39;</span><span class="p">,</span> <span class="n">files</span><span class="o">=</span><span class="n">payload</span><span class="p">,</span> <span class="n">frmt</span><span class="o">=</span><span class="s2">&quot;json&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">response</span></div>

<div class="viewcode-block" id="ENRICHR.export"><a class="viewcode-back" href="../comprior.html#knowledgebases.ENRICHR.export">[docs]</a>    <span class="k">def</span> <span class="nf">export</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Download file of enrichment results.</span>
<span class="sd">           Requires a userListId that was retrieved from a prior query.</span>

<span class="sd">           :param params: list of parameters to use for that query (userListId: Identifier returned from addList endpoint, filename: Name of text file download, backgroundType: Gene set library for which to download results)</span>
<span class="sd">           :type params: list of str</span>
<span class="sd">           :return: text file containing enrichment results.</span>
<span class="sd">           :rtype: str</span>
<span class="sd">           &quot;&quot;&quot;</span>

        <span class="n">params</span><span class="p">[</span><span class="s2">&quot;stream&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;true&quot;</span>

        <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">http_get</span><span class="p">(</span><span class="s2">&quot;export&quot;</span><span class="p">,</span> <span class="n">params</span> <span class="o">=</span> <span class="n">params</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">response</span></div>

<div class="viewcode-block" id="ENRICHR.genemap"><a class="viewcode-back" href="../comprior.html#knowledgebases.ENRICHR.genemap">[docs]</a>    <span class="k">def</span> <span class="nf">genemap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Finds all terms, their descriptions, and optional categorizations, for a given gene identifier.</span>

<span class="sd">           :param params: list of parameters to be used for the query (gene	Gene to use in search for terms, json (optional): Set &quot;true&quot; to return JSON rather plaintext, setup (optional): Set &quot;true&quot; to category information for the libraries)</span>
<span class="sd">           :type params: list of str</span>
<span class="sd">           :return: json object of all terms containing the specified gene and their descriptions.</span>
<span class="sd">           :rtype: dict of str</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">http_get</span><span class="p">(</span><span class="s2">&quot;genemap&quot;</span><span class="p">,</span> <span class="n">params</span> <span class="o">=</span> <span class="n">params</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">response</span></div>

<div class="viewcode-block" id="ENRICHR.enrich"><a class="viewcode-back" href="../comprior.html#knowledgebases.ENRICHR.enrich">[docs]</a>    <span class="k">def</span> <span class="nf">enrich</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns all that are terms available in library (specified by backgroundType param) and enriched in the given set of genes (specified by userListId param).</span>

<span class="sd">            :param params: list of parameters to be used for the query (userListId: Identifier returned from addList endpoint; backgroundType: Gene set library to enrich against)</span>
<span class="sd">            :type params: list of str</span>
<span class="sd">            :return: dataframe object of all enriched terms (unsorted, unfiltered.</span>
<span class="sd">            :rtype: dataframe</span>
<span class="sd">            &quot;&quot;&quot;</span>

        <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">http_get</span><span class="p">(</span><span class="s2">&quot;enrich&quot;</span><span class="p">,</span> <span class="n">params</span> <span class="o">=</span> <span class="n">params</span><span class="p">)</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">values</span><span class="p">())[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">res</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Rank&quot;</span><span class="p">,</span> <span class="s2">&quot;Term name&quot;</span><span class="p">,</span> <span class="s2">&quot;P-value&quot;</span><span class="p">,</span> <span class="s2">&quot;Z-score&quot;</span><span class="p">,</span> <span class="s2">&quot;Combined score&quot;</span><span class="p">,</span> <span class="s2">&quot;Overlapping genes&quot;</span><span class="p">,</span> <span class="s2">&quot;Adjusted p-value&quot;</span><span class="p">,</span> <span class="s2">&quot;Old p-value&quot;</span><span class="p">,</span> <span class="s2">&quot;Old adjusted p-value&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">res</span><span class="p">[[</span><span class="s2">&quot;Term name&quot;</span><span class="p">,</span> <span class="s2">&quot;P-value&quot;</span><span class="p">,</span> <span class="s2">&quot;Z-score&quot;</span><span class="p">,</span> <span class="s2">&quot;Combined score&quot;</span><span class="p">,</span> <span class="s2">&quot;Overlapping genes&quot;</span><span class="p">,</span> <span class="s2">&quot;Adjusted p-value&quot;</span><span class="p">,</span> <span class="s2">&quot;Old p-value&quot;</span><span class="p">,</span> <span class="s2">&quot;Old adjusted p-value&quot;</span><span class="p">]]</span></div></div>


<div class="viewcode-block" id="UMLS_AUTH"><a class="viewcode-back" href="../comprior.html#knowledgebases.UMLS_AUTH">[docs]</a><span class="k">class</span> <span class="nc">UMLS_AUTH</span><span class="p">(</span><span class="n">REST</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Singleton class.</span>
<span class="sd">       Python code encapsulates it in a way that is not shown in Sphinx, so have a look at the descriptions in the source code.</span>

<span class="sd">       Authentication service to get access to the UMLS database UMLS database (which we need for retrieving CUI disease codes for querying DisGeNET).</span>
<span class="sd">       You first have to get a ticket-granting ticket (tgt, valid for 8 hours) with the help of an API key.</span>
<span class="sd">       With the tgt, you can then request a service ticket for every new query to the UMLS database.</span>
<span class="sd">       The service ticket must then be used for the query.</span>
<span class="sd">       The task of this class is to generate a valid tgt and subsequent service ticket.</span>
<span class="sd">       Documentation on the authentication process: https://documentation.uts.nlm.nih.gov/rest/authentication.html</span>

<span class="sd">       :param config: configuration parameters for UMLS web service as specified in config file.</span>
<span class="sd">       :type config: dict</span>
<span class="sd">       :param tgt_timestamp: timestamp of the tgt. If it is older than 8 hours, we need to request a new tgt.</span>
<span class="sd">       :type tgt_timestamp: str</span>
<span class="sd">       :param tgt: id of the ticket-granting ticket (valid for 8 hours). With this ticket, we can then query the actual UMLS web service.</span>
<span class="sd">       :type tgt: list of str</span>
<span class="sd">       :param service: uri for the service login</span>
<span class="sd">       :type service: str</span>
<span class="sd">       &quot;&quot;&quot;</span>
    <span class="k">class</span> <span class="nc">__UMLS_AUTH</span><span class="p">(</span><span class="n">REST</span><span class="p">):</span>
        <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">config</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfig</span><span class="p">(</span><span class="s2">&quot;UMLS&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tgt_timestamp</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tgt</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">service</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;loginservice_uri&quot;</span><span class="p">]</span>
            <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="s2">&quot;UMLS_AUTH&quot;</span><span class="p">,</span> <span class="n">url</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;login_uri&quot;</span><span class="p">])</span>

        <span class="k">def</span> <span class="nf">get_tgt</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="sd">&quot;&quot;&quot;Get a ticket-granting ticket (tgt) from the authentication service.</span>

<span class="sd">               :return: a valid tgt (valid for 8 hours).</span>
<span class="sd">               :rtype: str</span>
<span class="sd">               &quot;&quot;&quot;</span>
            <span class="c1"># we use a different uri here, so use requests instead of the bioservices interface</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tgt_timestamp</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span>

            <span class="n">params</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;apikey&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;apikey&quot;</span><span class="p">]}</span>
            <span class="n">r</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">http_post</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;auth_endpoint&quot;</span><span class="p">],</span> <span class="n">data</span><span class="o">=</span><span class="n">params</span><span class="p">)</span>


            <span class="n">response</span> <span class="o">=</span> <span class="n">fromstring</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
            <span class="c1">## extract the entire URL needed from the HTML form (action attribute) returned - looks similar to https://utslogin.nlm.nih.gov/cas/v1/tickets/TGT-36471-aYqNLN2rFIJPXKzxwdTNC5ZT7z3B3cTAKfSc5ndHQcUxeaDOLN-cas</span>
            <span class="c1">## we make a POST call to this URL in the getst method</span>
            <span class="c1">#only extract the very last part of the uri</span>
            <span class="n">tgt_str</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">xpath</span><span class="p">(</span><span class="s1">&#39;//form/@action&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">tgt</span> <span class="o">=</span> <span class="n">tgt_str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">return</span> <span class="n">tgt</span>

        <span class="c1"># authentication function: get the service ticket - valid for up to 8 minutes</span>
        <span class="k">def</span> <span class="nf">get_st</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="sd">&quot;&quot;&quot;Get a service ticket from the authentication service (valid for 8 minutes), which can then be used in the actual query.</span>
<span class="sd">               In order to get a service ticket, a valid ticket-granting ticket (tgt) must be provided.</span>
<span class="sd">               If the last tgt is outdated, generate a new one.</span>

<span class="sd">               :return: service ticket.</span>
<span class="sd">               :rtype: str</span>
<span class="sd">               &quot;&quot;&quot;</span>
            <span class="c1"># check if a valid tgt exists (expires after 8 hours)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">tgt</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># request new tgt</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">tgt</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_tgt</span><span class="p">()</span>
            <span class="k">elif</span> <span class="p">(</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">tgt_timestamp</span><span class="p">)</span><span class="o">.</span><span class="n">total_seconds</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">28800</span><span class="p">:</span>
                <span class="c1"># request new tgt</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">tgt</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_tgt</span><span class="p">()</span>

            <span class="n">params</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;service&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">service</span><span class="p">}</span>
            <span class="n">h</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Content-type&quot;</span><span class="p">:</span> <span class="s2">&quot;application/x-www-form-urlencoded&quot;</span><span class="p">,</span> <span class="s2">&quot;Accept&quot;</span><span class="p">:</span> <span class="s2">&quot;text/plain&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;User-Agent&quot;</span><span class="p">:</span> <span class="s2">&quot;python&quot;</span><span class="p">}</span>
            <span class="n">st</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">http_post</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;auth_endpoint&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">tgt</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">params</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">st</span>

    <span class="n">instance</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">UMLS_AUTH</span><span class="o">.</span><span class="n">instance</span><span class="p">:</span>
            <span class="n">UMLS_AUTH</span><span class="o">.</span><span class="n">instance</span> <span class="o">=</span> <span class="n">UMLS_AUTH</span><span class="o">.</span><span class="n">__UMLS_AUTH</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">__getattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">instance</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span></div>

<span class="c1">#web service for retrieven UMLS CUI codes for labels (DisGeNET requires CUIs)</span>
<div class="viewcode-block" id="UMLS"><a class="viewcode-back" href="../comprior.html#knowledgebases.UMLS">[docs]</a><span class="k">class</span> <span class="nc">UMLS</span><span class="p">(</span><span class="n">REST</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Retrieves UMLS CUI codes for labels, which can then be used for querying DisGeNET.</span>

<span class="sd">       :param config: configuration parameters for UMLS web service (as specified in configuration file).</span>
<span class="sd">       :type config: dict</span>
<span class="sd">       :param auth: authentication component to generate a valid service ticket (required for every query).</span>
<span class="sd">       :type auth: :class:`UMLS_AUTH`</span>
<span class="sd">       &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">config</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfig</span><span class="p">(</span><span class="s2">&quot;UMLS&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">auth</span> <span class="o">=</span> <span class="n">UMLS_AUTH</span><span class="p">()</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="s2">&quot;UMLS&quot;</span><span class="p">,</span> <span class="n">url</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;webservice_uri&quot;</span><span class="p">])</span>

<div class="viewcode-block" id="UMLS.getCUIs"><a class="viewcode-back" href="../comprior.html#knowledgebases.UMLS.getCUIs">[docs]</a>    <span class="k">def</span> <span class="nf">getCUIs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get CUIs for the given labels.</span>

<span class="sd">           :param labels: list of identifiers for which to retrieve CUIs, e.g. disease names.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: list of CUIs.</span>
<span class="sd">           :rtype: list of str</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="c1"># do the actual request</span>
        <span class="n">cuis</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">:</span>
            <span class="c1">#get a valid service ticket</span>
            <span class="n">params</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ticket&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">auth</span><span class="o">.</span><span class="n">get_st</span><span class="p">()}</span>
            <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;/search/current?string=&quot;</span> <span class="o">+</span> <span class="n">label</span> <span class="o">+</span> <span class="s2">&quot;&amp;searchType=exact&quot;</span>

            <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">http_get</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">params</span> <span class="o">=</span> <span class="n">params</span><span class="p">)</span>

            <span class="c1">#get CUIs out of response</span>
            <span class="n">results</span> <span class="o">=</span> <span class="n">response</span><span class="p">[</span><span class="s2">&quot;result&quot;</span><span class="p">][</span><span class="s2">&quot;results&quot;</span><span class="p">]</span>
            <span class="n">label_cuis</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
                <span class="n">label_cuis</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;ui&quot;</span><span class="p">])</span>
            <span class="n">cuis</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">label_cuis</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">cuis</span><span class="p">)</span></div></div>

<div class="viewcode-block" id="DISGENET"><a class="viewcode-back" href="../comprior.html#knowledgebases.DISGENET">[docs]</a><span class="k">class</span> <span class="nc">DISGENET</span><span class="p">(</span><span class="n">REST</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Queries the DisGeNET web service for a given set of labels and retrieves association scores for all genes related to the query labels.</span>
<span class="sd">       DisGeNET API documentation: https://www.disgenet.org/api/</span>

<span class="sd">       :param umls: list of gene names to be mapped</span>
<span class="sd">       :type umls: :class:`UMLS` for transforming disease names to CUIs (required for query)</span>
<span class="sd">       &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">umls</span> <span class="o">=</span> <span class="n">UMLS</span><span class="p">()</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="s2">&quot;DisGeNET&quot;</span><span class="p">,</span> <span class="n">url</span><span class="o">=</span><span class="n">util</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;DisGeNET&quot;</span><span class="p">][</span><span class="s2">&quot;webservice_url&quot;</span><span class="p">])</span>

<div class="viewcode-block" id="DISGENET.getVersion"><a class="viewcode-back" href="../comprior.html#knowledgebases.DISGENET.getVersion">[docs]</a>    <span class="k">def</span> <span class="nf">getVersion</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get the current version of the DisGeNET API endpoint.</span>

<span class="sd">           :return: web service version infos.</span>
<span class="sd">           :rtype: json dict</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">ret</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">http_get</span><span class="p">(</span><span class="s2">&quot;/version&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ret</span></div>

<div class="viewcode-block" id="DISGENET.query"><a class="viewcode-back" href="../comprior.html#knowledgebases.DISGENET.query">[docs]</a>    <span class="k">def</span> <span class="nf">query</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Conducts the actual query to retrive gene-disease association scores for a given list of disease labels.</span>
<span class="sd">           Transforms the disease labels into CUIs before with the UMLS web service.</span>

<span class="sd">           :param labels: list of disease labels for which to retrieve gene-disease associations.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: DataFrame with gene-disease association scores.</span>
<span class="sd">           :rtype: :class:`pandas.DataFrame`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">cuis</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">umls</span><span class="o">.</span><span class="n">getCUIs</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>

        <span class="n">cui_string</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">cui_string</span> <span class="o">+=</span> <span class="s2">&quot;%2C&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cuis</span><span class="p">)</span>
        <span class="n">requestString</span> <span class="o">=</span> <span class="s2">&quot;/gda/disease/&quot;</span> <span class="o">+</span> <span class="n">cui_string</span>
        <span class="n">ret</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="c1">#retry query until we have to wait too long. DisGeNET sometimes seems to be quite slow</span>
        <span class="k">while</span> <span class="bp">self</span><span class="o">.</span><span class="n">TIMEOUT</span> <span class="o">&lt;</span> <span class="mi">250</span> <span class="ow">and</span> <span class="n">ret</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Trying to connect to DisGeNET...&quot;</span><span class="p">)</span>
            <span class="n">ret</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">http_get</span><span class="p">(</span><span class="n">requestString</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">TIMEOUT</span> <span class="o">+=</span> <span class="mi">30</span>

        <span class="c1">#check if 500 is returned</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ret</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="n">util</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;DisGeNET&quot;</span><span class="p">][</span><span class="s2">&quot;associationScore&quot;</span><span class="p">]])</span>

        <span class="c1"># bring the result into a readable format</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">ret</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span><span class="c1">#if the dataframe is empty, just return an empty dataframe</span>
            <span class="n">reduced_results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="n">util</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;DisGeNET&quot;</span><span class="p">][</span><span class="s2">&quot;associationScore&quot;</span><span class="p">]]]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="n">util</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;DisGeNET&quot;</span><span class="p">][</span><span class="s2">&quot;associationScore&quot;</span><span class="p">]])</span>

        <span class="n">reduced_results</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">]</span>

        <span class="c1">#save the results in an intermediate file</span>
        <span class="n">dir_name</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;DisGeNET&quot;</span><span class="p">][</span><span class="s2">&quot;outputDir&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s1">&#39;%Y-%m-</span><span class="si">%d</span><span class="s1">_%H-%M-%S&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
        <span class="n">util</span><span class="o">.</span><span class="n">createDirectory</span><span class="p">(</span><span class="n">dir_name</span><span class="p">)</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;queryResults.csv&quot;</span>

        <span class="n">reduced_results</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">dir_name</span> <span class="o">+</span> <span class="n">filename</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>

        <span class="c1">#map gene IDs to the desired format if necessary</span>
        <span class="n">desiredIDFormat</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;Dataset&quot;</span><span class="p">][</span><span class="s2">&quot;finalGeneIDFormat&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">desiredIDFormat</span> <span class="o">!=</span> <span class="s2">&quot;HGNC&quot;</span><span class="p">:</span>
            <span class="n">outputFile</span> <span class="o">=</span> <span class="n">dir_name</span> <span class="o">+</span> <span class="s2">&quot;mapped_&quot;</span> <span class="o">+</span> <span class="s2">&quot;HGNC_&quot;</span> <span class="o">+</span> <span class="n">desiredIDFormat</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">filename</span>
            <span class="n">util</span><span class="o">.</span><span class="n">createDirectory</span><span class="p">(</span><span class="n">dir_name</span> <span class="o">+</span> <span class="s2">&quot;mapped_&quot;</span> <span class="o">+</span> <span class="s2">&quot;HGNC_&quot;</span> <span class="o">+</span> <span class="n">desiredIDFormat</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span><span class="p">)</span>
            <span class="n">reduced_results</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">mapRanking</span><span class="p">(</span><span class="n">reduced_results</span><span class="p">,</span> <span class="s2">&quot;HGNC&quot;</span><span class="p">,</span> <span class="n">desiredIDFormat</span><span class="p">,</span> <span class="n">outputFile</span><span class="p">)</span>


        <span class="k">return</span> <span class="n">reduced_results</span></div></div>

<div class="viewcode-block" id="GCONVERT"><a class="viewcode-back" href="../comprior.html#knowledgebases.GCONVERT">[docs]</a><span class="k">class</span> <span class="nc">GCONVERT</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Queries the g:Convert web service to map a list of identifiers to a desired format.</span>
<span class="sd">       g:Convert makes use of the Ensembl build.</span>
<span class="sd">       g:Convert API documentation: https://biit.cs.ut.ee/gprofiler/page/apis</span>

<span class="sd">       :param url: API url as specified in the configuration file.</span>
<span class="sd">       :type url: str</span>
<span class="sd">       &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">url</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;gConvert&quot;</span><span class="p">][</span><span class="s2">&quot;webservice_url&quot;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;GCONVERT&quot;</span>

<div class="viewcode-block" id="GCONVERT.query"><a class="viewcode-back" href="../comprior.html#knowledgebases.GCONVERT.query">[docs]</a>    <span class="k">def</span> <span class="nf">query</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">items</span><span class="p">,</span> <span class="n">originalFormat</span><span class="p">,</span> <span class="n">desiredFormat</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Map a list of itendifiers to the desired format.</span>

<span class="sd">           :param items: list of identifiers to be mapped</span>
<span class="sd">           :type items: list of str</span>
<span class="sd">           :param originalFormat: current format of the identifiers</span>
<span class="sd">           :type originalFormat: str</span>
<span class="sd">           :param desiredFormat: desired identifier format</span>
<span class="sd">           :type desiredFormat: str</span>
<span class="sd">           :return: DataFrame containing the identifier mapping.</span>
<span class="sd">           :rtype: :class:`pandas.DataFrame`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">mapping</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">post</span><span class="p">(</span>
            <span class="n">url</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">url</span><span class="p">,</span>
            <span class="n">json</span><span class="o">=</span><span class="p">{</span>
                <span class="s1">&#39;organism&#39;</span><span class="p">:</span> <span class="s1">&#39;hsapiens&#39;</span><span class="p">,</span>
                <span class="s1">&#39;target&#39;</span><span class="p">:</span> <span class="n">desiredFormat</span><span class="p">,</span>
                <span class="s1">&#39;query&#39;</span><span class="p">:</span> <span class="n">items</span><span class="p">,</span>
            <span class="p">}</span>
        <span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">json</span><span class="p">()[</span><span class="s1">&#39;result&#39;</span><span class="p">]</span>
        <span class="n">desireds</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
        <span class="n">originals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">result</span><span class="p">:</span>
            <span class="n">desireds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;converted&quot;</span><span class="p">])</span>
            <span class="n">originals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;incoming&quot;</span><span class="p">])</span>
        <span class="n">mapping</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="n">originalFormat</span><span class="p">:</span> <span class="n">originals</span><span class="p">,</span> <span class="n">desiredFormat</span><span class="p">:</span> <span class="n">desireds</span><span class="p">})</span>
        <span class="k">return</span> <span class="n">mapping</span></div></div>



<div class="viewcode-block" id="PATHWAYCOMMONSWS"><a class="viewcode-back" href="../comprior.html#knowledgebases.PATHWAYCOMMONSWS">[docs]</a><span class="k">class</span> <span class="nc">PATHWAYCOMMONSWS</span><span class="p">(</span><span class="n">REST</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Queries the PathwayCommons web service.</span>
<span class="sd">       Bioservices&#39; existing implementation to query PathwayCommons was not used because it contained outdated values for _valid_formats for pathway retrieval,</span>
<span class="sd">       so we used the original code and adapted it to work correctly.</span>
<span class="sd">       &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">easyXMLConversion</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_default_extension</span> <span class="o">=</span> <span class="s2">&quot;json&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="s2">&quot;PathwayCommonsWS&quot;</span><span class="p">,</span> <span class="n">url</span><span class="o">=</span><span class="n">util</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;PathwayCommons&quot;</span><span class="p">][</span><span class="s2">&quot;webservice_url&quot;</span><span class="p">])</span>

    <span class="n">_valid_direction</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;BOTHSTREAM&quot;</span><span class="p">,</span> <span class="s2">&quot;DOWNSTREAM&quot;</span><span class="p">,</span> <span class="s2">&quot;UPSTREAM&quot;</span><span class="p">]</span>
<div class="viewcode-block" id="PATHWAYCOMMONSWS.getVersion"><a class="viewcode-back" href="../comprior.html#knowledgebases.PATHWAYCOMMONSWS.getVersion">[docs]</a>    <span class="k">def</span> <span class="nf">getVersion</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Map a list of genes to the desired format.</span>

<span class="sd">           :param genes: list of gene names to be mapped</span>
<span class="sd">           :type genes: list of str</span>
<span class="sd">           :return: list of mapped gene names.</span>
<span class="sd">           :rtype: list of str</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">ret</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">http_get</span><span class="p">(</span><span class="s2">&quot;/version&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ret</span></div>

    <span class="c1"># just a get/set to the default extension</span>
    <span class="k">def</span> <span class="nf">_set_default_ext</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ext</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Map a list of genes to the desired format.</span>

<span class="sd">           :param genes: list of gene names to be mapped</span>
<span class="sd">           :type genes: list of str</span>
<span class="sd">           :return: list of mapped gene names.</span>
<span class="sd">           :rtype: list of str</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">devtools</span><span class="o">.</span><span class="n">check_param_in_list</span><span class="p">(</span><span class="n">ext</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;json&quot;</span><span class="p">,</span> <span class="s2">&quot;xml&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_default_extension</span> <span class="o">=</span> <span class="n">ext</span>

    <span class="k">def</span> <span class="nf">_get_default_ext</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Map a list of genes to the desired format.</span>

<span class="sd">           :param genes: list of gene names to be mapped</span>
<span class="sd">           :type genes: list of str</span>
<span class="sd">           :return: list of mapped gene names.</span>
<span class="sd">           :rtype: list of str</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_default_extension</span>

    <span class="n">default_extension</span> <span class="o">=</span> <span class="nb">property</span><span class="p">(</span><span class="n">_get_default_ext</span><span class="p">,</span> <span class="n">_set_default_ext</span><span class="p">,</span>
                                     <span class="n">doc</span><span class="o">=</span><span class="s2">&quot;set extension of the requests (default is json). Can be &#39;json&#39; or &#39;xml&#39;&quot;</span><span class="p">)</span>

<div class="viewcode-block" id="PATHWAYCOMMONSWS.search"><a class="viewcode-back" href="../comprior.html#knowledgebases.PATHWAYCOMMONSWS.search">[docs]</a>    <span class="k">def</span> <span class="nf">search</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">q</span><span class="p">,</span> <span class="n">page</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">datasource</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Text search in PathwayCommons using Lucene query syntax</span>

<span class="sd">        Some of the parameters are BioPAX properties, others are composite</span>
<span class="sd">        relationships.</span>

<span class="sd">        All index fields are (case-sensitive): comment, ecnumber,</span>
<span class="sd">        keyword, name, pathway, term, xrefdb, xrefid, dataSource, and organism.</span>

<span class="sd">        The pathway field maps to all participants of pathways that contain</span>
<span class="sd">        the keyword(s) in any of its text fields.</span>

<span class="sd">        Finally, keyword is a transitive aggregate field that includes all</span>
<span class="sd">        searchable keywords of that element and its child elements.</span>

<span class="sd">        All searches can also be filtered by data source and organism.</span>

<span class="sd">        It is also possible to restrict the domain class using the</span>
<span class="sd">        &#39;type&#39; parameter.</span>

<span class="sd">        This query can be used standalone or to retrieve starting points</span>
<span class="sd">        for graph searches.</span>


<span class="sd">        :param str q: requires a keyword , name, external identifier, or a</span>
<span class="sd">            Lucene query string.</span>
<span class="sd">        :param int page: (N&gt;=0, default is 0), search result page number.</span>
<span class="sd">        :param str datasource: filter by data source (use names or URIs of</span>
<span class="sd">            pathway data sources or of any existing Provenance object). If</span>
<span class="sd">            multiple data source values are specified, a union of hits from</span>
<span class="sd">            specified sources is returned. datasource=[reactome,pid] returns</span>
<span class="sd">            hits associated with Reactome or PID.</span>
<span class="sd">        :param str organism: The organism can be specified either by</span>
<span class="sd">            official name, e.g. &quot;homo sapiens&quot; or by NCBI taxonomy id,</span>
<span class="sd">            e.g. &quot;9606&quot;. Similar to data sources, if multiple organisms</span>
<span class="sd">            are declared a union of all hits from specified organisms</span>
<span class="sd">            is returned. For example organism=[9606, 10016] returns results</span>
<span class="sd">            for both human and mice.</span>
<span class="sd">        :param str type: BioPAX class filter</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_extension</span> <span class="o">==</span> <span class="s2">&quot;xml&quot;</span><span class="p">:</span>
            <span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;search.xml?q=</span><span class="si">%s</span><span class="s2">&quot;</span>  <span class="o">%</span> <span class="n">q</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_extension</span> <span class="o">==</span> <span class="s2">&quot;json&quot;</span><span class="p">:</span>
            <span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;search.json?q=</span><span class="si">%s</span><span class="s2">&quot;</span>  <span class="o">%</span> <span class="n">q</span>

        <span class="n">params</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="n">page</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
            <span class="n">params</span><span class="p">[</span><span class="s1">&#39;page&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">page</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;page should be &gt;=0&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">datasource</span><span class="p">:</span>
            <span class="n">params</span><span class="p">[</span><span class="s1">&#39;datasource&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">datasource</span>

        <span class="k">if</span> <span class="nb">type</span><span class="p">:</span>
            <span class="n">params</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">type</span>

        <span class="k">if</span> <span class="n">organism</span><span class="p">:</span>
            <span class="n">params</span><span class="p">[</span><span class="s1">&#39;organism&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">organism</span>

        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">http_get</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">frmt</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">default_extension</span><span class="p">,</span>
                <span class="n">params</span><span class="o">=</span><span class="n">params</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_extension</span> <span class="o">==</span> <span class="s2">&quot;xml&quot;</span><span class="p">:</span>
            <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">easyXML</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">res</span></div>


<div class="viewcode-block" id="PATHWAYCOMMONSWS.get"><a class="viewcode-back" href="../comprior.html#knowledgebases.PATHWAYCOMMONSWS.get">[docs]</a>    <span class="k">def</span> <span class="nf">get</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">uri</span><span class="p">,</span> <span class="n">frmt</span><span class="o">=</span><span class="s2">&quot;BIOPAX&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Retrieves full pathway information for a set of elements</span>

<span class="sd">        elements can be for example pathway, interaction or physical</span>
<span class="sd">        entity given the RDF IDs. Get commands only</span>
<span class="sd">        retrieve the BioPAX elements that are directly mapped to the ID.</span>
<span class="sd">        Use the :meth:`traverse` query to traverse BioPAX graph and</span>
<span class="sd">        obtain child/owner elements.</span>

<span class="sd">        :param str uri: valid/existing BioPAX element&#39;s URI (RDF ID; for</span>
<span class="sd">            utility classes that were &quot;normalized&quot;, such as entity refereneces</span>
<span class="sd">            and controlled vocabularies, it is usually a Identifiers.org URL.</span>
<span class="sd">            Multiple IDs can be provided using list</span>
<span class="sd">            uri=[http://identifiers.org/uniprot/Q06609,</span>
<span class="sd">            http://identifiers.org/uniprot/Q549Z0&#39;]</span>
<span class="sd">            See also about MIRIAM and Identifiers.org.</span>
<span class="sd">        :param str format: output format (values)</span>

<span class="sd">        :return: a complete BioPAX representation for the record</span>
<span class="sd">            pointed to by the given URI is returned. Other output</span>
<span class="sd">            formats are produced by converting the BioPAX record on</span>
<span class="sd">            demand and can be specified by the optional format</span>
<span class="sd">            parameter. Please be advised that with some output formats</span>
<span class="sd">            it might return &quot;no result found&quot; error if the conversion is</span>
<span class="sd">            not applicable for the BioPAX result. For example,</span>
<span class="sd">            BINARY_SIF output usually works if there are some</span>
<span class="sd">            interactions, complexes, or pathways in the retrieved set</span>
<span class="sd">            and not only physical entities.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># validates the URIs</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">uri</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;get?uri=&quot;</span> <span class="o">+</span><span class="n">uri</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">uri</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;get?uri=&quot;</span> <span class="o">+</span><span class="n">uri</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">uri</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">uri</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
                    <span class="n">url</span> <span class="o">+=</span> <span class="s2">&quot;&amp;uri=&quot;</span> <span class="o">+</span> <span class="n">u</span>

        <span class="k">if</span> <span class="n">frmt</span> <span class="o">!=</span> <span class="s2">&quot;BIOPAX&quot;</span><span class="p">:</span>
            <span class="n">url</span> <span class="o">+=</span> <span class="s2">&quot;&amp;format=</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">frmt</span>

        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">http_get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">res</span></div></div>


<div class="viewcode-block" id="KnowledgeBaseFactory"><a class="viewcode-back" href="../comprior.html#knowledgebases.KnowledgeBaseFactory">[docs]</a><span class="k">class</span> <span class="nc">KnowledgeBaseFactory</span><span class="p">():</span><span class="c1">#singleton class</span>
    <span class="sd">&quot;&quot;&quot;Singleton class.</span>
<span class="sd">       Python code encapsulates it in a way that is not shown in Sphinx, so have a look at the descriptions in the source code.</span>

<span class="sd">       Creates knowledge bases based on the provided name and creates all corresponding objects, e.g. web service endpoints.</span>
<span class="sd">       Every knowledge base implementation must be registered here, otherwise it will not be accessible.</span>
<span class="sd">       &quot;&quot;&quot;</span>
    <span class="k">class</span> <span class="nc">__KnowledgeBaseFactory</span><span class="p">():</span>
        <span class="k">def</span> <span class="nf">createKnowledgeBase</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">knowledgebase</span><span class="p">):</span>
            <span class="sd">&quot;&quot;&quot;Creates knowledge base based on a given name.</span>

<span class="sd">               :param knowledgebase: name of the knowledge base to be created.</span>
<span class="sd">               :type knowledgebase: str</span>
<span class="sd">               :return: knowledge base object.</span>
<span class="sd">               :rtype: :class:`KnowledgeBase` or inheriting classes</span>
<span class="sd">               &quot;&quot;&quot;</span>
            <span class="k">if</span> <span class="n">knowledgebase</span> <span class="o">==</span> <span class="s2">&quot;DisGeNET&quot;</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">Disgenet</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">knowledgebase</span> <span class="o">==</span> <span class="s2">&quot;KEGG&quot;</span><span class="p">:</span>
                <span class="c1">#create a pathway mapper</span>
                <span class="n">pathwayparser</span> <span class="o">=</span> <span class="n">KEGGPathwayParser</span><span class="p">()</span>
                <span class="k">return</span> <span class="n">Kegg</span><span class="p">(</span><span class="n">pathwayparser</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">knowledgebase</span> <span class="o">==</span> <span class="s2">&quot;Enrichr&quot;</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">Enrichr</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">knowledgebase</span> <span class="o">==</span> <span class="s2">&quot;OpenTargets&quot;</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">OpenTargets</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">knowledgebase</span> <span class="o">==</span> <span class="s2">&quot;PathwayCommons&quot;</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">Pathwaycommons</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">knowledgebase</span> <span class="o">==</span> <span class="s2">&quot;Biomart&quot;</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">BioMART</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">knowledgebase</span> <span class="o">==</span> <span class="s2">&quot;gConvert&quot;</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">Gconvert</span><span class="p">()</span>

            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The listed knowledge base is not available. See the documentation for available knowledge bases.&quot;</span><span class="p">)</span>
            <span class="n">exit</span><span class="p">()</span>

    <span class="n">instance</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">KnowledgeBaseFactory</span><span class="o">.</span><span class="n">instance</span><span class="p">:</span>
            <span class="n">KnowledgeBaseFactory</span><span class="o">.</span><span class="n">instance</span> <span class="o">=</span> <span class="n">KnowledgeBaseFactory</span><span class="o">.</span><span class="n">__KnowledgeBaseFactory</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">__getattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">instance</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span></div>


<span class="c1">############################### KNOWLEDGE BASES ###############################</span>

<div class="viewcode-block" id="KnowledgeBase"><a class="viewcode-back" href="../comprior.html#knowledgebases.KnowledgeBase">[docs]</a><span class="k">class</span> <span class="nc">KnowledgeBase</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Super class for every knowledge base implementation.</span>
<span class="sd">       If a new knowledge base is implemented, it must inherit from this class and implement methods :meth:`KnowledgeBase.getRelevantGenes()`, :meth:`KnowledgeBase.getGeneScores()`, and :meth:`KnowledgeBase.getRelevantPathways()`.</span>

<span class="sd">       :param name: name of the knowledge base</span>
<span class="sd">       :type name: str</span>
<span class="sd">       :param config: configuration parameter of the knowledge base as specified in the config file.</span>
<span class="sd">       :type config: dict</span>
<span class="sd">       :param webservice: web service querying object</span>
<span class="sd">       :type webservice: :class:`bioservices.REST` or inheriting classes.</span>
<span class="sd">       :param hasGeneInformation: true if the knowledge base provides gene association information, false otherwise</span>
<span class="sd">       :type hasGeneInformation: bool</span>
<span class="sd">       :param hasPathwayInformation: true if the knowledge base also provides pathway information, false otherwise</span>
<span class="sd">       :type hasPathwayInformation: bool</span>
<span class="sd">       &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">kb_config</span><span class="p">,</span> <span class="n">webservice</span><span class="p">,</span> <span class="n">geneInfo</span><span class="p">,</span> <span class="n">pathwayInfo</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">config</span> <span class="o">=</span> <span class="n">kb_config</span>
        <span class="n">util</span><span class="o">.</span><span class="n">createDirectory</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;outputDir&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span> <span class="o">=</span> <span class="n">webservice</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hasGeneInformation</span> <span class="o">=</span> <span class="n">geneInfo</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hasPathwayInformation</span> <span class="o">=</span> <span class="n">pathwayInfo</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>

<div class="viewcode-block" id="KnowledgeBase.getRelevantGenes"><a class="viewcode-back" href="../comprior.html#knowledgebases.KnowledgeBase.getRelevantGenes">[docs]</a>    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">getRelevantGenes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Abstract.</span>
<span class="sd">           Get all genes that are associated to a list of labels, e.g. disease names.</span>

<span class="sd">           :param labels: list of labels for which to retrieve the genes.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: list of associated genes.</span>
<span class="sd">           :rtype: list of str</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">pass</span></div>

<div class="viewcode-block" id="KnowledgeBase.getGeneScores"><a class="viewcode-back" href="../comprior.html#knowledgebases.KnowledgeBase.getGeneScores">[docs]</a>    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">getGeneScores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Abstract.</span>
<span class="sd">           Get all genes and their association scores for a given list of disease names.</span>

<span class="sd">           :param labels: list of disease names for which to get gene-disease-association scores.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: DataFrame of genes and their association scores.</span>
<span class="sd">           :rtype: :class:`pandas.DataFrame`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">pass</span></div>

<div class="viewcode-block" id="KnowledgeBase.getRelevantPathways"><a class="viewcode-back" href="../comprior.html#knowledgebases.KnowledgeBase.getRelevantPathways">[docs]</a>    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">getRelevantPathways</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get all pathways related to a set of labels, e.g disease names.</span>

<span class="sd">           :param labels: list of labels for which to find related pathways.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: dict of pathway names and pathway representations.</span>
<span class="sd">           :rtype: dict with :class:`pypath.Network` as values</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">pass</span></div>

<div class="viewcode-block" id="KnowledgeBase.getName"><a class="viewcode-back" href="../comprior.html#knowledgebases.KnowledgeBase.getName">[docs]</a>    <span class="k">def</span> <span class="nf">getName</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns the name of the knowledge base.</span>

<span class="sd">           :return: knowledge base name.</span>
<span class="sd">           :rtype: str</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span></div>

<div class="viewcode-block" id="KnowledgeBase.hasPathways"><a class="viewcode-back" href="../comprior.html#knowledgebases.KnowledgeBase.hasPathways">[docs]</a>    <span class="k">def</span> <span class="nf">hasPathways</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns if knowledge base retrieves pathway information, i.e. if :meth:`KnowledgeBase.getRelevantPathways()` is implemented..</span>

<span class="sd">           :return: true if knowledge base provides pathway information, false otherwise.</span>
<span class="sd">           :rtype: bool</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">hasPathwayInformation</span></div>

<div class="viewcode-block" id="KnowledgeBase.hasGenes"><a class="viewcode-back" href="../comprior.html#knowledgebases.KnowledgeBase.hasGenes">[docs]</a>    <span class="k">def</span> <span class="nf">hasGenes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns if knowledge base retrieves gene information, i.e. if :meth:`KnowledgeBase.getRelevantGenes()` :meth:`KnowledgeBase.getGeneScores()` are implemented.</span>

<span class="sd">           :return: true if knowledge base provides gene information, false otherwise.</span>
<span class="sd">           :rtype: bool</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">hasGeneInformation</span></div></div>


<div class="viewcode-block" id="Enrichr"><a class="viewcode-back" href="../comprior.html#knowledgebases.Enrichr">[docs]</a><span class="k">class</span> <span class="nc">Enrichr</span><span class="p">(</span><span class="n">KnowledgeBase</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Special knowledge base not intended to be used by feature selection approaches.</span>
<span class="sd">       Instead, it is used for evaluation purposes to annotate and enrich rankings.</span>

<span class="sd">       :param name: name of the knowledge base</span>
<span class="sd">       :type name: str</span>
<span class="sd">       :param config: configuration parameter of the knowledge base as specified in the config file.</span>
<span class="sd">       :type config: dict</span>
<span class="sd">       :param webservice: web service querying object</span>
<span class="sd">       :type webservice: :class:`bioservices.REST` or inheriting classes.</span>
<span class="sd">       :param hasGeneInformation: true if the knowledge base provides gene association information, false otherwise</span>
<span class="sd">       :type hasGeneInformation: bool</span>
<span class="sd">       :param hasPathwayInformation: true if the knowledge base also provides pathway information, false otherwise</span>
<span class="sd">       :type hasPathwayInformation: bool</span>
<span class="sd">       &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="s2">&quot;Enrichr&quot;</span><span class="p">,</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfig</span><span class="p">(</span><span class="s2">&quot;Enrichr&quot;</span><span class="p">),</span> <span class="n">ENRICHR</span><span class="p">(),</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>

<div class="viewcode-block" id="Enrichr.downloadEnrichedTerms"><a class="viewcode-back" href="../comprior.html#knowledgebases.Enrichr.downloadEnrichedTerms">[docs]</a>    <span class="k">def</span> <span class="nf">downloadEnrichedTerms</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">userIdList</span><span class="p">,</span> <span class="n">filePrefix</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Downloads enriched terms from a former query into a file.</span>
<span class="sd">           Filters these terms for those with an adjusted p-value &gt; 0.05, then sorts by combined score in descending order.</span>

<span class="sd">           :param userIdList: userIdList to retrieve enrichment/annotation results from the original query.</span>
<span class="sd">           :type userIdList: str</span>
<span class="sd">           :param filePrefix: prefix to use in filename.</span>
<span class="sd">           :type filePrefix: str</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">geneSetLibrary</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;geneSetLibrary&quot;</span><span class="p">]</span>

        <span class="n">outputDir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;outputDir&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s1">&#39;%Y-%m-</span><span class="si">%d</span><span class="s1">_%H-%M-%S&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
        <span class="n">util</span><span class="o">.</span><span class="n">createDirectory</span><span class="p">(</span><span class="n">outputDir</span><span class="p">)</span>
        <span class="n">outputFile</span> <span class="o">=</span> <span class="n">filePrefix</span> <span class="o">+</span> <span class="s2">&quot;_enrichedTerms.txt&quot;</span>

        <span class="n">params</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;userListId&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">userIdList</span><span class="p">),</span> <span class="s2">&quot;backgroundType&quot;</span><span class="p">:</span> <span class="n">geneSetLibrary</span><span class="p">}</span>
        <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">enrich</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>
        <span class="c1"># first, filter enriched terms by q-score &gt; 0.05</span>
        <span class="c1"># second, order by combined score in descending order</span>
        <span class="c1"># see also this best practices recommendation: https://www.researchgate.net/post/Enrichr_what_value_of_combined_score_is_significant</span>
        <span class="n">final_terms</span> <span class="o">=</span> <span class="n">response</span><span class="p">[(</span><span class="n">response</span><span class="p">[</span><span class="s2">&quot;Adjusted p-value&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">)]</span>
        <span class="n">final_terms</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;Combined score&quot;</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

        <span class="n">final_terms</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">outputFile</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span></div>

<div class="viewcode-block" id="Enrichr.getRelevantGenes"><a class="viewcode-back" href="../comprior.html#knowledgebases.Enrichr.getRelevantGenes">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantGenes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Is not implemented for EnrichR.</span>

<span class="sd">           :param labels: list of gene names to be mapped</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: :class:`NotImplementedError` as this knowledge base is not intended to be used for such analyses.</span>
<span class="sd">           :rtype: :class:`NotImplementedError`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;EnrichR is currently only used for subsequent annotation and not intended to be used during analysis.&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="Enrichr.getGeneScores"><a class="viewcode-back" href="../comprior.html#knowledgebases.Enrichr.getGeneScores">[docs]</a>    <span class="k">def</span> <span class="nf">getGeneScores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Is not implemented for EnrichR.</span>

<span class="sd">           :param labels: list of gene names to be mapped</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: :class:`NotImplementedError` as this knowledge base is not intended to be used for such analyses.</span>
<span class="sd">           :rtype: :class:`NotImplementedError`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;EnrichR is currently only used for subsequent annotation and not intended to be used during analysis.&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="Enrichr.getRelevantPathways"><a class="viewcode-back" href="../comprior.html#knowledgebases.Enrichr.getRelevantPathways">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantPathways</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Is not implemented for EnrichR.</span>

<span class="sd">           :param labels: list of labels for which to find related pathways.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: :class:`NotImplementedError` as this knowledge base is not intended to be used for such analyses.</span>
<span class="sd">           :rtype: :class:`NotImplementedError`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;EnrichR is currently only used for subsequent annotation and not intended to be used during analysis.&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="Enrichr.enrichGeneset"><a class="viewcode-back" href="../comprior.html#knowledgebases.Enrichr.enrichGeneset">[docs]</a>    <span class="k">def</span> <span class="nf">enrichGeneset</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">geneList</span><span class="p">,</span> <span class="n">filePrefix</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Sends a list of identifies (here, genes) to EnrichR web service and stores all term enrichments in a file.</span>

<span class="sd">           :param geneList: list of gene names for which to retrieve enrichments.</span>
<span class="sd">           :type geneList: list of str</span>
<span class="sd">           :param filePrefix: prefix to use in file name (to store enrichments).</span>
<span class="sd">           :type filePrefix: str</span>
<span class="sd">           &quot;&quot;&quot;</span>

        <span class="c1">#submit list for analysis</span>
        <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">addlist</span><span class="p">(</span><span class="n">geneList</span><span class="p">)</span>
        <span class="n">userListID</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;userListId&quot;</span><span class="p">]</span>

        <span class="c1">#download files of enriched terms</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">downloadEnrichedTerms</span><span class="p">(</span><span class="n">userListID</span><span class="p">,</span> <span class="n">filePrefix</span><span class="p">)</span></div>

<div class="viewcode-block" id="Enrichr.annotateGene"><a class="viewcode-back" href="../comprior.html#knowledgebases.Enrichr.annotateGene">[docs]</a>    <span class="k">def</span> <span class="nf">annotateGene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Annotates a gene with terms.</span>

<span class="sd">           :param gene: gene name.</span>
<span class="sd">           :type gene: str</span>
<span class="sd">           :return: list of all annotations to the provided gene.</span>
<span class="sd">           :rtype: list of str</span>
<span class="sd">           &quot;&quot;&quot;</span>

        <span class="n">params</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;json&quot;</span><span class="p">:</span> <span class="s2">&quot;true&quot;</span><span class="p">,</span> <span class="s2">&quot;setup&quot;</span><span class="p">:</span> <span class="s2">&quot;true&quot;</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="p">}</span>

        <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">genemap</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">annotations</span> <span class="o">=</span> <span class="n">response</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
            <span class="n">geneSetLibrary</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;geneSetLibrary&quot;</span><span class="p">]</span>
            <span class="n">annotation</span> <span class="o">=</span> <span class="n">annotations</span><span class="p">[</span><span class="n">geneSetLibrary</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="c1">#no annotation was found</span>
            <span class="n">annotation</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">return</span> <span class="n">annotation</span></div>

<div class="viewcode-block" id="Enrichr.annotateGenes"><a class="viewcode-back" href="../comprior.html#knowledgebases.Enrichr.annotateGenes">[docs]</a>    <span class="k">def</span> <span class="nf">annotateGenes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">geneList</span><span class="p">,</span> <span class="n">filePrefix</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Annotates a list of genes with relevant terms.</span>

<span class="sd">           :param geneList: list of gene names to annotate.</span>
<span class="sd">           :type geneList: list of str</span>
<span class="sd">           :param filePrefix: prefix to use when storing results in a file.</span>
<span class="sd">           :type filePrefix: str</span>
<span class="sd">           :return: dict of gene names and lists of their annotations.</span>
<span class="sd">           :rtype: dict</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">annotations</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="c1"># if false then annotate each gene individually</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">geneList</span><span class="p">:</span>
            <span class="n">annotations</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotateGene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="n">outputDir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;outputDir&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s1">&#39;%Y-%m-</span><span class="si">%d</span><span class="s1">_%H-%M-%S&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
        <span class="n">util</span><span class="o">.</span><span class="n">createDirectory</span><span class="p">(</span><span class="n">outputDir</span><span class="p">)</span>
        <span class="n">outputFile</span> <span class="o">=</span> <span class="n">filePrefix</span> <span class="o">+</span> <span class="s2">&quot;_annotatedGenes.txt&quot;</span>

        <span class="c1"># write gene annotations to file</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outputFile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;attributeName</span><span class="se">\t</span><span class="s2">annotation</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">annotation</span> <span class="ow">in</span> <span class="n">annotations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">anno</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">annotation</span><span class="p">)</span>
                <span class="n">g</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">anno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">anno</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">annotations</span></div></div>

<div class="viewcode-block" id="BioMART"><a class="viewcode-back" href="../comprior.html#knowledgebases.BioMART">[docs]</a><span class="k">class</span> <span class="nc">BioMART</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Maps a identifiers or data sets with identifiers to the desired format by using BiomaRt.</span>
<span class="sd">       Wrapper class that internally invokes BiomaRt&#39;s R code.</span>
<span class="sd">       Very unstable, so currently not used.</span>
<span class="sd">       However, it can be exchanged in :meth:`benchutils.retrieveMappings()` function.</span>
<span class="sd">       &quot;&quot;&quot;</span>

<div class="viewcode-block" id="BioMART.mapItems"><a class="viewcode-back" href="../comprior.html#knowledgebases.BioMART.mapItems">[docs]</a>    <span class="k">def</span> <span class="nf">mapItems</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">itemList</span><span class="p">,</span> <span class="n">originalFormat</span><span class="p">,</span> <span class="n">desiredFormat</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Map a list of identifiers to the desired format.</span>
<span class="sd">           Internally invokes external R code that uses the BiomaRt package.</span>

<span class="sd">           :param itemList: list of identifiers to be mapped</span>
<span class="sd">           :type itemList: list of str</span>
<span class="sd">           :param originalFormat: original identifier format.</span>
<span class="sd">           :type originalFormat: str</span>
<span class="sd">           :param desiredFormat: format to which to map identifiers.</span>
<span class="sd">           :type desiredFormat: str</span>
<span class="sd">           :return: mapping data frame of identifiers (with original and desired format)</span>
<span class="sd">           :rtype: :class:`pandas.DataFrame`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="c1"># write results into intermediate file</span>
        <span class="n">outputDir</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfigValue</span><span class="p">(</span><span class="s2">&quot;Biomart&quot;</span><span class="p">,</span> <span class="s2">&quot;outputDir&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s1">&#39;%Y-%m-</span><span class="si">%d</span><span class="s1">_%H-%M-%S&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
        <span class="n">util</span><span class="o">.</span><span class="n">createDirectory</span><span class="p">(</span><span class="n">outputDir</span><span class="p">)</span>
        <span class="n">util</span><span class="o">.</span><span class="n">createDirectory</span><span class="p">(</span><span class="n">outputDir</span> <span class="o">+</span> <span class="s2">&quot;input/&quot;</span><span class="p">)</span>
        <span class="n">util</span><span class="o">.</span><span class="n">createDirectory</span><span class="p">(</span><span class="n">outputDir</span> <span class="o">+</span> <span class="s2">&quot;output/&quot;</span><span class="p">)</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;mapping.csv&quot;</span>

        <span class="c1">#put itemList into input file</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outputDir</span> <span class="o">+</span> <span class="s2">&quot;input/&quot;</span> <span class="o">+</span> <span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">itemList</span><span class="p">:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">item</span><span class="p">)</span>

        <span class="n">params</span> <span class="o">=</span> <span class="p">[</span><span class="n">originalFormat</span><span class="p">,</span> <span class="n">desiredFormat</span><span class="p">,</span> <span class="n">outputDir</span> <span class="o">+</span> <span class="s2">&quot;input/&quot;</span> <span class="o">+</span> <span class="n">filename</span><span class="p">,</span> <span class="n">outputDir</span> <span class="o">+</span> <span class="s2">&quot;output/&quot;</span> <span class="o">+</span> <span class="n">filename</span><span class="p">]</span>
        <span class="n">util</span><span class="o">.</span><span class="n">runRCommand</span><span class="p">(</span><span class="n">util</span><span class="o">.</span><span class="n">getConfig</span><span class="p">(</span><span class="s2">&quot;R&quot;</span><span class="p">),</span> <span class="s2">&quot;IdentifierMapping.R&quot;</span><span class="p">,</span> <span class="n">params</span><span class="p">)</span>
        <span class="n">mappedItems</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">outputDir</span> <span class="o">+</span> <span class="s2">&quot;output/&quot;</span> <span class="o">+</span> <span class="n">filename</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">mappedItems</span></div>

<div class="viewcode-block" id="BioMART.getRelevantGenes"><a class="viewcode-back" href="../comprior.html#knowledgebases.BioMART.getRelevantGenes">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantGenes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Is not implemented for BiomaRt.</span>

<span class="sd">           :param labels: list of gene names to be mapped</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: :class:`NotImplementedError` as this knowledge base is not intended to be used for such analyses.</span>
<span class="sd">           :rtype: :class:`NotImplementedError`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;Biomart is currently only used for identifier mapping and not intended to be used during analysis.&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="BioMART.getGeneScores"><a class="viewcode-back" href="../comprior.html#knowledgebases.BioMART.getGeneScores">[docs]</a>    <span class="k">def</span> <span class="nf">getGeneScores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Is not implemented for BiomaRt.</span>

<span class="sd">           :param labels: list of gene names to be mapped</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: :class:`NotImplementedError` as this knowledge base is not intended to be used for such analyses.</span>
<span class="sd">           :rtype: :class:`NotImplementedError`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
            <span class="s2">&quot;Biomart is currently only used for identifier mapping and not intended to be used during analysis.&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="BioMART.getRelevantPathways"><a class="viewcode-back" href="../comprior.html#knowledgebases.BioMART.getRelevantPathways">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantPathways</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Is not implemented for BiomaRt.</span>

<span class="sd">           :param labels: list of labels for which to find related pathways.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: :class:`NotImplementedError` as this knowledge base is not intended to be used for such analyses.</span>
<span class="sd">           :rtype: :class:`NotImplementedError`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
            <span class="s2">&quot;Biomart is currently only used for identifier mapping and not intended to be used during analysis.&quot;</span><span class="p">)</span></div></div>

<div class="viewcode-block" id="Gconvert"><a class="viewcode-back" href="../comprior.html#knowledgebases.Gconvert">[docs]</a><span class="k">class</span> <span class="nc">Gconvert</span><span class="p">(</span><span class="n">KnowledgeBase</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Maps identifiers or data sets containing identifiers to the desired format by using the g:Convert web service.</span>

<span class="sd">       :param name: name of the knowledge base</span>
<span class="sd">       :type name: str</span>
<span class="sd">       :param config: configuration parameter of the knowledge base as specified in the config file.</span>
<span class="sd">       :type config: dict</span>
<span class="sd">       :param webservice: web service querying object.</span>
<span class="sd">       :type webservice: :class:`bioservices.REST` or inheriting classes</span>
<span class="sd">       :param hasGeneInformation: true if the knowledge base provides gene association information, false otherwise</span>
<span class="sd">       :type hasGeneInformation: bool</span>
<span class="sd">       :param hasPathwayInformation: true if the knowledge base also provides pathway information, false otherwise</span>
<span class="sd">       :type hasPathwayInformation: bool</span>
<span class="sd">       &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="s2">&quot;gConvert&quot;</span><span class="p">,</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfig</span><span class="p">(</span><span class="s2">&quot;gConvert&quot;</span><span class="p">),</span> <span class="n">GCONVERT</span><span class="p">(),</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>


<div class="viewcode-block" id="Gconvert.mapItems"><a class="viewcode-back" href="../comprior.html#knowledgebases.Gconvert.mapItems">[docs]</a>    <span class="k">def</span> <span class="nf">mapItems</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">itemList</span><span class="p">,</span> <span class="n">originalFormat</span><span class="p">,</span> <span class="n">desiredFormat</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Map a list of identifiers to the desired format.</span>

<span class="sd">           :param itemList: list of identifiers to be mapped.</span>
<span class="sd">           :type itemList: list of str</span>
<span class="sd">           :param originalFormat: current format of the identifiers.</span>
<span class="sd">           :type originalFormat: str</span>
<span class="sd">           :param desiredFormat: desired format to which to map identifiers.</span>
<span class="sd">           :type desiredFormat: str</span>
<span class="sd">           :return: DataFrame table containing mappings of the identifiers from original to desired format.</span>
<span class="sd">           :rtype: :class:`pandas.DataFrame`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="c1"># write results into intermediate file</span>
        <span class="n">mapping</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">itemList</span><span class="p">,</span> <span class="n">originalFormat</span><span class="p">,</span> <span class="n">desiredFormat</span><span class="p">)</span>
        <span class="c1">#remove nans in mapping</span>
        <span class="n">indexNames</span> <span class="o">=</span> <span class="n">mapping</span><span class="p">[</span><span class="n">mapping</span><span class="p">[</span><span class="n">desiredFormat</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;nan&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>
        <span class="n">mapping</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">indexNames</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">mapping</span></div>

<div class="viewcode-block" id="Gconvert.getRelevantGenes"><a class="viewcode-back" href="../comprior.html#knowledgebases.Gconvert.getRelevantGenes">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantGenes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Is not implemented for g:Convert.</span>

<span class="sd">           :param labels: list of gene names to be mapped</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: :class:`NotImplementedError` as this knowledge base is not intended to be used for such analyses.</span>
<span class="sd">           :rtype: :class:`NotImplementedError`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;g:Convert is currently only used for identifier mapping and not intended to be used during analysis.&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="Gconvert.getGeneScores"><a class="viewcode-back" href="../comprior.html#knowledgebases.Gconvert.getGeneScores">[docs]</a>    <span class="k">def</span> <span class="nf">getGeneScores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Is not implemented for g:Convert.</span>

<span class="sd">           :param labels: list of gene names to be mapped</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: :class:`NotImplementedError` as this knowledge base is not intended to be used for such analyses.</span>
<span class="sd">           :rtype: :class:`NotImplementedError`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
            <span class="s2">&quot;g:Convert is currently only used for identifier mapping and not intended to be used during analysis.&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="Gconvert.getRelevantPathways"><a class="viewcode-back" href="../comprior.html#knowledgebases.Gconvert.getRelevantPathways">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantPathways</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Is not implemented for g:Convert.</span>

<span class="sd">           :param labels: list of labels for which to find related pathways.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: :class:`NotImplementedError` as this knowledge base is not intended to be used for such analyses.</span>
<span class="sd">           :rtype: :class:`NotImplementedError`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
            <span class="s2">&quot;g:Convert is currently only used for identifier mapping and not intended to be used during analysis.&quot;</span><span class="p">)</span></div></div>



<div class="viewcode-block" id="OpenTargets"><a class="viewcode-back" href="../comprior.html#knowledgebases.OpenTargets">[docs]</a><span class="k">class</span> <span class="nc">OpenTargets</span><span class="p">(</span><span class="n">KnowledgeBase</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Knowledge base implementation of OpenTargets.</span>
<span class="sd">       Uses the OpenTargetsClient Python implementation provided by OpenTargets to query the web service API.</span>

<span class="sd">       :param name: name of the knowledge base</span>
<span class="sd">       :type name: str</span>
<span class="sd">       :param config: configuration parameter of the knowledge base as specified in the config file.</span>
<span class="sd">       :type config: dict</span>
<span class="sd">       :param webservice: web service querying implementation.</span>
<span class="sd">       :type webservice: :class:`opentargets.OpenTargetsClient`</span>
<span class="sd">       :param hasGeneInformation: true if the knowledge base provides gene association information, false otherwise</span>
<span class="sd">       :type hasGeneInformation: bool</span>
<span class="sd">       :param hasPathwayInformation: true if the knowledge base also provides pathway information, false otherwise</span>
<span class="sd">       :type hasPathwayInformation: bool</span>
<span class="sd">       &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="s2">&quot;OpenTargets&quot;</span><span class="p">,</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfig</span><span class="p">(</span><span class="s2">&quot;OpenTargets&quot;</span><span class="p">),</span> <span class="n">OpenTargetsClient</span><span class="p">(),</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>

<div class="viewcode-block" id="OpenTargets.getAssociations"><a class="viewcode-back" href="../comprior.html#knowledgebases.OpenTargets.getAssociations">[docs]</a>    <span class="k">def</span> <span class="nf">getAssociations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get all relevant information for a given set of labels, sorted by their association scores in descending order.</span>
<span class="sd">           Writes web service results into an intermediate file and maps the identifiers to have the correct format for further processing.</span>

<span class="sd">           :param labels: list of labels, e.g. disease names.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: DataFrame containing all related genes and their association scores.</span>
<span class="sd">           :rtype: :class:`pandas.DataFrame`</span>
<span class="sd">           &quot;&quot;&quot;</span>

        <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">]</span>
        <span class="n">associated_genes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">cols</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">a_for_disease</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">get_associations_for_disease</span><span class="p">(</span><span class="n">term</span><span class="p">)</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;SEEMS SOME ERROR OCCURED FOR &quot;</span> <span class="o">+</span> <span class="n">term</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="c1">#check if an error code was returned</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">a_for_disease</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ERROR: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">a_for_disease</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; RETURNED.&quot;</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="n">geneIDs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
            <span class="n">assocscores</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">a_for_disease</span><span class="p">:</span>
                <span class="n">geneID</span> <span class="o">=</span> <span class="n">a</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
                <span class="c1">#the ID is currently attached by an EFO ID - remove it</span>
                <span class="n">geneID</span> <span class="o">=</span> <span class="n">geneID</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">geneIDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">geneID</span><span class="p">)</span>
                <span class="n">score</span> <span class="o">=</span> <span class="n">a</span><span class="p">[</span><span class="s1">&#39;association_score&#39;</span><span class="p">][</span><span class="s1">&#39;overall&#39;</span><span class="p">]</span>
                <span class="n">assocscores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>

            <span class="n">associated_genes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">:</span> <span class="n">geneIDs</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">:</span> <span class="n">assocscores</span><span class="p">})</span>

        <span class="c1">#write results into intermediate file</span>
        <span class="n">outputDir</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfigValue</span><span class="p">(</span><span class="s2">&quot;OpenTargets&quot;</span><span class="p">,</span> <span class="s2">&quot;outputDir&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s1">&#39;%Y-%m-</span><span class="si">%d</span><span class="s1">_%H-%M-%S&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
        <span class="n">util</span><span class="o">.</span><span class="n">createDirectory</span><span class="p">(</span><span class="n">outputDir</span><span class="p">)</span>
        <span class="n">outputFile</span> <span class="o">=</span> <span class="s2">&quot;genes.csv&quot;</span>
        <span class="n">associated_genes</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">outputDir</span> <span class="o">+</span> <span class="n">outputFile</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
        <span class="c1"># check if gene ID format from expression data set matches OpenTargets&#39; Ensembl Gene ID format</span>
        <span class="n">desiredIDFormat</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;Dataset&quot;</span><span class="p">][</span><span class="s2">&quot;finalGeneIDFormat&quot;</span><span class="p">]</span>
        <span class="c1">#if desiredIDFormat != &quot;ensembl_gene_id&quot;:</span>
        <span class="k">if</span> <span class="n">desiredIDFormat</span> <span class="o">!=</span> <span class="s2">&quot;ENSG&quot;</span><span class="p">:</span>
            <span class="n">outputFile</span> <span class="o">=</span> <span class="n">outputDir</span> <span class="o">+</span> <span class="s2">&quot;mapped_&quot;</span> <span class="o">+</span> <span class="s2">&quot;ENSG_&quot;</span> <span class="o">+</span> <span class="n">desiredIDFormat</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">outputFile</span>
            <span class="n">util</span><span class="o">.</span><span class="n">createDirectory</span><span class="p">(</span><span class="n">outputDir</span> <span class="o">+</span> <span class="s2">&quot;mapped_&quot;</span> <span class="o">+</span> <span class="s2">&quot;ENSG_&quot;</span> <span class="o">+</span> <span class="n">desiredIDFormat</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span><span class="p">)</span>
            <span class="n">associated_genes</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">mapRanking</span><span class="p">(</span><span class="n">associated_genes</span><span class="p">,</span> <span class="s2">&quot;ENSG&quot;</span><span class="p">,</span> <span class="n">desiredIDFormat</span><span class="p">,</span> <span class="n">outputFile</span><span class="p">)</span>

        <span class="c1">#order by score</span>
        <span class="n">associated_genes</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;score&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1">#remove duplicates</span>
        <span class="n">associated_genes</span> <span class="o">=</span> <span class="n">associated_genes</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s2">&quot;gene_symbol&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">associated_genes</span></div>

<div class="viewcode-block" id="OpenTargets.getRelevantGenes"><a class="viewcode-back" href="../comprior.html#knowledgebases.OpenTargets.getRelevantGenes">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantGenes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get all genes that are somehow associated to the given labels, e.g. disease names.</span>

<span class="sd">           :param labels: list of identifiers, e.g. disease names, for which to find associated genes.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: list of associated genes.</span>
<span class="sd">           :rtype: list of str</span>
<span class="sd">           &quot;&quot;&quot;</span>

        <span class="n">genes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">getAssociations</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">genes</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span></div>

<div class="viewcode-block" id="OpenTargets.getGeneScores"><a class="viewcode-back" href="../comprior.html#knowledgebases.OpenTargets.getGeneScores">[docs]</a>    <span class="k">def</span> <span class="nf">getGeneScores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get all genes and their association scores that are related to the given labels, e.g. disease names.</span>

<span class="sd">           :param labels: list of identifiers, e.g. disease names, for which to find associated genes.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: DataFrame of associated genes and their association scores, in descending order.</span>
<span class="sd">           :rtype: :class:`pandas.DataFrame`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">geneScores</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">getAssociations</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">geneScores</span></div>


<div class="viewcode-block" id="OpenTargets.getRelevantPathways"><a class="viewcode-back" href="../comprior.html#knowledgebases.OpenTargets.getRelevantPathways">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantPathways</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;As OpenTargets currently does not provide pathway information, this feature is not implemented for OpenTargets.</span>

<span class="sd">           :param labels: list of labels for which to find related pathways.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: :class:`NotImplementedError` as this knowledge base is not intended to be used for such analyses.</span>
<span class="sd">           :rtype: :class:`NotImplementedError`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;OpenTargets cannot return pathways at the moment (although they also provide references to pathways, but this feature is not supported in the API.&quot;</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="Kegg"><a class="viewcode-back" href="../comprior.html#knowledgebases.Kegg">[docs]</a><span class="k">class</span> <span class="nc">Kegg</span><span class="p">(</span><span class="n">KnowledgeBase</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Knowledge base implementation for KEGG.</span>
<span class="sd">       Uses the KEGG web service implementation provided by bioservices.</span>
<span class="sd">       Requires an instance of :class:`KEGGPathwayParser` to be able to map retrieved pathways into the internal pathway format.</span>

<span class="sd">       :param name: name of the knowledge base</span>
<span class="sd">       :type name: str</span>
<span class="sd">       :param config: configuration parameter of the knowledge base as specified in the config file.</span>
<span class="sd">       :type config: dict</span>
<span class="sd">       :param webservice: web service querying implementation.</span>
<span class="sd">       :type webservice: :class:`opentargets.OpenTargetsClient`</span>
<span class="sd">       :param hasGeneInformation: true if the knowledge base provides gene association information, false otherwise</span>
<span class="sd">       :type hasGeneInformation: bool</span>
<span class="sd">       :param hasPathwayInformation: true if the knowledge base also provides pathway information, false otherwise</span>
<span class="sd">       :type hasPathwayInformation: bool</span>
<span class="sd">       :param pathwayparser: pathway mapping class that transforms KEGG pathways in SIF format into the internally used pathway format.</span>
<span class="sd">       :type pathwayparser: :class:`KEGGPathwayParser`</span>
<span class="sd">       &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pathwayparser</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pathwayparser</span> <span class="o">=</span> <span class="n">pathwayparser</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="s2">&quot;KEGG&quot;</span><span class="p">,</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfig</span><span class="p">(</span><span class="s2">&quot;KEGG&quot;</span><span class="p">),</span> <span class="n">KEGG</span><span class="p">(),</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>

<div class="viewcode-block" id="Kegg.getPathwayNames"><a class="viewcode-back" href="../comprior.html#knowledgebases.Kegg.getPathwayNames">[docs]</a>    <span class="k">def</span> <span class="nf">getPathwayNames</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Retrieve all pathway names related to the given labels, e.g. disease names.</span>

<span class="sd">           :param labels: list labels, e.g. disease names, for which to find pathways.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: list of pathway names.</span>
<span class="sd">           :rtype: list of str</span>
<span class="sd">           &quot;&quot;&quot;</span>

        <span class="n">pathways</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">:</span>
            <span class="n">pathwayListString</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;pathway&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\&quot;</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">label</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># pathway are returned in the form path:pathwayid\tdescription\n</span>
            <span class="k">if</span> <span class="n">pathwayListString</span> <span class="o">==</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pathwayListString</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                <span class="n">pathwayList</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="c1">#print(&quot;EMPTY PATHWAYS:&quot; + str(pathwayListString))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">pathwayList</span> <span class="o">=</span> <span class="n">pathwayListString</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">pathway</span> <span class="ow">in</span> <span class="n">pathwayList</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">count</span> <span class="o">&lt;=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;maxNumPathways&quot;</span><span class="p">]):</span>
                    <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="c1"># strip off the path prefix</span>
                    <span class="n">pathwayName</span> <span class="o">=</span> <span class="n">pathway</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">pathwayName</span> <span class="o">=</span> <span class="n">pathwayName</span><span class="p">[</span><span class="mi">5</span><span class="p">:]</span>
                    <span class="c1"># replace &quot;map&quot; by &quot;hsa&quot; (because we want the reference organisms for homo sapiens)</span>
                    <span class="n">pathwayName</span> <span class="o">=</span> <span class="n">pathwayName</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;map&quot;</span><span class="p">,</span> <span class="s2">&quot;hsa&quot;</span><span class="p">)</span>
                    <span class="n">pathways</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pathwayName</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">break</span>
        <span class="k">return</span> <span class="n">pathways</span></div>

<div class="viewcode-block" id="Kegg.getRelevantGenes"><a class="viewcode-back" href="../comprior.html#knowledgebases.Kegg.getRelevantGenes">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantGenes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get all genes that are related to a set of labels, e.g. disease names.</span>
<span class="sd">           For KEGG, this means we retrieve all genes that are contained in pathways associated to these labels.</span>

<span class="sd">           :param labels: list of identifiers, e.g. disease names, for which to find associated genes.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: list of associated genes.</span>
<span class="sd">           :rtype: list of str</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">pathways</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">getRelevantPathways</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">pathway</span> <span class="ow">in</span> <span class="n">pathways</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">nodes</span> <span class="o">=</span> <span class="n">pathway</span><span class="o">.</span><span class="n">nodes</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">nodes</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">label</span><span class="p">)</span>

        <span class="c1">#write genes to file</span>
        <span class="n">outputDir</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfigValue</span><span class="p">(</span><span class="s2">&quot;KEGG&quot;</span><span class="p">,</span> <span class="s2">&quot;outputDir&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s1">&#39;%Y-%m-</span><span class="si">%d</span><span class="s1">_%H-%M-%S&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
        <span class="n">util</span><span class="o">.</span><span class="n">createDirectory</span><span class="p">(</span><span class="n">outputDir</span><span class="p">)</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;queryResults.txt&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outputDir</span> <span class="o">+</span> <span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;attributeName,score</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;,0.0</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">genes</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">genes</span></div>

<div class="viewcode-block" id="Kegg.getGeneScores"><a class="viewcode-back" href="../comprior.html#knowledgebases.Kegg.getGeneScores">[docs]</a>    <span class="k">def</span> <span class="nf">getGeneScores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get association scores for all genes that are related to the provided labels, e.g. disease names.</span>
<span class="sd">           For KEGG, the association score for a gene is the sum of its degree percentile rank for every pathway, normalized by the overall number of pathways retrieved.</span>
<span class="sd">           This favors hub genes/genes having many interactions with other genes.</span>

<span class="sd">           :param labels: list of identifiers, e.g. disease names, for which to find associated genes.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: DataFrame of associated genes and their association scores, in descending order.</span>
<span class="sd">           :rtype: :class:`pandas.DataFrame`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">pathways</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">getRelevantPathways</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>
        <span class="n">overallCount</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">pathways</span><span class="p">)</span>
        <span class="n">occurrenceCounts</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">pathway</span> <span class="ow">in</span> <span class="n">pathways</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">nodes</span> <span class="o">=</span> <span class="n">pathway</span><span class="o">.</span><span class="n">nodes</span>
            <span class="n">interactions</span> <span class="o">=</span> <span class="n">pathway</span><span class="o">.</span><span class="n">interactions_by_nodes</span>
            <span class="n">reduced_interactions</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="c1">#update interactions to only store the interactions counts</span>
            <span class="k">for</span> <span class="n">entity</span> <span class="ow">in</span> <span class="n">interactions</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">reduced_interactions</span><span class="p">[</span><span class="n">entity</span><span class="o">.</span><span class="n">label</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">interactions</span><span class="p">[</span><span class="n">entity</span><span class="p">])</span>
            <span class="c1">#make dataframe from count</span>
            <span class="n">interactions_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">reduced_interactions</span><span class="p">,</span> <span class="n">orient</span> <span class="o">=</span> <span class="s1">&#39;index&#39;</span><span class="p">)</span>
            <span class="n">interactions_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;degree&quot;</span><span class="p">]</span>
            <span class="n">percentiles</span> <span class="o">=</span> <span class="n">interactions_df</span><span class="p">[</span><span class="s2">&quot;degree&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rank</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">pct</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">percentiles</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
                <span class="n">perc_score</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">feature</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">occurrenceCounts</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">occurrenceCounts</span><span class="p">[</span><span class="n">feature</span><span class="p">]</span> <span class="o">+=</span> <span class="n">perc_score</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">occurrenceCounts</span><span class="p">[</span><span class="n">feature</span><span class="p">]</span> <span class="o">=</span> <span class="n">perc_score</span>

        <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
        <span class="n">scores</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">geneID</span> <span class="ow">in</span> <span class="n">occurrenceCounts</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">geneID</span><span class="p">)</span>
            <span class="n">score</span> <span class="o">=</span> <span class="n">occurrenceCounts</span><span class="p">[</span><span class="n">geneID</span><span class="p">]</span> <span class="o">/</span> <span class="n">overallCount</span>
            <span class="n">scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>

        <span class="n">geneScores</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">:</span> <span class="n">genes</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">:</span> <span class="n">scores</span><span class="p">})</span>
        <span class="n">geneScores</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;score&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
        <span class="c1">#write genes to file</span>
        <span class="n">outputDir</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfigValue</span><span class="p">(</span><span class="s2">&quot;KEGG&quot;</span><span class="p">,</span> <span class="s2">&quot;outputDir&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s1">&#39;%Y-%m-</span><span class="si">%d</span><span class="s1">_%H-%M-%S&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
        <span class="n">util</span><span class="o">.</span><span class="n">createDirectory</span><span class="p">(</span><span class="n">outputDir</span><span class="p">)</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;queryResults.txt&quot;</span>
        <span class="n">geneScores</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">outputDir</span> <span class="o">+</span> <span class="n">filename</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">geneScores</span></div>

<div class="viewcode-block" id="Kegg.getRelevantPathways"><a class="viewcode-back" href="../comprior.html#knowledgebases.Kegg.getRelevantPathways">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantPathways</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get all pathways related to a set of labels, e.g. disease names.</span>
<span class="sd">           Uses the :class:`KEGGPathwayParser` to map KEGG&#39;s pathways from SIF to :class:`pypath.Network`.</span>

<span class="sd">           :param labels: list of gene names to be mapped</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: dict of pathway names and their internal representation as :class:`pypath.Network`.</span>
<span class="sd">           :rtype: dict</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">pathway_graphs</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="n">pathways</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">pathway_graphs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">:</span>
            <span class="n">pathwayListString</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;pathway&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\&quot;</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">label</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># pathway are returned in the form path:pathwayid\tdescription\n</span>
            <span class="k">if</span> <span class="p">((</span><span class="n">pathwayListString</span> <span class="o">==</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">pathwayListString</span><span class="p">,</span> <span class="nb">int</span><span class="p">))</span> <span class="ow">or</span> <span class="p">(</span><span class="n">pathwayListString</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)):</span>
                <span class="n">pathwayList</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="c1">#print(&quot;EMPTY PATHWAYS:&quot; + str(pathwayListString))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">pathwayList</span> <span class="o">=</span> <span class="n">pathwayListString</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">pathway</span> <span class="ow">in</span> <span class="n">pathwayList</span><span class="p">:</span>
                <span class="c1"># strip off the path prefix</span>
                <span class="n">pathwayName</span> <span class="o">=</span> <span class="n">pathway</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">pathwayName</span> <span class="o">=</span> <span class="n">pathwayName</span><span class="p">[</span><span class="mi">5</span><span class="p">:]</span>
                <span class="c1"># replace &quot;map&quot; by &quot;hsa&quot; (because we want the reference organisms for homo sapiens)</span>
                <span class="n">pathwayName</span> <span class="o">=</span> <span class="n">pathwayName</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;map&quot;</span><span class="p">,</span> <span class="s2">&quot;hsa&quot;</span><span class="p">)</span>

                <span class="n">pathway_resource</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">pathwayName</span><span class="p">,</span> <span class="s2">&quot;kgml&quot;</span><span class="p">)</span>
                <span class="c1">#only read the pathway if it contains meaningful information</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">pathway_resource</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">count</span> <span class="o">&lt;=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;maxNumPathways&quot;</span><span class="p">]):</span>
                        <span class="n">xmlPathway</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">parse_kgml_pathway</span><span class="p">(</span><span class="n">pathwayName</span><span class="p">,</span> <span class="n">pathway_resource</span><span class="p">)</span>
                        <span class="n">parsed_pathway</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pathwayparser</span><span class="o">.</span><span class="n">parsePathway</span><span class="p">(</span><span class="n">xmlPathway</span><span class="p">,</span> <span class="n">pathwayName</span><span class="p">)</span>
                        <span class="c1">#only add if the pathway is not empty</span>
                        <span class="k">if</span> <span class="n">parsed_pathway</span><span class="o">.</span><span class="n">vcount</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                            <span class="n">pathway_graphs</span><span class="p">[</span><span class="n">pathwayName</span><span class="p">]</span> <span class="o">=</span> <span class="n">parsed_pathway</span>
                            <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">break</span>

        <span class="k">return</span> <span class="n">pathway_graphs</span></div></div>


<div class="viewcode-block" id="Disgenet"><a class="viewcode-back" href="../comprior.html#knowledgebases.Disgenet">[docs]</a><span class="k">class</span> <span class="nc">Disgenet</span><span class="p">(</span><span class="n">KnowledgeBase</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Knowledge base implementation for DisGeNET.</span>

<span class="sd">       :param name: name of the knowledge base</span>
<span class="sd">       :type name: str</span>
<span class="sd">       :param config: configuration parameter of the knowledge base as specified in the config file.</span>
<span class="sd">       :type config: dict</span>
<span class="sd">       :param webservice: web service querying implementation.</span>
<span class="sd">       :type webservice: :class:`DISGENET`</span>
<span class="sd">       :param hasGeneInformation: true if the knowledge base provides gene association information, false otherwise</span>
<span class="sd">       :type hasGeneInformation: bool</span>
<span class="sd">       :param hasPathwayInformation: true if the knowledge base also provides pathway information, false otherwise</span>
<span class="sd">       :type hasPathwayInformation: bool</span>
<span class="sd">       &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="s2">&quot;DisGeNET&quot;</span><span class="p">,</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfig</span><span class="p">(</span><span class="s2">&quot;DisGeNET&quot;</span><span class="p">),</span> <span class="n">DISGENET</span><span class="p">(),</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>

<div class="viewcode-block" id="Disgenet.getRelevantGenes"><a class="viewcode-back" href="../comprior.html#knowledgebases.Disgenet.getRelevantGenes">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantGenes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get all genes that are related to a set of labels, e.g. disease names.</span>

<span class="sd">           :param labels: list of identifiers, e.g. disease names, for which to find associated genes.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: list of associated genes.</span>
<span class="sd">           :rtype: list of str</span>
<span class="sd">           &quot;&quot;&quot;</span>

        <span class="n">assocs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>

        <span class="c1">#check if error code was returned</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">assocs</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ERROR &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">assocs</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; RETURNED FOR LABELS: &quot;</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">labels</span><span class="p">))</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">])</span>

        <span class="c1"># extract only the genes from the results</span>
        <span class="c1"># for now, just merge all gene sets from the query into one (no matter how many genes were associated to a particular disease</span>
        <span class="c1"># (=no interleaving)</span>
        <span class="n">relevantGenes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">assocs</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">])</span>

        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">relevantGenes</span><span class="p">)</span></div>

<div class="viewcode-block" id="Disgenet.getGeneScores"><a class="viewcode-back" href="../comprior.html#knowledgebases.Disgenet.getGeneScores">[docs]</a>    <span class="k">def</span> <span class="nf">getGeneScores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get association scores for all genes that are related to the provided labels, e.g. disease names.</span>
<span class="sd">           DisGeNET provides a couple of association scores to its genes (https://www.disgenet.org/dbinfo).</span>
<span class="sd">           Which score to use can be defined by the user in the config file.</span>

<span class="sd">           :param labels: list of identifiers, e.g. disease names, for which to find associated genes.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: DataFrame of associated genes and their association scores, in descending order.</span>
<span class="sd">           :rtype: :class:`pandas.DataFrame`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">assocs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>
        <span class="c1"># check if an error code was returned</span>
        <span class="c1"># check if error code was returned</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">assocs</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ERROR &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">assocs</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; RETURNED FOR LABELS: &quot;</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">labels</span><span class="p">))</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">assocs</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">])</span>

        <span class="n">assocs</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">]</span>
        <span class="c1">#sort by assoc score</span>
        <span class="n">assocs</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;score&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
        <span class="c1"># for duplicate gene entries, select entries with highest association score</span>
        <span class="n">assocs</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">assocs</span></div>

<div class="viewcode-block" id="Disgenet.getRelevantPathways"><a class="viewcode-back" href="../comprior.html#knowledgebases.Disgenet.getRelevantPathways">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantPathways</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;As DisGeNET currently does not provide pathway information, this feature is not implemented.</span>

<span class="sd">           :param labels: list of labels for which to find related pathways.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: :class:`NotImplementedError` as this knowledge base is not intended to be used for such analyses.</span>
<span class="sd">           :rtype: :class:`NotImplementedError`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;DisGeNET is not a pathway database. Querying for pathways is not possible.&quot;</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="Pathwaycommons"><a class="viewcode-back" href="../comprior.html#knowledgebases.Pathwaycommons">[docs]</a><span class="k">class</span> <span class="nc">Pathwaycommons</span><span class="p">(</span><span class="n">KnowledgeBase</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Knowledge base implementation for PathwayCommons.</span>

<span class="sd">       :param name: name of the knowledge base</span>
<span class="sd">       :type name: str</span>
<span class="sd">       :param config: configuration parameter of the knowledge base as specified in the config file.</span>
<span class="sd">       :type config: dict</span>
<span class="sd">       :param webservice: web service querying implementation.</span>
<span class="sd">       :type webservice: :class:`opentargets.OpenTargetsClient`</span>
<span class="sd">       :param hasGeneInformation: true if the knowledge base provides gene association information, false otherwise</span>
<span class="sd">       :type hasGeneInformation: bool</span>
<span class="sd">       :param hasPathwayInformation: true if the knowledge base also provides pathway information, false otherwise</span>
<span class="sd">       :type hasPathwayInformation: bool</span>
<span class="sd">       &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="s2">&quot;PathwayCommons&quot;</span><span class="p">,</span> <span class="n">util</span><span class="o">.</span><span class="n">getConfig</span><span class="p">(</span><span class="s2">&quot;PathwayCommons&quot;</span><span class="p">),</span> <span class="n">PATHWAYCOMMONSWS</span><span class="p">(),</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>

<div class="viewcode-block" id="Pathwaycommons.getGeneScores"><a class="viewcode-back" href="../comprior.html#knowledgebases.Pathwaycommons.getGeneScores">[docs]</a>    <span class="k">def</span> <span class="nf">getGeneScores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get association scores for all genes that are related to the provided labels, e.g. disease names.</span>
<span class="sd">           For PathwayCommons, the association score for a gene is the sum of its degree percentile rank for every pathway, normalized by the overall number of pathways retrieved.</span>
<span class="sd">           This favors hub genes/genes having many interactions with other genes.</span>

<span class="sd">           :param labels: list of identifiers, e.g. disease names, for which to find associated genes.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: DataFrame of associated genes and their association scores, in descending order.</span>
<span class="sd">           :rtype: :class:`pandas.DataFrame`</span>
<span class="sd">           &quot;&quot;&quot;</span>

        <span class="n">pathways</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">getRelevantPathways</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>
        <span class="n">overallCount</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">pathways</span><span class="p">)</span>
        <span class="n">occurrenceCounts</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">pathway</span> <span class="ow">in</span> <span class="n">pathways</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">nodes</span> <span class="o">=</span> <span class="n">pathway</span><span class="o">.</span><span class="n">nodes</span>
            <span class="n">interactions</span> <span class="o">=</span> <span class="n">pathway</span><span class="o">.</span><span class="n">interactions_by_nodes</span>
            <span class="n">reduced_interactions</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="c1"># update interactions to only store the interactions counts</span>
            <span class="k">for</span> <span class="n">entity</span> <span class="ow">in</span> <span class="n">interactions</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">reduced_interactions</span><span class="p">[</span><span class="n">entity</span><span class="o">.</span><span class="n">label</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">interactions</span><span class="p">[</span><span class="n">entity</span><span class="p">])</span>
            <span class="c1"># make dataframe from count</span>
            <span class="n">interactions_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">reduced_interactions</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">)</span>
            <span class="n">interactions_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;degree&quot;</span><span class="p">]</span>
            <span class="n">percentiles</span> <span class="o">=</span> <span class="n">interactions_df</span><span class="p">[</span><span class="s2">&quot;degree&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rank</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">pct</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">percentiles</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
                <span class="n">perc_score</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">feature</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">occurrenceCounts</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">occurrenceCounts</span><span class="p">[</span><span class="n">feature</span><span class="p">]</span> <span class="o">+=</span> <span class="n">perc_score</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">occurrenceCounts</span><span class="p">[</span><span class="n">feature</span><span class="p">]</span> <span class="o">=</span> <span class="n">perc_score</span>

        <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
        <span class="n">scores</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">geneID</span> <span class="ow">in</span> <span class="n">occurrenceCounts</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">geneID</span><span class="p">)</span>
            <span class="n">score</span> <span class="o">=</span> <span class="n">occurrenceCounts</span><span class="p">[</span><span class="n">geneID</span><span class="p">]</span> <span class="o">/</span> <span class="n">overallCount</span>
            <span class="n">scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>

        <span class="n">geneScores</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">:</span> <span class="n">genes</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">:</span> <span class="n">scores</span><span class="p">})</span>
        <span class="n">geneScores</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;score&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">geneScores</span></div>

<div class="viewcode-block" id="Pathwaycommons.getRelevantGenes"><a class="viewcode-back" href="../comprior.html#knowledgebases.Pathwaycommons.getRelevantGenes">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantGenes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get all genes that are related to a set of labels, e.g. disease names.</span>
<span class="sd">           For PathwayCommons, this means we retrieve all genes that are contained in pathways associated to these labels.</span>

<span class="sd">           :param labels: list of identifiers, e.g. disease names, for which to find associated genes.</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: list of associated genes.</span>
<span class="sd">           :rtype: list of str</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">pathways</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">getRelevantPathways</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">pathway</span> <span class="ow">in</span> <span class="n">pathways</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">nodes</span> <span class="o">=</span> <span class="n">pathway</span><span class="o">.</span><span class="n">nodes</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">nodes</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">label</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">genes</span></div>

<div class="viewcode-block" id="Pathwaycommons.readPathway"><a class="viewcode-back" href="../comprior.html#knowledgebases.Pathwaycommons.readPathway">[docs]</a>    <span class="k">def</span> <span class="nf">readPathway</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pathway</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Reads a pathway to create :class:`pypath.Network`.</span>

<span class="sd">           :param pathway: pathway string to parse</span>
<span class="sd">           :type pathway: str</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="n">interactions</span> <span class="o">=</span> <span class="n">pathway</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">interaction</span> <span class="ow">in</span> <span class="n">interactions</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">()</span></div>

<div class="viewcode-block" id="Pathwaycommons.getRelevantPathways"><a class="viewcode-back" href="../comprior.html#knowledgebases.Pathwaycommons.getRelevantPathways">[docs]</a>    <span class="k">def</span> <span class="nf">getRelevantPathways</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get all pathways related to a set of labels, e.g. disease names as :class:`pypath.Network`.</span>

<span class="sd">           :param labels: list of gene names to be mapped</span>
<span class="sd">           :type labels: list of str</span>
<span class="sd">           :return: dict of pathway names and their internal representation as :class:`pypath.Network`.</span>
<span class="sd">           :rtype: dict</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="c1">#collect pathway IDs first</span>
        <span class="n">overall</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">pathways</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">p_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">:</span>
            <span class="n">pathwayIDs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">term</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;homo sapiens&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;pathway&quot;</span><span class="p">)</span>
            <span class="c1"># if no error code was returned</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">pathwayIDs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pathwayIDs</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                <span class="k">continue</span>

            <span class="n">numHits</span> <span class="o">=</span> <span class="n">pathwayIDs</span><span class="p">[</span><span class="s2">&quot;numHits&quot;</span><span class="p">]</span>
            <span class="n">maxHits</span> <span class="o">=</span> <span class="n">pathwayIDs</span><span class="p">[</span><span class="s2">&quot;maxHitsPerPage&quot;</span><span class="p">]</span>
            <span class="n">pages</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">numHits</span> <span class="o">/</span> <span class="n">maxHits</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">pages</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">pages</span><span class="p">):</span>
                    <span class="n">pathwayIDs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">term</span><span class="p">,</span> <span class="n">page</span><span class="o">=</span><span class="n">i</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;homo sapiens&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;pathway&quot;</span><span class="p">)</span>

                    <span class="c1">#if no error code was returned</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">pathwayIDs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pathwayIDs</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                        <span class="k">continue</span>

                    <span class="n">items</span> <span class="o">=</span> <span class="n">pathwayIDs</span><span class="p">[</span><span class="s2">&quot;searchHit&quot;</span><span class="p">]</span>
                    <span class="n">overall</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">items</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">items</span><span class="p">:</span>
                        <span class="n">p_ids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;uri&quot;</span><span class="p">])</span>



        <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="nb">id</span> <span class="ow">in</span> <span class="n">p_ids</span><span class="p">:</span>
            <span class="n">pathway_sif</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">webservice</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="nb">id</span><span class="p">,</span> <span class="n">frmt</span> <span class="o">=</span> <span class="s2">&quot;SIF&quot;</span><span class="p">)</span>
            <span class="c1">#if server does not return an error code and the pathway has a sif version</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">pathway_sif</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">):</span>
                <span class="k">continue</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pathway_sif</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span><span class="n">pathway_sif</span><span class="o">.</span><span class="n">text</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
                <span class="c1">#only load the pathway if the maxNumPathways count is not reached yet</span>
                <span class="k">if</span> <span class="n">count</span> <span class="o">&lt;=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;maxNumPathways&quot;</span><span class="p">]):</span>
                    <span class="n">params</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;pathway&quot;</span><span class="p">:</span> <span class="n">pathway_sif</span><span class="p">}</span>

                    <span class="c1">#create pypath pathway from SIF</span>
                    <span class="nb">input</span> <span class="o">=</span> <span class="n">input_formats</span><span class="o">.</span><span class="n">NetworkInput</span><span class="p">(</span>
                        <span class="n">name</span><span class="o">=</span><span class="nb">id</span><span class="p">,</span>
                        <span class="nb">input</span><span class="o">=</span> <span class="n">Pathwaycommons</span><span class="o">.</span><span class="n">readPathway</span><span class="p">,</span>
                        <span class="n">input_args</span><span class="o">=</span><span class="n">params</span><span class="p">,</span>
                        <span class="n">separator</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span>
                        <span class="n">id_col_a</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                        <span class="n">id_col_b</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                        <span class="n">id_type_a</span><span class="o">=</span><span class="s1">&#39;genesymbol&#39;</span><span class="p">,</span>
                        <span class="n">id_type_b</span><span class="o">=</span><span class="s1">&#39;genesymbol&#39;</span><span class="p">,</span>
                        <span class="n">sign</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>
                    <span class="p">)</span>

                    <span class="n">pathway</span> <span class="o">=</span> <span class="n">network</span><span class="o">.</span><span class="n">Network</span><span class="p">()</span>
                    <span class="n">pathway</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">input</span><span class="p">)</span>
                    <span class="c1">#only add pathway if it has nodes</span>
                    <span class="k">if</span> <span class="n">pathway</span><span class="o">.</span><span class="n">vcount</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">pathways</span><span class="p">[</span><span class="nb">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">pathway</span>
                        <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="k">else</span><span class="p">:</span>
                    <span class="k">break</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">continue</span>

        <span class="k">return</span> <span class="n">pathways</span></div></div>

<span class="c1">############################### PATHWAY MAPPERS ###############################</span>

<div class="viewcode-block" id="PathwayParser"><a class="viewcode-back" href="../comprior.html#knowledgebases.PathwayParser">[docs]</a><span class="k">class</span> <span class="nc">PathwayParser</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Super class that maps a pathway from its original format (provided by a knowledge base) to the internally used :class:`pypath.Network`.</span>
<span class="sd">       When having to map pathways from a knowledge base, implement a new class that inherits from this one and implements :meth:`PathwayParser.parsePathway()`.</span>
<span class="sd">       &quot;&quot;&quot;</span>

    <span class="c1">#abstract method for pathway parsing. returns an igraph object for the input pathway</span>
<div class="viewcode-block" id="PathwayParser.parsePathway"><a class="viewcode-back" href="../comprior.html#knowledgebases.PathwayParser.parsePathway">[docs]</a>    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">parsePathway</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pathway</span><span class="p">,</span> <span class="n">pathwayID</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Abstract method.</span>
<span class="sd">           Parse a pathway to the internally used format of :class:`pypath.Network`.</span>

<span class="sd">           :param pathway: pathway string to parse</span>
<span class="sd">           :type pathway: str</span>
<span class="sd">           :param pathwayID: name of the pathway</span>
<span class="sd">           :type pathwayID: str</span>
<span class="sd">           :return: pathway in the internally used format..</span>
<span class="sd">           :rtype: :class:`pypath.Network`</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">pass</span></div></div>

<div class="viewcode-block" id="KEGGPathwayParser"><a class="viewcode-back" href="../comprior.html#knowledgebases.KEGGPathwayParser">[docs]</a><span class="k">class</span> <span class="nc">KEGGPathwayParser</span><span class="p">(</span><span class="n">PathwayParser</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Parse KEGG pathways, which are returned in KGML format.</span>
<span class="sd">       &quot;&quot;&quot;</span>

<div class="viewcode-block" id="KEGGPathwayParser.readInteractions"><a class="viewcode-back" href="../comprior.html#knowledgebases.KEGGPathwayParser.readInteractions">[docs]</a>    <span class="k">def</span> <span class="nf">readInteractions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">interactions</span><span class="p">,</span> <span class="n">geneIds</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Parses interactions for a set of genes.</span>

<span class="sd">           :param interactions: interactions to parse</span>
<span class="sd">           :type interactions: list</span>
<span class="sd">           :param geneIds: gene ids whose interactions to add</span>
<span class="sd">           :type geneIds: list of str</span>
<span class="sd">           &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">interaction</span> <span class="ow">in</span> <span class="n">interactions</span><span class="p">:</span>
            <span class="c1"># filter pathway relations for protein-protein (PPrel) and gene expression (GErel) interactions</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">interaction</span><span class="p">[</span><span class="s2">&quot;link&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;PPrel&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">interaction</span><span class="p">[</span><span class="s2">&quot;link&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;GErel&quot;</span><span class="p">):</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">source</span> <span class="o">=</span> <span class="n">geneIds</span><span class="p">[</span><span class="n">interaction</span><span class="p">[</span><span class="s2">&quot;entry1&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">label</span>
                    <span class="n">target</span> <span class="o">=</span> <span class="n">geneIds</span><span class="p">[</span><span class="n">interaction</span><span class="p">[</span><span class="s2">&quot;entry2&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">label</span>
                    <span class="n">relation_name</span> <span class="o">=</span> <span class="n">interaction</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
                    <span class="n">interactionString</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">source</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">relation_name</span><span class="p">])</span>
                    <span class="k">yield</span> <span class="n">interactionString</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                <span class="k">except</span><span class="p">:</span>
                    <span class="k">continue</span></div>

<div class="viewcode-block" id="KEGGPathwayParser.parsePathway"><a class="viewcode-back" href="../comprior.html#knowledgebases.KEGGPathwayParser.parsePathway">[docs]</a>    <span class="k">def</span> <span class="nf">parsePathway</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kgml_pathway</span><span class="p">,</span> <span class="n">pathwayID</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Parse KEGG pathway to the internally used format of :class:`pypath.Network`.</span>

<span class="sd">           :param pathway: pathway string to parse</span>
<span class="sd">           :type pathway: str</span>
<span class="sd">           :param pathwayID: name of the pathway</span>
<span class="sd">           :type pathwayID: str</span>
<span class="sd">           :return: pathway in the internally used format..</span>
<span class="sd">           :rtype: :class:`pypath.Network`</span>
<span class="sd">           &quot;&quot;&quot;</span>

        <span class="c1">#create gene IDs first</span>
        <span class="n">geneIds</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">kgml_pathway</span><span class="p">[</span><span class="s2">&quot;entries&quot;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">entry</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span><span class="p">:</span>
                <span class="c1"># select first of the gene names in the list to be the alias</span>
                <span class="n">node</span> <span class="o">=</span> <span class="n">entity</span><span class="o">.</span><span class="n">Entity</span><span class="p">(</span><span class="n">entry</span><span class="p">[</span><span class="s2">&quot;gene_names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s2">&quot;...&quot;</span><span class="p">))</span>
                <span class="n">geneIds</span><span class="p">[</span><span class="n">entry</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span><span class="o">=</span> <span class="n">node</span>

        <span class="n">params</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;interactions&quot;</span><span class="p">:</span> <span class="n">kgml_pathway</span><span class="p">[</span><span class="s2">&quot;relations&quot;</span><span class="p">],</span> <span class="s2">&quot;geneIds&quot;</span><span class="p">:</span> <span class="n">geneIds</span><span class="p">}</span>

        <span class="c1"># create pypath pathway from SIF</span>
        <span class="nb">input</span> <span class="o">=</span> <span class="n">input_formats</span><span class="o">.</span><span class="n">NetworkInput</span><span class="p">(</span>
            <span class="n">name</span><span class="o">=</span><span class="n">pathwayID</span><span class="p">,</span>
            <span class="nb">input</span><span class="o">=</span><span class="n">KEGGPathwayParser</span><span class="o">.</span><span class="n">readInteractions</span><span class="p">,</span>
            <span class="n">input_args</span><span class="o">=</span><span class="n">params</span><span class="p">,</span>
            <span class="n">separator</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span>
            <span class="n">id_col_a</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">id_col_b</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">id_type_a</span><span class="o">=</span><span class="s1">&#39;genesymbol&#39;</span><span class="p">,</span>
            <span class="n">id_type_b</span><span class="o">=</span><span class="s1">&#39;genesymbol&#39;</span><span class="p">,</span>
            <span class="n">sign</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="n">pathway</span> <span class="o">=</span> <span class="n">network</span><span class="o">.</span><span class="n">Network</span><span class="p">()</span>
        <span class="n">pathway</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">input</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">pathway</span></div></div>

</pre></div>

           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Cindy Perscheid.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>