

<!DOCTYPE html>
<html class="writer-html4" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Python Code Documentation &mdash; Comprior  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript">
          var DOCUMENTATION_OPTIONS = {
              URL_ROOT:'./',
              VERSION:'',
              LANGUAGE:'None',
              COLLAPSE_INDEX:false,
              FILE_SUFFIX:'.html',
              HAS_SOURCE:  true,
              SOURCELINK_SUFFIX: '.txt'
          };
      </script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="R Code Documentation" href="r-doc.html" />
    <link rel="prev" title="Extending Comprior - How-Tos" href="howtos.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Comprior
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation and Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases.html">Example Use Cases</a></li>
<li class="toctree-l1"><a class="reference internal" href="inputdata.html">Input Data Sets</a></li>
<li class="toctree-l1"><a class="reference internal" href="configparams.html">Configuration Parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="outputstructure.html">Folder Structure  - Where to find what Files (In- and Output)</a></li>
<li class="toctree-l1"><a class="reference internal" href="knowledgebases.html">Knowledge Bases</a></li>
<li class="toctree-l1"><a class="reference internal" href="priorknowledge.html">Prior Knowledge Approaches</a></li>
<li class="toctree-l1"><a class="reference internal" href="howtos.html">Extending Comprior - How-Tos</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Python Code Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#pipeline-module">pipeline module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#benchutils-module">benchutils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#preprocessing-module">preprocessing module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#featureselection-module">featureselection module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#knowledgebases-module">knowledgebases module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#evaluation-module">evaluation module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="r-doc.html">R Code Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="java-doc.html">Java Code Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="architecture.html">System Architecture</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Comprior</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Python Code Documentation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/comprior.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="python-code-documentation">
<h1>Python Code Documentation<a class="headerlink" href="#python-code-documentation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="pipeline-module">
<h2>pipeline module<a class="headerlink" href="#pipeline-module" title="Permalink to this headline">¶</a></h2>
<p>The framework module is responsible for orchestrating the complete benchmarking process.
It is the starting point that is invoked when running Comprior.
It tidies up and prepares working directories, creates and coordinates the execution order of preprocessing modules, feature selectors, and evaluation procedures.</p>
<span class="target" id="module-pipeline"></span><dl class="class">
<dt id="pipeline.Pipeline">
<em class="property">class </em><code class="descclassname">pipeline.</code><code class="descname">Pipeline</code><span class="sig-paren">(</span><em>userConfig</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class that executes the complete benchmarking pipeline.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>outputRootPath</strong> (<em>str</em>) – absolute path to the overall output directory (will be extended by own folders by every <a class="reference internal" href="#evaluation.Evaluator" title="evaluation.Evaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.Evaluator</span></code></a>).</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="pipeline.Pipeline.prepareExecution">
<code class="descname">prepareExecution</code><span class="sig-paren">(</span><em>userConfig</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline.prepareExecution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline.prepareExecution" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepares the pipeline execution by loading the configuration file, clearing intermediate directories, and creating output directories.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>userConfig</strong> (<em>str</em>) – absolute path to an additional user configuration file (config.ini will always be used by default) to overwrite default configuration.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pipeline.Pipeline.evaluateInputData">
<code class="descname">evaluateInputData</code><span class="sig-paren">(</span><em>inputfile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline.evaluateInputData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline.evaluateInputData" title="Permalink to this definition">¶</a></dt>
<dd><p>Run <a class="reference internal" href="#evaluation.DatasetEvaluator" title="evaluation.DatasetEvaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.DatasetEvaluator</span></code></a> to create plots as specified by the config’s Evaluation-preanalysis parameter.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>inputfile</strong> (<em>str</em>) – absolute path to the input data set to be analyzed.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pipeline.Pipeline.evaluateKnowledgeBases">
<code class="descname">evaluateKnowledgeBases</code><span class="sig-paren">(</span><em>labeledInputDataPath</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline.evaluateKnowledgeBases"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline.evaluateKnowledgeBases" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluates knowledge base coverage for all knowledge bases that are used in the specified feature selection methods.
Uses the class labels and alternativeSearchTerms from the config, queries the knowledge bases and creates corresponding plots regarding coverage of theses search terms.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labeledInputDataPath</strong> (<em>str</em>) – absolute path to the labeled input data set.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pipeline.Pipeline.runFeatureSelector">
<code class="descname">runFeatureSelector</code><span class="sig-paren">(</span><em>selector</em>, <em>datasetLocation</em>, <em>outputDir</em>, <em>loggingDir</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline.runFeatureSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline.runFeatureSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs a given feature selector.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>selector</strong> (<a class="reference internal" href="#featureselection.FeatureSelector" title="featureselection.FeatureSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.FeatureSelector</span></code></a>) – Any feature selector that inherits from <a class="reference internal" href="#featureselection.FeatureSelector" title="featureselection.FeatureSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.FeatureSelector</span></code></a>.</li>
<li><strong>datasetLocation</strong> (<em>str</em>) – absolute path to the input data set (from which features should be selected).</li>
<li><strong>outputDir</strong> (<em>str</em>) – absolute path to the selector’s output directory (where ranking will be written to).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pipeline.Pipeline.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><em>datasetLocation</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates and runs all feature selectors that are listed in the config file.
Applies parallelization by running as much feature selectors in parallel as stated in the config’s General–&gt;numCores attribute.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>datasetLocation</strong> (<em>str</em>) – absolute path to the input data set (from which features should be selected).</li>
<li><strong>outputRootPath</strong> (<em>str</em>) – absolute path to the selector’s output directory (where ranking will be written to).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">absolute path to directory that contains generated feature rankings.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">str</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pipeline.Pipeline.assignColors">
<code class="descname">assignColors</code><span class="sig-paren">(</span><em>methods</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline.assignColors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline.assignColors" title="Permalink to this definition">¶</a></dt>
<dd><p>Assigns each (feature selection) method a unique color.
Will be delivered later on to every <a class="reference internal" href="#evaluation.Evaluator" title="evaluation.Evaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.Evaluator</span></code></a> instance to create visualizations with consistent coloring for evaluated approaches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>methods</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of str) – List of method names.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Dictionary containing hex color codes for every method</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dict</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pipeline.Pipeline.assignMarkers">
<code class="descname">assignMarkers</code><span class="sig-paren">(</span><em>approaches</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline.assignMarkers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline.assignMarkers" title="Permalink to this definition">¶</a></dt>
<dd><p>Assigns each (feature selection) method a unique color.
Will be delivered later on to every <a class="reference internal" href="#evaluation.Evaluator" title="evaluation.Evaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.Evaluator</span></code></a> instance to create visualizations with consistent coloring for evaluated approaches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>methods</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of str) – List of method names.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Dictionary containing hex color codes for every method</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dict</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pipeline.Pipeline.evaluateBiomarkers">
<code class="descname">evaluateBiomarkers</code><span class="sig-paren">(</span><em>inputDir</em>, <em>dataset</em>, <em>rankingsDir</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline.evaluateBiomarkers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline.evaluateBiomarkers" title="Permalink to this definition">¶</a></dt>
<dd><p>Covers the evaluation phase.
Processes input data to only contain the top k selected features per feature selection approach via the <a class="reference internal" href="#evaluation.AttributeRemover" title="evaluation.AttributeRemover"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.AttributeRemover</span></code></a>.
Runs all selected evaluation strategies that cover assessment of rankings (<a class="reference internal" href="#evaluation.RankingsEvaluator" title="evaluation.RankingsEvaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.RankingsEvaluator</span></code></a>), annotations(<a class="reference internal" href="#evaluation.AnnotationEvaluator" title="evaluation.AnnotationEvaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.AnnotationEvaluator</span></code></a>), and classification performance (<a class="reference internal" href="#evaluation.ClassificationEvaluator" title="evaluation.ClassificationEvaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.ClassificationEvaluator</span></code></a>).
If selected, also conducts cross-validation across data sets with <a class="reference internal" href="#evaluation.CrossEvaluator" title="evaluation.CrossEvaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.CrossEvaluator</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>inputDir</strong> (<em>str</em>) – absolute path to the directory where input data sets are located (for <a class="reference internal" href="#evaluation.AttributeRemover" title="evaluation.AttributeRemover"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.AttributeRemover</span></code></a>).</li>
<li><strong>dataset</strong> (<em>str</em>) – absolute file path to the input data set (from which features should be selected).</li>
<li><strong>rankingsDir</strong> (<em>str</em>) – absolute path to the directory that contains all rankings.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pipeline.Pipeline.preprocessData">
<code class="descname">preprocessData</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline.preprocessData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline.preprocessData" title="Permalink to this definition">¶</a></dt>
<dd><p>Preprocesses the input data set specified in the config file.
Preprocessing consists of a) transposing the data so that features are in the columns (if necessary), b) mapping the features to the right format (if necessary), c) labeling the data with the user-specified metadata attribute, d) filtering features or samples that have too few information (optional, specified via config), and finally e) putting the analysis-ready data set to the right location for further processing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">A tuple consisting of the absolute path to the analysis-ready data set and the absolute path to the mapped input final_filename and mapped_input</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">tuple(str,str)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pipeline.Pipeline.loadConfig">
<code class="descname">loadConfig</code><span class="sig-paren">(</span><em>userConfig</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline.loadConfig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline.loadConfig" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads the config files.
config.ini will always be loaded as default config file, all other config files provided by userConfig overwrite corresponding values.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>userConfig</strong> (str or <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of str, optional) – absolute path(s) to user-defined config files that should be used. If config files specify the same parameter, the value specified by the last config file in the list will be used.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pipeline.Pipeline.prepareDirectories">
<code class="descname">prepareDirectories</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline.prepareDirectories"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline.prepareDirectories" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepares directory structure for benchmarking run.
Creates all necessary directories in the output folder.
Also cleans up intermediate directory so that no old data is accidentially used.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the directory where all results from this run will be stored.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pipeline.Pipeline.executePipeline">
<code class="descname">executePipeline</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pipeline.html#Pipeline.executePipeline"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pipeline.Pipeline.executePipeline" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>The entry point for the overall benchmarking process.</dt>
<dd>This method is invoked when running the framework, and from here all other steps of the benchmarking process are encapsulated in own methods.</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>userConfig</strong> (<em>str</em>) – absolute path to an additional user configuration file (config.ini will always be used by default) to overwrite default configuration.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="benchutils-module">
<h2>benchutils module<a class="headerlink" href="#benchutils-module" title="Permalink to this headline">¶</a></h2>
<p>Utility module that provides functionality that is repeatedly used across the system, e.g. directory handling and file loading, identifier mapping, logging, and running external code from R or Java.
It also loads and stores the configuration parameters.</p>
<span class="target" id="module-benchutils"></span><dl class="function">
<dt id="benchutils.loadConfig">
<code class="descclassname">benchutils.</code><code class="descname">loadConfig</code><span class="sig-paren">(</span><em>path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#loadConfig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.loadConfig" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads the config files.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path</strong> (<em>str</em><em> or </em><em>list of str</em>) – absolute path or list of absolute paths to the config files. For multiple config files specifying the same parameters, the ones from the last config file in the list will be used.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.getConfig">
<code class="descclassname">benchutils.</code><code class="descname">getConfig</code><span class="sig-paren">(</span><em>category</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#getConfig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.getConfig" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the config entries for a particular category.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>category</strong> (<em>str</em>) – category name.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">all parameters for that config category</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dict</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.getConfigValue">
<code class="descclassname">benchutils.</code><code class="descname">getConfigValue</code><span class="sig-paren">(</span><em>category</em>, <em>identifier</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#getConfigValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.getConfigValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the value for a given config parameter.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>category</strong> (<em>str</em>) – the parameter’s category name.</li>
<li><strong>identifier</strong> (<em>str</em>) – the parameter name.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the parameter value.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">str</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.getConfigBoolean">
<code class="descclassname">benchutils.</code><code class="descname">getConfigBoolean</code><span class="sig-paren">(</span><em>category</em>, <em>identifier</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#getConfigBoolean"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.getConfigBoolean" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the boolean value for a given config parameter.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>category</strong> (<em>str</em>) – the parameter’s category name.</li>
<li><strong>identifier</strong> (<em>str</em>) – the parameter name.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the parameter boolean value.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">bool</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.loadRanking">
<code class="descclassname">benchutils.</code><code class="descname">loadRanking</code><span class="sig-paren">(</span><em>rankingFile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#loadRanking"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.loadRanking" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a feature ranking from a file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>rankingFile</strong> (<em>str</em>) – absolute path to the file containing a feature ranking.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the feature ranking as a DataFrame.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.createOrClearDirectory">
<code class="descclassname">benchutils.</code><code class="descname">createOrClearDirectory</code><span class="sig-paren">(</span><em>directoryLocation</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#createOrClearDirectory"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.createOrClearDirectory" title="Permalink to this definition">¶</a></dt>
<dd><p>If the provided directory location is already existing, remove all files in that directory.
Create a new directory otherwise.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>directoryLocation</strong> (<em>str</em>) – absolute path to the directory that must be cleared or created.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.createDirectory">
<code class="descclassname">benchutils.</code><code class="descname">createDirectory</code><span class="sig-paren">(</span><em>directoryLocation</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#createDirectory"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.createDirectory" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a directory.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>directoryLocation</strong> (<em>str</em>) – absolute path to the directory to be created.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.removeDirectoryContent">
<code class="descclassname">benchutils.</code><code class="descname">removeDirectoryContent</code><span class="sig-paren">(</span><em>directoryLocation</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#removeDirectoryContent"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.removeDirectoryContent" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove the files inside a directory.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>directoryLocation</strong> (<em>str</em>) – absolute path to the directory that must be cleared.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.removeFile">
<code class="descclassname">benchutils.</code><code class="descname">removeFile</code><span class="sig-paren">(</span><em>file</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#removeFile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.removeFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete a file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>file</strong> (<em>str</em>) – absolute path to the file that must be deleted.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.cleanupResults">
<code class="descclassname">benchutils.</code><code class="descname">cleanupResults</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#cleanupResults"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.cleanupResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove all intermediate files from former runs, e.g. generated during preprocessing or mapping.</p>
</dd></dl>

<dl class="function">
<dt id="benchutils.createLogger">
<code class="descclassname">benchutils.</code><code class="descname">createLogger</code><span class="sig-paren">(</span><em>outputPath</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#createLogger"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.createLogger" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a logger for Comprior. Creates two handlers for this logger: one for console output that only contains high-level status update logs and error messages.
Warnings and other tracing information is written to an extra log file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>outputPath</strong> (<em>String</em>) – absoulte path to where the log file will be stored.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.logDebug">
<code class="descclassname">benchutils.</code><code class="descname">logDebug</code><span class="sig-paren">(</span><em>message</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#logDebug"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.logDebug" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a log at debug level.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>message</strong> (<em>String</em>) – the log message to print.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.logInfo">
<code class="descclassname">benchutils.</code><code class="descname">logInfo</code><span class="sig-paren">(</span><em>message</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#logInfo"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.logInfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a log at info level.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>message</strong> (<em>String</em>) – the log message to print.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.logWarning">
<code class="descclassname">benchutils.</code><code class="descname">logWarning</code><span class="sig-paren">(</span><em>message</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#logWarning"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.logWarning" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a log at warning level.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>message</strong> (<em>String</em>) – the log message to print.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.logError">
<code class="descclassname">benchutils.</code><code class="descname">logError</code><span class="sig-paren">(</span><em>message</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#logError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.logError" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a log at error level.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>message</strong> (<em>String</em>) – the log message to print.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.createTimeLog">
<code class="descclassname">benchutils.</code><code class="descname">createTimeLog</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#createTimeLog"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.createTimeLog" title="Permalink to this definition">¶</a></dt>
<dd><p>Create the data structure for tracing runtimes of feature selection approaches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the logging data structure.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.flushTimeLog">
<code class="descclassname">benchutils.</code><code class="descname">flushTimeLog</code><span class="sig-paren">(</span><em>timeLogs</em>, <em>outputFilePath</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#flushTimeLog"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.flushTimeLog" title="Permalink to this definition">¶</a></dt>
<dd><p>Write the whole log (of runtimes) to a file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>timeLogs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – the logs in a DataFrame.</li>
<li><strong>outputFilePath</strong> (<em>str</em>) – absolute path to the log file.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.logRuntime">
<code class="descclassname">benchutils.</code><code class="descname">logRuntime</code><span class="sig-paren">(</span><em>timeLogs</em>, <em>start</em>, <em>end</em>, <em>message</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#logRuntime"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.logRuntime" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a runtime log entry and add it to the runtime log data structure.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>timeLogs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – logs to which the new entry should be added</li>
<li><strong>start</strong> (<em>str</em>) – starting time.</li>
<li><strong>end</strong> (<em>str</em>) – ending time.</li>
<li><strong>message</strong> (<em>str</em>) – description of that entry.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">updated logs.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.runRCommand">
<code class="descclassname">benchutils.</code><code class="descname">runRCommand</code><span class="sig-paren">(</span><em>rConfig</em>, <em>scriptName</em>, <em>params</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#runRCommand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.runRCommand" title="Permalink to this definition">¶</a></dt>
<dd><p>Run external R code.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>rConfig</strong> (<em>dict</em>) – R config parameters (store paths to Rscript and the R code).</li>
<li><strong>scriptName</strong> (<em>str</em>) – name of the R script to be executed.</li>
<li><strong>params</strong> (<em>list of str</em>) – list of parameters that will be forwarded to the R script.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.runJavaCommand">
<code class="descclassname">benchutils.</code><code class="descname">runJavaCommand</code><span class="sig-paren">(</span><em>javaConfig</em>, <em>scriptName</em>, <em>params</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#runJavaCommand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.runJavaCommand" title="Permalink to this definition">¶</a></dt>
<dd><p>Run external Java code.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>javaConfig</strong> (<em>dict</em>) – java config parameters (store paths to java and the java code).</li>
<li><strong>scriptName</strong> (<em>str</em>) – name of the jar to be executed.</li>
<li><strong>params</strong> (<em>list of str</em>) – list of parameters that will be forwarded to the jar.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.mapIdentifiers">
<code class="descclassname">benchutils.</code><code class="descname">mapIdentifiers</code><span class="sig-paren">(</span><em>itemList</em>, <em>originalFormat</em>, <em>desiredFormat</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#mapIdentifiers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.mapIdentifiers" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a log entry and add it to the log data structure.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>itemList</strong> (<em>list of str</em>) – list of identifiers, e.g. gene names, to be mapped</li>
<li><strong>originalFormat</strong> (<em>str</em>) – current format of the identifiers.</li>
<li><strong>desiredFormat</strong> (<em>str</em>) – desired format to which the identifiers should be mapped.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">mapping table where every item from itemList is now mapped to desiredFormat.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.mapGeneList">
<code class="descclassname">benchutils.</code><code class="descname">mapGeneList</code><span class="sig-paren">(</span><em>genes</em>, <em>originalFormat</em>, <em>desiredFormat</em>, <em>outputFile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#mapGeneList"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.mapGeneList" title="Permalink to this definition">¶</a></dt>
<dd><p>Map a list of genes to the desired format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>genes</strong> (<em>list of str</em>) – list of gene names to be mapped</li>
<li><strong>originalFormat</strong> (<em>str</em>) – current format of the gene names.</li>
<li><strong>desiredFormat</strong> (<em>str</em>) – desired format to which the gene names should be mapped.</li>
<li><strong>outputFile</strong> (<em>str</em>) – absolute path to the output file in which the mapping should be stored.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">list of mapped gene names.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list of str</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.mapRanking">
<code class="descclassname">benchutils.</code><code class="descname">mapRanking</code><span class="sig-paren">(</span><em>ranking</em>, <em>originalFormat</em>, <em>desiredFormat</em>, <em>outputFile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#mapRanking"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.mapRanking" title="Permalink to this definition">¶</a></dt>
<dd><p>Map the feature names of a ranking to the desired format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>ranking</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – DataFrame of the ranking.</li>
<li><strong>originalFormat</strong> (<em>str</em>) – current format of the feature names in the ranking.</li>
<li><strong>desiredFormat</strong> (<em>str</em>) – desired format to which the feature names should be mapped.</li>
<li><strong>outputFile</strong> (<em>str</em>) – absolute path to the output file in which the mapped feature ranking should be stored.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">mapped feature ranking.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.retrieveMappings">
<code class="descclassname">benchutils.</code><code class="descname">retrieveMappings</code><span class="sig-paren">(</span><em>itemList</em>, <em>originalFormat</em>, <em>desiredFormat</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#retrieveMappings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.retrieveMappings" title="Permalink to this definition">¶</a></dt>
<dd><p>Query the knowledge base to map the identifiers.
We have mapping via BiomaRt and gConvert available.
gConvert is currently used because BiomaRt is unstable and blocks when parallel queries are sent.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>itemList</strong> (<em>list of str</em>) – list of identifier names to be mapped</li>
<li><strong>originalFormat</strong> (<em>str</em>) – current format of the identifiers.</li>
<li><strong>desiredFormat</strong> (<em>str</em>) – desired format to which the identifiers should be mapped.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">mapping table for all identifiers.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="benchutils.mapDataMatrix">
<code class="descclassname">benchutils.</code><code class="descname">mapDataMatrix</code><span class="sig-paren">(</span><em>inputMatrix</em>, <em>genesInColumns</em>, <em>originalFormat</em>, <em>desiredFormat</em>, <em>outputFile</em>, <em>labeled</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/benchutils.html#mapDataMatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#benchutils.mapDataMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Map the features of a data set to the desired format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>inputMatrix</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – DataFrame of the ranking.</li>
<li><strong>genesInColumns</strong> (<em>bool</em>) – if the genes/features are located in the columns.</li>
<li><strong>originalFormat</strong> (<em>str</em>) – current format of the feature names in the data set.</li>
<li><strong>desiredFormat</strong> (<em>str</em>) – desired format to which the feature names should be mapped.</li>
<li><strong>outputFile</strong> (<em>str</em>) – absolute path to the output file in which the mapped data set should be stored.</li>
<li><strong>labeled</strong> (<em>bool</em>) – if the data matrix is additionally labeled.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">mapped data set.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="preprocessing-module">
<h2>preprocessing module<a class="headerlink" href="#preprocessing-module" title="Permalink to this headline">¶</a></h2>
<p>Contains all classes related to preprocessing.
All classes providing preprocessing functionality have to inherit from the abstract class <a class="reference internal" href="#preprocessing.Preprocessor" title="preprocessing.Preprocessor"><code class="xref py py-class docutils literal notranslate"><span class="pre">preprocessing.Preprocessor</span></code></a> and implement its <a class="reference internal" href="#preprocessing.Preprocessor.preprocess" title="preprocessing.Preprocessor.preprocess"><code class="xref py py-meth docutils literal notranslate"><span class="pre">preprocessing.Preprocessor.preprocess()</span></code></a> method.
For a detailed look at the class architecture, have a look at ADD CLASS ARCHITECTURE LINK HERE.</p>
<span class="target" id="module-preprocessing"></span><dl class="class">
<dt id="preprocessing.Preprocessor">
<em class="property">class </em><code class="descclassname">preprocessing.</code><code class="descname">Preprocessor</code><span class="sig-paren">(</span><em>input</em>, <em>metadata</em>, <em>output</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#Preprocessor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.Preprocessor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Super class of all preprocessor implementations.
Inherit from this class and implement <a class="reference internal" href="#preprocessing.Preprocessor.preprocess" title="preprocessing.Preprocessor.preprocess"><code class="xref py py-meth docutils literal notranslate"><span class="pre">preprocessing.Preprocessor.preprocess()</span></code></a> if you want to add a new preprocessor class.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to the input file.</li>
<li><strong>metadata</strong> (<em>str</em>) – absolute path to the metadata file.</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to the output directory.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="preprocessing.Preprocessor.preprocess">
<code class="descname">preprocess</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#Preprocessor.preprocess"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.Preprocessor.preprocess" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract method.
Interface method that is invoked externally to trigger preprocessing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the preprocessed output file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="preprocessing.MappingPreprocessor">
<em class="property">class </em><code class="descclassname">preprocessing.</code><code class="descname">MappingPreprocessor</code><span class="sig-paren">(</span><em>input</em>, <em>output</em>, <em>currentFormat</em>, <em>desiredFormat</em>, <em>labeled</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#MappingPreprocessor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.MappingPreprocessor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#preprocessing.Preprocessor" title="preprocessing.Preprocessor"><code class="xref py py-class docutils literal notranslate"><span class="pre">preprocessing.Preprocessor</span></code></a></p>
<p>Maps the input data set to a desired format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to the input file.</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to the output directory.</li>
<li><strong>currentFormat</strong> (<em>str</em>) – current identifier format.</li>
<li><strong>desiredFormat</strong> (<em>str</em>) – desired identifier format.</li>
<li><strong>labeled</strong> (<em>bool</em>) – boolean value if the input data is labeled.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="preprocessing.MappingPreprocessor.preprocess">
<code class="descname">preprocess</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#MappingPreprocessor.preprocess"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.MappingPreprocessor.preprocess" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps the identifiers in the input dataset to the desired format that was specified when constructing the preprocessor.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the mapped file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="preprocessing.FilterPreprocessor">
<em class="property">class </em><code class="descclassname">preprocessing.</code><code class="descname">FilterPreprocessor</code><span class="sig-paren">(</span><em>input</em>, <em>metadata</em>, <em>output</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#FilterPreprocessor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.FilterPreprocessor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#preprocessing.Preprocessor" title="preprocessing.Preprocessor"><code class="xref py py-class docutils literal notranslate"><span class="pre">preprocessing.Preprocessor</span></code></a></p>
<p>Filters features or samples above a user-defined threshold of missing values.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to the input file.</li>
<li><strong>metadata</strong> (<em>str</em>) – absolute path to the metadata file.</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to the output directory.</li>
<li><strong>config</strong> (<em>str</em>) – configuration parameter for preprocessing as specified in the config file.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="preprocessing.FilterPreprocessor.preprocess">
<code class="descname">preprocess</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#FilterPreprocessor.preprocess"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.FilterPreprocessor.preprocess" title="Permalink to this definition">¶</a></dt>
<dd><p>Depending on what is specified in the config file, filter samples and/or features.
Remove all samples/features that have missing values above the threshold specified in the config.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the filtered output file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="preprocessing.FilterPreprocessor.filterMissings">
<code class="descname">filterMissings</code><span class="sig-paren">(</span><em>threshold</em>, <em>data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#FilterPreprocessor.filterMissings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.FilterPreprocessor.filterMissings" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter the data for entries that have missing information above the given threshold.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>threshold</strong> (<em>str</em>) – maximum percentage of allowed missing items as string.</li>
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – a DataFrame to be filtered</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">filtered DataFrame.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="preprocessing.DataTransformationPreprocessor">
<em class="property">class </em><code class="descclassname">preprocessing.</code><code class="descname">DataTransformationPreprocessor</code><span class="sig-paren">(</span><em>input</em>, <em>metadata</em>, <em>output</em>, <em>dataSeparator</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#DataTransformationPreprocessor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.DataTransformationPreprocessor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#preprocessing.Preprocessor" title="preprocessing.Preprocessor"><code class="xref py py-class docutils literal notranslate"><span class="pre">preprocessing.Preprocessor</span></code></a></p>
<p>Transform the input data to have features in the columns for subsequent processing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to the input file.</li>
<li><strong>metadata</strong> (<em>str</em>) – absolute path to the metadata file.</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to the output directory.</li>
<li><strong>dataSeparator</strong> (<em>str</em>) – delimiter to use when parsing the input file.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="preprocessing.DataTransformationPreprocessor.preprocess">
<code class="descname">preprocess</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#DataTransformationPreprocessor.preprocess"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.DataTransformationPreprocessor.preprocess" title="Permalink to this definition">¶</a></dt>
<dd><p>If not already so, transpose the input data to have the features in the columns.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the correctly formatted output file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="preprocessing.MetaDataPreprocessor">
<em class="property">class </em><code class="descclassname">preprocessing.</code><code class="descname">MetaDataPreprocessor</code><span class="sig-paren">(</span><em>input</em>, <em>metadata</em>, <em>output</em>, <em>separator</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#MetaDataPreprocessor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.MetaDataPreprocessor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#preprocessing.Preprocessor" title="preprocessing.Preprocessor"><code class="xref py py-class docutils literal notranslate"><span class="pre">preprocessing.Preprocessor</span></code></a></p>
<p>Add labels to input data.
Get labels from meta data attribute that was specified in the user config.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to the input file.</li>
<li><strong>metadata</strong> (<em>str</em>) – absolute path to the metadata file.</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to the output directory.</li>
<li><strong>dataSeparator</strong> (<em>str</em>) – delimiter to use when parsing the input and metadata file.</li>
<li><strong>diseaseColumn</strong> (<em>str</em>) – column name of the class labels.</li>
<li><strong>transposeMetadataMatrix</strong> (<em>bool</em>) – boolean value if the identifier names are located in the columns, as specified in the config file.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="preprocessing.MetaDataPreprocessor.preprocess">
<code class="descname">preprocess</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#MetaDataPreprocessor.preprocess"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.MetaDataPreprocessor.preprocess" title="Permalink to this definition">¶</a></dt>
<dd><p>Labels all samples of a data set.
Labels are taken from the corresponding metadata file and the metadata attribute that was specified in the config file.
Samples without metadata information well be assigned to class “NotAvailable”.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the labeled data set.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="preprocessing.DataMovePreprocessor">
<em class="property">class </em><code class="descclassname">preprocessing.</code><code class="descname">DataMovePreprocessor</code><span class="sig-paren">(</span><em>input</em>, <em>output</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#DataMovePreprocessor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.DataMovePreprocessor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#preprocessing.Preprocessor" title="preprocessing.Preprocessor"><code class="xref py py-class docutils literal notranslate"><span class="pre">preprocessing.Preprocessor</span></code></a></p>
<p>Moves the input data set to the specified location.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to the input file.</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to the output directory.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="preprocessing.DataMovePreprocessor.preprocess">
<code class="descname">preprocess</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/preprocessing.html#DataMovePreprocessor.preprocess"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preprocessing.DataMovePreprocessor.preprocess" title="Permalink to this definition">¶</a></dt>
<dd><p>Moves a file (self.input) to another location (self.output).
Typically used at the end of preprocessing, when the final data set is moved to a new location for the actual analysis.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the new file location.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="featureselection-module">
<h2>featureselection module<a class="headerlink" href="#featureselection-module" title="Permalink to this headline">¶</a></h2>
<p>Contains all classes related to feature selection.
Each feature selection approach must be implemented in its own class inheriting from the abstract super class <a class="reference internal" href="#featureselection.FeatureSelector" title="featureselection.FeatureSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.FeatureSelector</span></code></a> or one of its abstract subclasses, e.g. for including R or Java code.
Each feature selection class must implement setParams() and selectFeatures(), as input or output parameters are just set at runtime.</p>
<p>Feature extraction methods are implemented in the same structure, except that they need to have an instance of a class inheriting from <code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.PathwayMapper</span></code> assigned to them so that the feature space can be transformed from the original to the new, e.g. pathways.</p>
<p>The creation of feature selectors is encapsulated by the class <a class="reference internal" href="#featureselection.FeatureSelectorFactory" title="featureselection.FeatureSelectorFactory"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.FeatureSelectorFactory</span></code></a> that takes care that every selector is equipped correspondingly, e.g. with a knowledge base or another feature selector.
For a detailed look at the class architecture and the inheritance structure, have a look at ADD CLASS ARCHITECTURE LINK HERE.</p>
<span class="target" id="module-featureselection"></span><dl class="class">
<dt id="featureselection.FeatureSelectorFactory">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">FeatureSelectorFactory</code><a class="reference internal" href="_modules/featureselection.html#FeatureSelectorFactory"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelectorFactory" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Singleton class.
Python code encapsulates it in a way that is not shown in Sphinx, so have a look at the descriptions in the source code.</p>
<p>Creates feature selector object based on a given name.
New feature selection approaches must be registered here.
Names for feature selectors must follow to a particular scheme, with keywords separated by _:
- first keyword is the actual selector name
- if needed, second keyword is the knowledge base
- if needed, third keyword is the (traditional) approach to be combined
Examples:
- Traditional Approaches have only one keyword, e.g. InfoGain or ANOVA
- LassoPenalty_KEGG provides KEGG information to the LassoPenalty feature selection approach
- Weighted_KEGG_InfoGain –&gt; Factory creates an instance of KBweightedSelector which uses KEGG as knowledge base and InfoGain as traditional selector.
While the focus here lies on the combination of traditional approaches with prior biological knowledge, it is theoretically possible to use ANY selector object for combination that inherits from <a class="reference internal" href="#featureselection.FeatureSelector" title="featureselection.FeatureSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">FeatureSelector</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>config</strong> (<em>dict</em>) – configuration parameters for UMLS web service as specified in config file.</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="featureselection.FeatureSelectorFactory.instance">
<code class="descname">instance</code><em class="property"> = None</em><a class="headerlink" href="#featureselection.FeatureSelectorFactory.instance" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.FeatureSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">FeatureSelector</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Abstract super class for feature selection functionality.
Every feature selection class has to inherit from this class and implement its <a class="reference internal" href="#featureselection.FeatureSelector.selectFeatures" title="featureselection.FeatureSelector.selectFeatures"><code class="xref py py-meth docutils literal notranslate"><span class="pre">FeatureSelector.selectFeatures()</span></code></a> method and - if necessary - its <a class="reference internal" href="#featureselection.FeatureSelector.setParams" title="featureselection.FeatureSelector.setParams"><code class="xref py py-meth docutils literal notranslate"><span class="pre">FeatureSelector.setParams()</span></code></a> method.
Once created, feature selection can be triggered by first setting parameters (input, output, etc) as needed with <a class="reference internal" href="#featureselection.FeatureSelector.setParams" title="featureselection.FeatureSelector.setParams"><code class="xref py py-meth docutils literal notranslate"><span class="pre">FeatureSelector.setParams()</span></code></a>.
The actual feature selection is triggered by invoking <a class="reference internal" href="#featureselection.FeatureSelector.selectFeatures" title="featureselection.FeatureSelector.selectFeatures"><code class="xref py py-meth docutils literal notranslate"><span class="pre">FeatureSelector.selectFeatures()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to input dataset.</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to output directory (where the ranking will be stored).</li>
<li><strong>dataset</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – the dataset for which to select features. Will be loaded dynamically based on self.input at first usage.</li>
<li><strong>dataConfig</strong> (<em>dict</em>) – config parameters for input data set.</li>
<li><strong>name</strong> (<em>str</em>) – selector name</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="featureselection.FeatureSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract.
Invoke feature selection functionality in this method when implementing a new selector</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the output ranking file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureSelector.getTimeLogs">
<code class="descname">getTimeLogs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.getTimeLogs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.getTimeLogs" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets all logs for this selector.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">dataframe of logged events containing start/end time, duration, and a short description.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureSelector.setTimeLogs">
<code class="descname">setTimeLogs</code><span class="sig-paren">(</span><em>newTimeLogs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.setTimeLogs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.setTimeLogs" title="Permalink to this definition">¶</a></dt>
<dd><p>Overwrites the current logs with new ones.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>newTimeLogs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – new dataframe of logged events containing start/end time, duration, and a short description.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureSelector.disableLogFlush">
<code class="descname">disableLogFlush</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.disableLogFlush"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.disableLogFlush" title="Permalink to this definition">¶</a></dt>
<dd><p>Disables log flushing (i.e., writing the log to a separate file) of the selector at the end of feature selection.
This is needed when a <a class="reference internal" href="#featureselection.CombiningSelector" title="featureselection.CombiningSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">CombiningSelector</span></code></a> uses a second selector and wants to avoid that its log messages are written, potentially overwriting logs from another selector of the same name.</p>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureSelector.enableLogFlush">
<code class="descname">enableLogFlush</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.enableLogFlush"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.enableLogFlush" title="Permalink to this definition">¶</a></dt>
<dd><p>Enables log flushing, i.e. writing the logs to a separate file at the end of feature selection.</p>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureSelector.getName">
<code class="descname">getName</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.getName"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.getName" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the selector’s name.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">selector name.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureSelector.getData">
<code class="descname">getData</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.getData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.getData" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the labeled dataset from which to select features.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">dataframe containing the dataset with class labels.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureSelector.getUnlabeledData">
<code class="descname">getUnlabeledData</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.getUnlabeledData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.getUnlabeledData" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the dataset without labels.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">dataframe containing the dataset without class labels.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureSelector.getFeatures">
<code class="descname">getFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.getFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.getFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets features from the dataset.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of features.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureSelector.getUniqueLabels">
<code class="descname">getUniqueLabels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.getUniqueLabels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.getUniqueLabels" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the unique class labels available in the dataset.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of distinct class labels.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureSelector.getLabels">
<code class="descname">getLabels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.getLabels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.getLabels" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the labels in the data set.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">all labels from the dataset.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureSelector.setParams">
<code class="descname">setParams</code><span class="sig-paren">(</span><em>inputPath</em>, <em>outputDir</em>, <em>loggingDir</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.setParams"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.setParams" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets parameters for the feature selection run: path to the input datast and path to the output directory.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>inputPath</strong> (<em>str</em>) – absolute path to the input file containing the dataset for analysis.</li>
<li><strong>outputDir</strong> (<em>str</em>) – absolute path to the output directory (where to store the ranking)</li>
<li><strong>loggingDir</strong> (<em>str</em>) – absolute path to the logging directory (where to store log files)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureSelector.writeRankingToFile">
<code class="descname">writeRankingToFile</code><span class="sig-paren">(</span><em>ranking</em>, <em>outputFile</em>, <em>index=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureSelector.writeRankingToFile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureSelector.writeRankingToFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a given ranking to a specified file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>ranking</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – dataframe with the ranking.</li>
<li><strong>outputFile</strong> (<em>str</em>) – absolute path of the file where ranking will be stored.</li>
<li><strong>index</strong> (<em>bool</em><em>, </em><em>default False</em>) – whether to write the dataframe’s index or not.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.PythonSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">PythonSelector</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PythonSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PythonSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.FeatureSelector" title="featureselection.FeatureSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.FeatureSelector</span></code></a></p>
<p>Abstract.
Inherit from this class when implementing a feature selector using any of scikit-learn’s functionality.
As functionality invocation, input preprocessing and output postprocessing are typically very similar/the same for such implementations, this class already encapsulates it.
Instead of implementing <a class="reference internal" href="#featureselection.PythonSelector.selectFeatures" title="featureselection.PythonSelector.selectFeatures"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PythonSelector.selectFeatures()</span></code></a>, implement <a class="reference internal" href="#featureselection.PythonSelector.runSelector" title="featureselection.PythonSelector.runSelector"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PythonSelector.runSelector()</span></code></a>.</p>
<dl class="method">
<dt id="featureselection.PythonSelector.runSelector">
<code class="descname">runSelector</code><span class="sig-paren">(</span><em>data</em>, <em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PythonSelector.runSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PythonSelector.runSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract - implement this method when inheriting from this class.
Runs the actual feature selector of scikit-learn.
Is invoked by <a class="reference internal" href="#featureselection.PythonSelector.selectFeatures" title="featureselection.PythonSelector.selectFeatures"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PythonSelector.selectFeatures()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – dataframe containing the unlabeled dataset.</li>
<li><strong>labels</strong> (<em>list of int</em>) – numerically encoded class labels.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">sklearn/mlxtend selector that ran the selection (containing coefficients etc.).</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.PythonSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PythonSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PythonSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Executes the feature selection procedure.
Prepares the input data set to match scikit-learn’s expected formats and postprocesses the output to create a ranking.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the output ranking file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.PythonSelector.prepareInput">
<code class="descname">prepareInput</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PythonSelector.prepareInput"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PythonSelector.prepareInput" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepares the input data set before running any of scikit-learn’s selectors.
Removes the labels from the input data set and encodes the labels in numbers.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">dataset (without labels) and labels encoded in numbers.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> and list of int</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.PythonSelector.prepareOutput">
<code class="descname">prepareOutput</code><span class="sig-paren">(</span><em>outputFile</em>, <em>data</em>, <em>selector</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PythonSelector.prepareOutput"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PythonSelector.prepareOutput" title="Permalink to this definition">¶</a></dt>
<dd><p>Transforms the selector output to a valid ranking and stores it into the specified file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>outputFile</strong> (<em>str</em>) – absolute path of the file to which to write the ranking.</li>
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – input dataset.</li>
<li><strong>selector</strong> – selector object from scikit-learn.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.RSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">RSelector</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#RSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.RSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.FeatureSelector" title="featureselection.FeatureSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.FeatureSelector</span></code></a></p>
<p>Selector class for invoking R code for feature selection.
Inherit from this class if you want to use R code, implement <a class="reference internal" href="#featureselection.RSelector.createParams" title="featureselection.RSelector.createParams"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RSelector.createParams()</span></code></a> with what your script requires, and set self.scriptName accordingly.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>rConfig</strong> (<em>dict</em>) – config parameters to execute R code.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="featureselection.RSelector.createParams">
<code class="descname">createParams</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#RSelector.createParams"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.RSelector.createParams" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract.
Implement this method to set the parameters your R script requires.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>filename</strong> (<em>str</em>) – absolute path of the output file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of parameters to use for R code execution, e.g. input and output filenames.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.RSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#RSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.RSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Triggers the feature selection.
Actually a wrapper method that invokes external R code.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the result file containing the ranking.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.JavaSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">JavaSelector</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#JavaSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.JavaSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.FeatureSelector" title="featureselection.FeatureSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.FeatureSelector</span></code></a></p>
<p>Selector class for invoking R code for feature selection.
Inherit from this class if you want to use R code, implement <a class="reference internal" href="#featureselection.RSelector.createParams" title="featureselection.RSelector.createParams"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RSelector.createParams()</span></code></a> with what your script requires, and set self.scriptName accordingly.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>javaConfig</strong> (<em>dict</em>) – config parameters to execute java code.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="featureselection.JavaSelector.createParams">
<code class="descname">createParams</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#JavaSelector.createParams"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.JavaSelector.createParams" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract.
Implement this method to set the parameters your java code requires.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of parameters to use for java code execution, e.g. input and output filenames.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.JavaSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#JavaSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.JavaSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Triggers the feature selection.
Actually a wrapper method that invokes external java code.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the result file containing the ranking.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.PriorKnowledgeSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">PriorKnowledgeSelector</code><span class="sig-paren">(</span><em>name</em>, <em>knowledgebase</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PriorKnowledgeSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PriorKnowledgeSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.FeatureSelector" title="featureselection.FeatureSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.FeatureSelector</span></code></a></p>
<p>Super class for all prior knowledge approaches.
If you want to implement an own prior knowledge approach that uses a knowledge base (but not a second selector and no network approaches), inherit from this class.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>knowledgebase</strong> (<a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a> or inheriting class) – instance of a knowledge base.</li>
<li><strong>alternativeSearchTerms</strong> (<em>list of str</em>) – list of alternative search terms to use for querying the knowledge base.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="featureselection.PriorKnowledgeSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PriorKnowledgeSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PriorKnowledgeSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract.
Implement this method when inheriting from this class.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the output ranking file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.PriorKnowledgeSelector.collectAlternativeSearchTerms">
<code class="descname">collectAlternativeSearchTerms</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PriorKnowledgeSelector.collectAlternativeSearchTerms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PriorKnowledgeSelector.collectAlternativeSearchTerms" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets all alternative search terms that were specified in the config file and put them into a list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of alternative search terms to use for querying the knowledge base.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.PriorKnowledgeSelector.getSearchTerms">
<code class="descname">getSearchTerms</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PriorKnowledgeSelector.getSearchTerms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PriorKnowledgeSelector.getSearchTerms" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets all search terms to use for querying a knowledge base.
Search terms that will be used are a) the class labels in the dataset, and b) the alternative search terms that were specified in the config file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of search terms to use for querying the knowledge base.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.PriorKnowledgeSelector.getName">
<code class="descname">getName</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PriorKnowledgeSelector.getName"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PriorKnowledgeSelector.getName" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the full name (including applied knowledge base) of this selector.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">selector name.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.CombiningSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">CombiningSelector</code><span class="sig-paren">(</span><em>name</em>, <em>knowledgebase</em>, <em>tradApproach</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#CombiningSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.CombiningSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.PriorKnowledgeSelector" title="featureselection.PriorKnowledgeSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.PriorKnowledgeSelector</span></code></a></p>
<p>Super class for prior knoweldge approaches that use a knowledge base AND combine it with any kind of selector, e.g. a traditional approach.
Inherit from this class if you want to implement a feature selector that requires both a knowledge base and another selector, e.g. because it combines information from both.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>knowledgebase</strong> (<a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a> or inheriting class) – instance of a knowledge base.</li>
<li><strong>tradApproach</strong> (<a class="reference internal" href="#featureselection.FeatureSelector" title="featureselection.FeatureSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">FeatureSelector</span></code></a>) – any feature selector implementation to use internally, e.g. a traditional approach like ANOVA</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="featureselection.CombiningSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#CombiningSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.CombiningSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract.
Implement this method as desired when inheriting from this class.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the output ranking file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.CombiningSelector.getName">
<code class="descname">getName</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#CombiningSelector.getName"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.CombiningSelector.getName" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the full name (including applied knowledge base and feature selector) of this selector.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">selector name.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.CombiningSelector.getExternalGenes">
<code class="descname">getExternalGenes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#CombiningSelector.getExternalGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.CombiningSelector.getExternalGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets all genes related to the provided search terms from the knowledge base.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of gene names.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.NetworkSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">NetworkSelector</code><span class="sig-paren">(</span><em>name</em>, <em>knowledgebase</em>, <em>featuremapper</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#NetworkSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.NetworkSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.PriorKnowledgeSelector" title="featureselection.PriorKnowledgeSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.PriorKnowledgeSelector</span></code></a></p>
<p>Abstract.
Inherit from this method if you want to implement a new network approach that actually conducts feature EXTRACTION, i.e. maps the original data set to have pathway/subnetworks.
Instead of  <a class="reference internal" href="#featureselection.FeatureSelector.selectFeatures" title="featureselection.FeatureSelector.selectFeatures"><code class="xref py py-meth docutils literal notranslate"><span class="pre">FeatureSelector.selectFeatures()</span></code></a> implement <a class="reference internal" href="#featureselection.NetworkSelector.selectPathways" title="featureselection.NetworkSelector.selectPathways"><code class="xref py py-meth docutils literal notranslate"><span class="pre">NetworkSelector.selectPathways()</span></code></a> when inheriting from this class.</p>
<p>Instances of <a class="reference internal" href="#featureselection.NetworkSelector" title="featureselection.NetworkSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">NetworkSelector</span></code></a> and inheriting classes also require a <code class="xref py py-class docutils literal notranslate"><span class="pre">PathwayMapper</span></code> object that transfers the dataset to the new feature space.
Custom implementations thus need to implement a) a selection strategy to select pathways and b) a mapping strategy to compute new feature values for the selected pathways.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>featureMapper</strong> (<a class="reference internal" href="#featureselection.FeatureMapper" title="featureselection.FeatureMapper"><code class="xref py py-class docutils literal notranslate"><span class="pre">FeatureMapper</span></code></a> or inheriting class) – feature mapping object that transfers the feature space.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="featureselection.NetworkSelector.selectPathways">
<code class="descname">selectPathways</code><span class="sig-paren">(</span><em>pathways</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#NetworkSelector.selectPathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.NetworkSelector.selectPathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Selects the pathways that will become the new features of the data set.
Implement this method (instead of <a class="reference internal" href="#featureselection.FeatureSelector.selectFeatures" title="featureselection.FeatureSelector.selectFeatures"><code class="xref py py-meth docutils literal notranslate"><span class="pre">FeatureSelector.selectFeatures()</span></code></a> when inheriting from this class.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pathways</strong> (<em>dict</em>) – dict of pathways (pathway names as keys) to select from.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">pathway ranking as dataframe</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.NetworkSelector.writeMappedFile">
<code class="descname">writeMappedFile</code><span class="sig-paren">(</span><em>mapped_data</em>, <em>fileprefix</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#NetworkSelector.writeMappedFile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.NetworkSelector.writeMappedFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the mapped dataset with new feature values to the same directory as the original file is located (it will be automatically processed then).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>mapped_data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – dataframe containing the dataset with mapped feature space.</li>
<li><strong>fileprefix</strong> (<em>str</em>) – prefix of the file name, e.g. the directory path</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">absolute path of the file name to store the mapped data set.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">str</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.NetworkSelector.getName">
<code class="descname">getName</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#NetworkSelector.getName"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.NetworkSelector.getName" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the selector name (including the knowledge base).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">selector name.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.NetworkSelector.filterPathways">
<code class="descname">filterPathways</code><span class="sig-paren">(</span><em>pathways</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#NetworkSelector.filterPathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.NetworkSelector.filterPathways" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="featureselection.NetworkSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#NetworkSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.NetworkSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Instead of selecting existing features, instances of <a class="reference internal" href="#featureselection.NetworkSelector" title="featureselection.NetworkSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">NetworkSelector</span></code></a> select pathways or submodules as features.
For that, it first queries its knowledge base for pathways.
It then selects the top k pathways (strategy to be implemented in <a class="reference internal" href="#featureselection.NetworkSelector.selectPathways" title="featureselection.NetworkSelector.selectPathways"><code class="xref py py-meth docutils literal notranslate"><span class="pre">NetworkSelector.selectPathways()</span></code></a>) and subsequently maps the dataset to its new feature space.
The mapping will be conducted by an object of <code class="xref py py-class docutils literal notranslate"><span class="pre">PathwayMapper</span></code> or inheriting classes.
If a second dataset for cross-validation is available, the feature space of this dataset will also be transformed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the pathway ranking.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.RandomSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">RandomSelector</code><a class="reference internal" href="_modules/featureselection.html#RandomSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.RandomSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.FeatureSelector" title="featureselection.FeatureSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.FeatureSelector</span></code></a></p>
<p>Baseline Selector: Randomly selects any features.</p>
<dl class="method">
<dt id="featureselection.RandomSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#RandomSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.RandomSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Randomly select any features from the feature space.
Assigns a score of 0.0 to every feature</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the ranking file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.AnovaSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">AnovaSelector</code><a class="reference internal" href="_modules/featureselection.html#AnovaSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.AnovaSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.PythonSelector" title="featureselection.PythonSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.PythonSelector</span></code></a></p>
<p>Runs ANOVA feature selection using scikit-learn implementation</p>
<dl class="method">
<dt id="featureselection.AnovaSelector.runSelector">
<code class="descname">runSelector</code><span class="sig-paren">(</span><em>data</em>, <em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#AnovaSelector.runSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.AnovaSelector.runSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs the ANOVA feature selector of scikit-learn.
Is invoked by <a class="reference internal" href="#featureselection.PythonSelector.selectFeatures" title="featureselection.PythonSelector.selectFeatures"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PythonSelector.selectFeatures()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – dataframe containing the unlabeled dataset.</li>
<li><strong>labels</strong> (<em>list of int</em>) – numerically encoded class labels.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">sklearn/mlxtend selector that ran the selection (containing coefficients etc.).</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.Variance2Selector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">Variance2Selector</code><a class="reference internal" href="_modules/featureselection.html#Variance2Selector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.Variance2Selector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.PythonSelector" title="featureselection.PythonSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.PythonSelector</span></code></a></p>
<p>Runs variance-based feature selection using scikit-learn.</p>
<dl class="method">
<dt id="featureselection.Variance2Selector.prepareOutput">
<code class="descname">prepareOutput</code><span class="sig-paren">(</span><em>outputFile</em>, <em>data</em>, <em>selector</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#Variance2Selector.prepareOutput"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.Variance2Selector.prepareOutput" title="Permalink to this definition">¶</a></dt>
<dd><p>Transforms the selector output to a valid ranking and stores it into the specified file.
We need to override this method because variance selector has no attribute scores but variances.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>outputFile</strong> (<em>str</em>) – absolute path of the file to which to write the ranking.</li>
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – input dataset.</li>
<li><strong>selector</strong> – selector object from scikit-learn.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.Variance2Selector.runSelector">
<code class="descname">runSelector</code><span class="sig-paren">(</span><em>data</em>, <em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#Variance2Selector.runSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.Variance2Selector.runSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs the actual variance-based feature selector of scikit-learn.
Is invoked by <a class="reference internal" href="#featureselection.PythonSelector.selectFeatures" title="featureselection.PythonSelector.selectFeatures"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PythonSelector.selectFeatures()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – dataframe containing the unlabeled dataset.</li>
<li><strong>labels</strong> (<em>list of int</em>) – numerically encoded class labels.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">sklearn/mlxtend selector that ran the selection (containing coefficients etc.).</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.MRMRSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">MRMRSelector</code><a class="reference internal" href="_modules/featureselection.html#MRMRSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.MRMRSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.RSelector" title="featureselection.RSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.RSelector</span></code></a></p>
<p>Runs maximum Relevance minimum Redundancy (mRMR) feature selection using the mRMRe R implementation: <a class="reference external" href="https://cran.r-project.org/web/packages/mRMRe/index.html">https://cran.r-project.org/web/packages/mRMRe/index.html</a>
Actually a wrapper class for invoking the R code.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>scriptName</strong> (<em>str</em>) – name of the R script to invoke.</li>
<li><strong>maxFeatures</strong> (<em>int</em>) – maximum number of features to select. Currently all features (=0) are ranked..</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="featureselection.MRMRSelector.createParams">
<code class="descname">createParams</code><span class="sig-paren">(</span><em>outputFile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#MRMRSelector.createParams"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.MRMRSelector.createParams" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the parameters the R script requires (input file, output file, maximum number of features).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of parameters to use for mRMR execution in R.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.VarianceSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">VarianceSelector</code><a class="reference internal" href="_modules/featureselection.html#VarianceSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.VarianceSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.RSelector" title="featureselection.RSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.RSelector</span></code></a></p>
<p>Runs variance-based feature selection using R genefilter library.
Actually a wrapper class for invoking the R code.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>scriptName</strong> (<em>str</em>) – name of the R script to invoke.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="featureselection.VarianceSelector.createParams">
<code class="descname">createParams</code><span class="sig-paren">(</span><em>outputFile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#VarianceSelector.createParams"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.VarianceSelector.createParams" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the parameters the R script requires (input file, output file).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>outputFile</strong> (<em>str</em>) – absolute path to the output file that will contain the ranking.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of parameters to use for mRMR execution in R.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.InfoGainSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">InfoGainSelector</code><a class="reference internal" href="_modules/featureselection.html#InfoGainSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.InfoGainSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.JavaSelector" title="featureselection.JavaSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.JavaSelector</span></code></a></p>
<p>Runs InfoGain feature selection as provided by WEKA: <a class="reference external" href="https://www.cs.waikato.ac.nz/ml/weka/">https://www.cs.waikato.ac.nz/ml/weka/</a>
Actually a wrapper class for invoking java code.</p>
<dl class="method">
<dt id="featureselection.InfoGainSelector.createParams">
<code class="descname">createParams</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#InfoGainSelector.createParams"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.InfoGainSelector.createParams" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the parameters the java program requires (input file, output file, selector name).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of parameters to use for InfoGain execution in java.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.ReliefFSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">ReliefFSelector</code><a class="reference internal" href="_modules/featureselection.html#ReliefFSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.ReliefFSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.JavaSelector" title="featureselection.JavaSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.JavaSelector</span></code></a></p>
<p>Runs ReliefF feature selection as provided by WEKA: <a class="reference external" href="https://www.cs.waikato.ac.nz/ml/weka/">https://www.cs.waikato.ac.nz/ml/weka/</a>
Actually a wrapper class for invoking java code.</p>
<dl class="method">
<dt id="featureselection.ReliefFSelector.createParams">
<code class="descname">createParams</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#ReliefFSelector.createParams"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.ReliefFSelector.createParams" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the parameters the java program requires (input file, output file, selector name).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of parameters to use for InfoGain execution in java.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.KbSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">KbSelector</code><span class="sig-paren">(</span><em>knowledgebase</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#KbSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.KbSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.PriorKnowledgeSelector" title="featureselection.PriorKnowledgeSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.PriorKnowledgeSelector</span></code></a></p>
<p>Knowledge base selector.
Selects features exclusively based the information retrieved from a knowledge base.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>knowledgebase</strong> (<a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a>) – instance of a knowledge base.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="featureselection.KbSelector.updateScores">
<code class="descname">updateScores</code><span class="sig-paren">(</span><em>entry</em>, <em>newGeneScores</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#KbSelector.updateScores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.KbSelector.updateScores" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates a score entry with the new score retrieved from the knowledge base.
Used by apply function.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>entry</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code>) – a gene score entry consisting of the gene name and its score</li>
<li><strong>newGeneScores</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – dataframe containing gene scores retrieved from the knowledge base.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">updated series element.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.KbSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#KbSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.KbSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Does the actual feature selection.
Retrieves association scores for genes from the knowledge base based on the given search terms.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the resulting ranking file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.KBweightedSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">KBweightedSelector</code><span class="sig-paren">(</span><em>knowledgebase</em>, <em>tradApproach</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#KBweightedSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.KBweightedSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.CombiningSelector" title="featureselection.CombiningSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.CombiningSelector</span></code></a></p>
<p>Selects features based on association scores retrieved from the knowledge base and the relevance score retrieved by the (traditional) approach.
Computes the final score via tradScore * assocScore.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>knowledgebase</strong> (<a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a> or inheriting class) – instance of a knowledge base.</li>
<li><strong>tradApproach</strong> (<a class="reference internal" href="#featureselection.FeatureSelector" title="featureselection.FeatureSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">FeatureSelector</span></code></a>) – any feature selector implementation to use internally, e.g. a traditional approach like ANOVA</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="featureselection.KBweightedSelector.updateScores">
<code class="descname">updateScores</code><span class="sig-paren">(</span><em>entry</em>, <em>newGeneScores</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#KBweightedSelector.updateScores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.KBweightedSelector.updateScores" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates a score entry with the new score retrieved from the knowledge base.
Used by apply function.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>entry</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code>) – a gene score entry consisting of the gene name and its score</li>
<li><strong>newGeneScores</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – dataframe containing gene scores retrieved from the knowledge base.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">updated series element.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.KBweightedSelector.getName">
<code class="descname">getName</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#KBweightedSelector.getName"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.KBweightedSelector.getName" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the selector name (including the knowledge base and (traditional) selector).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">selector name.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.KBweightedSelector.computeStatisticalRankings">
<code class="descname">computeStatisticalRankings</code><span class="sig-paren">(</span><em>intermediateDir</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#KBweightedSelector.computeStatisticalRankings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.KBweightedSelector.computeStatisticalRankings" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the statistical relevance score of all features using the (traditional) selector.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>intermediateDir</strong> (<em>str</em>) – absolute path to output directory for (traditional) selector (where to write the statistical rankings).</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">dataframe with statistical ranking.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.KBweightedSelector.computeExternalRankings">
<code class="descname">computeExternalRankings</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#KBweightedSelector.computeExternalRankings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.KBweightedSelector.computeExternalRankings" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the association scores for every gene using the knowledge base.
Genes for which no entry could be found receive a default score of 0.000001.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">dataframe with statistical ranking.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.KBweightedSelector.combineRankings">
<code class="descname">combineRankings</code><span class="sig-paren">(</span><em>externalRankings</em>, <em>statisticalRankings</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#KBweightedSelector.combineRankings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.KBweightedSelector.combineRankings" title="Permalink to this definition">¶</a></dt>
<dd><p>Combines score rankings from both the knowledge base and the (traditional) selector (kb_score * trad_score) to retrieve a final score for every gene.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>externalRankings</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – dataframe with ranking from knowledge base.</li>
<li><strong>statisticalRankings</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – dataframe with statistical ranking.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">dataframe with final combined ranking.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.KBweightedSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#KBweightedSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.KBweightedSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs the feature selection process.
Retrieves scores from knowledge base and (traditional) selector and combines these to a single score.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to final output file containing the ranking.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.LassoPenalty">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">LassoPenalty</code><span class="sig-paren">(</span><em>knowledgebase</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#LassoPenalty"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.LassoPenalty" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.PriorKnowledgeSelector" title="featureselection.PriorKnowledgeSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.PriorKnowledgeSelector</span></code></a>, <a class="reference internal" href="#featureselection.RSelector" title="featureselection.RSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.RSelector</span></code></a></p>
<p>Runs feature selection by invoking xtune R package: <a class="reference external" href="https://cran.r-project.org/web/packages/xtune/index.html">https://cran.r-project.org/web/packages/xtune/index.html</a></p>
<p>xtune is a Lasso selector that uses feature-individual penalty scores.
These penalty scores are retrieved from the knowledge base.</p>
<dl class="method">
<dt id="featureselection.LassoPenalty.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#featureselection.LassoPenalty.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Triggers the feature selection.
Actually a wrapper method that invokes external R code.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the result file containing the ranking.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.LassoPenalty.getName">
<code class="descname">getName</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#featureselection.LassoPenalty.getName" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the full name (including applied knowledge base) of this selector.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">selector name.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.LassoPenalty.createParams">
<code class="descname">createParams</code><span class="sig-paren">(</span><em>outputFile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#LassoPenalty.createParams"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.LassoPenalty.createParams" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the parameters the xtune R script requires (input file, output file, filename containing rankings from knowledge base).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of parameters to use for xtune execution in R.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.LassoPenalty.computeExternalRankings">
<code class="descname">computeExternalRankings</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#LassoPenalty.computeExternalRankings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.LassoPenalty.computeExternalRankings" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the association scores for each feature based on the scores retrieved from the knowledge base.
Features that could not be found in the knowledge base receive a default score of 0.000001.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to the file containing the external rankings.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.WrapperSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">WrapperSelector</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#WrapperSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.WrapperSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.PythonSelector" title="featureselection.PythonSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.PythonSelector</span></code></a></p>
<p>Selector implementation for wrapper selectors using scikit-learn.
Currently implements recursive feature eliminatin (RFE) and sequential forward selection (SFS) strategies,
which can be combined with nearly any classifier offered by scikit-learn, e.g. SVM.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>selector</strong> – scikit-learn selector strategy (currently RFE and SFS)</li>
<li><strong>classifier</strong> – scikit-learn classifier to use for wrapper selection.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="featureselection.WrapperSelector.createClassifier">
<code class="descname">createClassifier</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#WrapperSelector.createClassifier"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.WrapperSelector.createClassifier" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a classifier instance (from scikit-learn) to be used during the selection process.
To enable the framework to use a new classifier, extend this method accordingly.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">scikit-learn classifier instance.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.WrapperSelector.createSelector">
<code class="descname">createSelector</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#WrapperSelector.createSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.WrapperSelector.createSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a selector instance that leads the selection process.
Currently, sequential forward selection (SFS) and recursive feature elimination (RFE) are implemented.
Extend this method if you want to add another selection strategy.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">scikit-learn selector instance.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.WrapperSelector.prepareOutput">
<code class="descname">prepareOutput</code><span class="sig-paren">(</span><em>outputFile</em>, <em>data</em>, <em>selector</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#WrapperSelector.prepareOutput"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.WrapperSelector.prepareOutput" title="Permalink to this definition">¶</a></dt>
<dd><p>Overwrites the inherited prepareOutput method because we need to access the particular selector’s coefficients.
The coefficients are extracted as feature scores and will be written to the rankings file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>outputFile</strong> (<em>str</em>) – selector name</li>
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – input dataset to get the feature names.</li>
<li><strong>selector</strong> – selector instance that is used during feature selection.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.WrapperSelector.runSelector">
<code class="descname">runSelector</code><span class="sig-paren">(</span><em>data</em>, <em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#WrapperSelector.runSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.WrapperSelector.runSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs the actual feature selector of scikit-learn.
Is invoked by <a class="reference internal" href="#featureselection.PythonSelector.selectFeatures" title="featureselection.PythonSelector.selectFeatures"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PythonSelector.selectFeatures()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – dataframe containing the unlabeled dataset.</li>
<li><strong>labels</strong> (<em>list of int</em>) – numerically encoded class labels.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">sklearn/mlxtend selector that ran the selection (containing coefficients etc.).</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.SVMRFESelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">SVMRFESelector</code><a class="reference internal" href="_modules/featureselection.html#SVMRFESelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.SVMRFESelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.JavaSelector" title="featureselection.JavaSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.JavaSelector</span></code></a></p>
<p>Executes SVM-RFE with poly-kernel.
Uses an efficient java implementation from WEKA and is thus just a wrapper class to invoke the corresponding jars.</p>
<dl class="method">
<dt id="featureselection.SVMRFESelector.createParams">
<code class="descname">createParams</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#SVMRFESelector.createParams"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.SVMRFESelector.createParams" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the parameters the java program requires (input file, output file, selector name).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of parameters to use for InfoGain execution in java.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.RandomForestSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">RandomForestSelector</code><a class="reference internal" href="_modules/featureselection.html#RandomForestSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.RandomForestSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.PythonSelector" title="featureselection.PythonSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.PythonSelector</span></code></a></p>
<p>Selector class that implements RandomForest as provided by scikit-learn.</p>
<dl class="method">
<dt id="featureselection.RandomForestSelector.prepareOutput">
<code class="descname">prepareOutput</code><span class="sig-paren">(</span><em>outputFile</em>, <em>data</em>, <em>selector</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#RandomForestSelector.prepareOutput"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.RandomForestSelector.prepareOutput" title="Permalink to this definition">¶</a></dt>
<dd><p>Overwrites the inherited prepareOutput method because we need to access the RandomForest selector’s feature importances.
These feature importances are extracted as feature scores and will be written to the rankings file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>outputFile</strong> (<em>str</em>) – selector name</li>
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – input dataset to get the feature names.</li>
<li><strong>selector</strong> – RandomForest selector instance that is used during feature selection.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.RandomForestSelector.runSelector">
<code class="descname">runSelector</code><span class="sig-paren">(</span><em>data</em>, <em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#RandomForestSelector.runSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.RandomForestSelector.runSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs the actual feature selection using scikit-learn’s RandomForest classifier.
Is invoked by <a class="reference internal" href="#featureselection.PythonSelector.selectFeatures" title="featureselection.PythonSelector.selectFeatures"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PythonSelector.selectFeatures()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – dataframe containing the unlabeled dataset.</li>
<li><strong>labels</strong> (<em>list of int</em>) – numerically encoded class labels.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">scikit-learn RandomForestClassifier that ran the selection.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.LassoSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">LassoSelector</code><a class="reference internal" href="_modules/featureselection.html#LassoSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.LassoSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.PythonSelector" title="featureselection.PythonSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.PythonSelector</span></code></a></p>
<p>Selector class that implements Lasso feature selection using scikit-learn.</p>
<dl class="method">
<dt id="featureselection.LassoSelector.prepareOutput">
<code class="descname">prepareOutput</code><span class="sig-paren">(</span><em>outputFile</em>, <em>data</em>, <em>selector</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#LassoSelector.prepareOutput"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.LassoSelector.prepareOutput" title="Permalink to this definition">¶</a></dt>
<dd><p>Overwrites the inherited prepareOutput method because we need to access Lasso’s coefficients.
These coefficients are extracted as feature scores and will be written to the rankings file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>outputFile</strong> (<em>str</em>) – selector name</li>
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – input dataset to get the feature names.</li>
<li><strong>selector</strong> – RandomForest selector instance that is used during feature selection.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.LassoSelector.runSelector">
<code class="descname">runSelector</code><span class="sig-paren">(</span><em>data</em>, <em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#LassoSelector.runSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.LassoSelector.runSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs the actual Lasso feature selector using scikit-learn.
Is invoked by <a class="reference internal" href="#featureselection.PythonSelector.selectFeatures" title="featureselection.PythonSelector.selectFeatures"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PythonSelector.selectFeatures()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – dataframe containing the unlabeled dataset.</li>
<li><strong>labels</strong> (<em>list of int</em>) – numerically encoded class labels.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">Lasso selector that ran the selection.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.PreFilterSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">PreFilterSelector</code><span class="sig-paren">(</span><em>knowledgebase</em>, <em>tradApproach</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PreFilterSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PreFilterSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.CombiningSelector" title="featureselection.CombiningSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.CombiningSelector</span></code></a></p>
<p>Applies a two-level prefiltering strategy for feature selection.
Filters all features that were not retrieved by a knowledge base based on the search terms provided in the config file.
Applies a (traditional) feature selector on the remaining features afterwards.</p>
<p>For traditional univariate filter approaches, the results retrieved by this class and <a class="reference internal" href="#featureselection.PostFilterSelector" title="featureselection.PostFilterSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">PostFilterSelector</span></code></a> will be the same.</p>
<dl class="method">
<dt id="featureselection.PreFilterSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PreFilterSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PreFilterSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Carries out feature selection.
First queries the assigned knowledge base to get genes that are associated to the given search terms.
Filter feature set of input data set to contain only features that are in the retrieved gene set.
Apply (traditional) selector on the filtered data set.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to rankings file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.PostFilterSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">PostFilterSelector</code><span class="sig-paren">(</span><em>knowledgebase</em>, <em>tradApproach</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PostFilterSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PostFilterSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.CombiningSelector" title="featureselection.CombiningSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.CombiningSelector</span></code></a></p>
<p>Applies a two-level postfiltering strategy for feature selection.
Applies (traditional) feature selection to the input data set.
Afterwards, removes all genes for which no information in the corresponding knowledge base was found based on the search terms provided in the config file.
For traditional univariate filter approaches, the results retrieved by this class and <a class="reference internal" href="#featureselection.PreFilterSelector" title="featureselection.PreFilterSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">PreFilterSelector</span></code></a> will be the same.</p>
<dl class="method">
<dt id="featureselection.PostFilterSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PostFilterSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PostFilterSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Carries out feature selection.
First executes (traditional) selector.
Then queries the assigned knowledge base to get genes that are associated to the given search terms.
Finally filters feature set to contain only features that are in the retrieved gene set.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to rankings file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.ExtensionSelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">ExtensionSelector</code><span class="sig-paren">(</span><em>knowledgebase</em>, <em>tradApproach</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#ExtensionSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.ExtensionSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.CombiningSelector" title="featureselection.CombiningSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.CombiningSelector</span></code></a></p>
<p>Selector implementation inspired by SOFOCLES:
“SoFoCles: Feature filtering for microarray classification based on Gene Ontology”, Papachristoudis et al., Journal of Biomedical Informatics, 2010</p>
<p>This selector carries out (traditional) feature selection and in parallel retrieves relevant genes from a knowledge base based on the provided search terms.
The ranking is then adapted by alternating the feature ranking retrieved by the (traditiona) selection approach and the externally retrieved genes.
This is kind of related to an extension approach, where a feature ranking that was retrieved by a traditional approach is extended by such external genes.</p>
<dl class="method">
<dt id="featureselection.ExtensionSelector.selectFeatures">
<code class="descname">selectFeatures</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#ExtensionSelector.selectFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.ExtensionSelector.selectFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Carries out feature selection.
Executes (traditional) selector and separately retrieves genes from the assigned knowledge base based on the search terms specified in the config.
Finally merges the two feature lists alternating to form an “extended” feature ranking.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">absolute path to rankings file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.NetworkActivitySelector">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">NetworkActivitySelector</code><span class="sig-paren">(</span><em>knowledgebase</em>, <em>featuremapper</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#NetworkActivitySelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.NetworkActivitySelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.NetworkSelector" title="featureselection.NetworkSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.NetworkSelector</span></code></a></p>
<p>Selector implementation that selects a set of pathways from the knowledge base and maps the feature space to the pathways.
Pathway ranking scores are computed based on the average ANOVA p-value of its member genes and the sample classes.
This method is also used by Chuang et al. and Tian et al. (Discovering statistically significant pathways in expression profiling studies)
Pathway feature values are computed with an instance of <a class="reference internal" href="#featureselection.FeatureMapper" title="featureselection.FeatureMapper"><code class="xref py py-class docutils literal notranslate"><span class="pre">FeatureMapper</span></code></a> or inheriting classes, whose mapping strategies can vary.
If pathways should be selected according to another strategy, use this class as an example implementation to implement a new class that inherits from <a class="reference internal" href="#featureselection.NetworkSelector" title="featureselection.NetworkSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">NetworkSelector</span></code></a>.</p>
<dl class="method">
<dt id="featureselection.NetworkActivitySelector.selectPathways">
<code class="descname">selectPathways</code><span class="sig-paren">(</span><em>pathways</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#NetworkActivitySelector.selectPathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.NetworkActivitySelector.selectPathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes a pathway ranking for the input pathways.
Computes a pathway score based on the average ANOVA’s f-test p-values of a pathway’s member genes and the sample classes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pathways</strong> (<em>str</em>) – selector name</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">pathway ranking with pathway scores</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.FeatureMapper">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">FeatureMapper</code><a class="reference internal" href="_modules/featureselection.html#FeatureMapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureMapper" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Abstract.
Inherit from this class and implement <a class="reference internal" href="#featureselection.FeatureMapper.mapFeatures" title="featureselection.FeatureMapper.mapFeatures"><code class="xref py py-meth docutils literal notranslate"><span class="pre">FeatureMapper.mapFeatures()</span></code></a> to implement a new mapping strategy.
Maps the feature space of the given input data to a given set of pathways.
Computes a new feature value for every feature and sample based on the implemented strategy.</p>
<dl class="method">
<dt id="featureselection.FeatureMapper.mapFeatures">
<code class="descname">mapFeatures</code><span class="sig-paren">(</span><em>original_data</em>, <em>pathways</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureMapper.mapFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureMapper.mapFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract method.
Implement this method when inheriting from this class.
Carries out the actual feature mapping.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>original_data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – the original data set of which to map the feature space.</li>
<li><strong>pathways</strong> (<em>dict</em>) – dict of pathway names as keys and corresponding pathway <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code> objects as values</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the transformed data set with new feature values</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureMapper.getUnlabeledData">
<code class="descname">getUnlabeledData</code><span class="sig-paren">(</span><em>dataset</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureMapper.getUnlabeledData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureMapper.getUnlabeledData" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes the labels from the data set.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>dataset</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – data set from which to remove the labels.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">data set without labels.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureMapper.getLabels">
<code class="descname">getLabels</code><span class="sig-paren">(</span><em>dataset</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureMapper.getLabels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureMapper.getLabels" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the dataset labels.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>dataset</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – data set from which to extract the labels.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">label vector of the data set.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureMapper.getFeatures">
<code class="descname">getFeatures</code><span class="sig-paren">(</span><em>dataset</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureMapper.getFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureMapper.getFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the features of a data set.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>dataset</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – data set from which to extract the features.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">feature vector of the data set.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureMapper.getSamples">
<code class="descname">getSamples</code><span class="sig-paren">(</span><em>dataset</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureMapper.getSamples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureMapper.getSamples" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets all samples in a data set.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>dataset</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – data set from which to extract the samples.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of samples from the data set.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.FeatureMapper.getPathwayGenes">
<code class="descname">getPathwayGenes</code><span class="sig-paren">(</span><em>pathway</em>, <em>genes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#FeatureMapper.getPathwayGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.FeatureMapper.getPathwayGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the intersection of a given set of genes and the genes contained in a given pathway.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pathway</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code>) – pathway object from which to get the genes.</li>
<li><strong>genes</strong> (<em>list of str</em>) – list of gene names.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">list of genes that are contained in both the pathway and the gene list.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list of str</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.CORGSActivityMapper">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">CORGSActivityMapper</code><a class="reference internal" href="_modules/featureselection.html#CORGSActivityMapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.CORGSActivityMapper" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.FeatureMapper" title="featureselection.FeatureMapper"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.FeatureMapper</span></code></a></p>
<p>Pathway mapper that implements the strategy described by Lee et al.: “Inferring Pathway Activity toward Precise Disease Classification”
Identifies CORGS genes for every pathway: uses random search to find the minimal set of genes for which the pathway activity score is maximal.
First, every sample receives an activity score, which is the average expression level of the (CORGS) genes / number of genes.
The computed activity scores are then used for f-testing with the class labels, and the p-values are the new pathway feature values.
These steps are executed again and again until the p-values are not decreasing anymore.</p>
<dl class="method">
<dt id="featureselection.CORGSActivityMapper.getANOVAscores">
<code class="descname">getANOVAscores</code><span class="sig-paren">(</span><em>data</em>, <em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#CORGSActivityMapper.getANOVAscores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.CORGSActivityMapper.getANOVAscores" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies ANOVA f-test to test the association/correlation of a feature (pathway) with a given label.
The feature has activity scores (computed from CORGS genes) for every sample, which are to be tested for the labels.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – the data set which to test for correlation with the labels (typically feature scores of a pathway for samples).</li>
<li><strong>labels</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code>) – class labels to use for f-test.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">series of p-values for every sample.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.CORGSActivityMapper.computeActivityScore">
<code class="descname">computeActivityScore</code><span class="sig-paren">(</span><em>sampleExpressionLevels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#CORGSActivityMapper.computeActivityScore"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.CORGSActivityMapper.computeActivityScore" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the activity score of a given set of genes for a specific sample.
The activity score of a sample is the mean expression value of the given genes divided by the overall number of given genes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sampleExpressionLevels</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – data set containing expression levels from a given set of genes for samples.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">activity scores for the given samples.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.CORGSActivityMapper.computeActivityVector">
<code class="descname">computeActivityVector</code><span class="sig-paren">(</span><em>expressionLevels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#CORGSActivityMapper.computeActivityVector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.CORGSActivityMapper.computeActivityVector" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the activity score of a given set of genes for a all samples.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>expressionLevels</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – input data set of expression levels for a given set of (CORGS) genes.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">instance of a feature selector implementation.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> or inheriting class</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.CORGSActivityMapper.mapFeatures">
<code class="descname">mapFeatures</code><span class="sig-paren">(</span><em>original_data</em>, <em>pathways</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#CORGSActivityMapper.mapFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.CORGSActivityMapper.mapFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Carries out the actual feature mapping.
Follows the strategy described by Lee et al.: “Inferring Pathway Activity toward Precise Disease Classification”
Identifies CORGS genes for every pathway: uses random search to find the minimal set of genes for which the pathway activity score is maximal.
First, every sample receives an activity score, which is the average expression level of the (CORGS) genes / number of genes.
The computed activity scores are then used for f-testing with the class labels, and the p-values are the new pathway feature values.
These steps are executed again and again until the p-values are not decreasing anymore.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>original_data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – the original data set of which to map the feature space.</li>
<li><strong>pathways</strong> (<em>dict</em>) – dict of pathway names as keys and corresponding pathway <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code> objects as values</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the transformed data set with new feature values</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="featureselection.PathwayActivityMapper">
<em class="property">class </em><code class="descclassname">featureselection.</code><code class="descname">PathwayActivityMapper</code><a class="reference internal" href="_modules/featureselection.html#PathwayActivityMapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PathwayActivityMapper" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#featureselection.FeatureMapper" title="featureselection.FeatureMapper"><code class="xref py py-class docutils literal notranslate"><span class="pre">featureselection.FeatureMapper</span></code></a></p>
<p>Pathway mapper that implements a strategy that is related to Vert and Kanehisa’s strategy: Vert, Jean-Philippe, and Minoru Kanehisa. “Graph-driven feature extraction from microarray data using diffusion kernels and kernel CCA.” NIPS. 2002.
Computes pathway activity scores for every sample and pathway as new feature values.
The feature value is the average of: expression level weighted by gene variance and neighbor correlation score)</p>
<dl class="method">
<dt id="featureselection.PathwayActivityMapper.getAverageCorrelation">
<code class="descname">getAverageCorrelation</code><span class="sig-paren">(</span><em>correlations</em>, <em>gene</em>, <em>neighbors</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PathwayActivityMapper.getAverageCorrelation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PathwayActivityMapper.getAverageCorrelation" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the average correlation from the correlations of a given gene and its neighbors.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>correlations</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – correlation matrix of all genes.</li>
<li><strong>gene</strong> (<em>str</em>) – gene name whose average neighbor correlation to compute.</li>
<li><strong>neighbors</strong> (<em>list of str</em>) – list of gene names that are neighbors of the given gene.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">average correlation value.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">float</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.PathwayActivityMapper.computeGeneVariances">
<code class="descname">computeGeneVariances</code><span class="sig-paren">(</span><em>data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PathwayActivityMapper.computeGeneVariances"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PathwayActivityMapper.computeGeneVariances" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the variances for every gene across all samples.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – data set with expression values.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">variance for every gene.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="featureselection.PathwayActivityMapper.mapFeatures">
<code class="descname">mapFeatures</code><span class="sig-paren">(</span><em>original_data</em>, <em>pathways</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/featureselection.html#PathwayActivityMapper.mapFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#featureselection.PathwayActivityMapper.mapFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Executes the actual feature mapping procedure.
A feature value  is the average of (for every gene in a pathway): (expression level weighted by gene variance and neighbor correlation score)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>original_data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – the original data set of which to map the feature space.</li>
<li><strong>pathways</strong> (<em>dict</em>) – dict of pathway names as keys and corresponding pathway <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code> objects as values</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the transformed data set with new feature values</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="knowledgebases-module">
<h2>knowledgebases module<a class="headerlink" href="#knowledgebases-module" title="Permalink to this headline">¶</a></h2>
<p>Contains all classes related to knowledge bases.
A knowledge base is realized with two classes:
* A class inheriting from <a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a> and implementing the three interface methods <code class="xref py py-meth docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase:getRelevantGenes()</span></code>, <code class="xref py py-meth docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase:getGeneScores()</span></code>, and <code class="xref py py-meth docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase:getRelevantPathways()</span></code>.
* A class that is responsible for querying the corresponding web service and inherits from Bioservice’s REST class.
Those knowledge bases that retrieve pathway information also need an additional PathwayMapper class, which transforms the original pathway results from the knowledge base (which can range from SIF to any other pathway specification format) into the pathway representation that is used throughout Comprior.
For Comprior’s internal pathway representation, we use pypath.</p>
<p>The creation of knowledge bases is encapsulated by the class <code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebase.KnowledgeBaseFactory</span></code> that takes care that every knowledge base is equipped with a web service querying class and, if needed, the right type of <code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebase.PathwayMapper</span></code>.
For a detailed look at the class architecture, have a look at ADD CLASS ARCHITECTURE LINK HERE.</p>
<span class="target" id="module-knowledgebases"></span><dl class="function">
<dt id="knowledgebases.suppress_stdout">
<code class="descclassname">knowledgebases.</code><code class="descname">suppress_stdout</code><span class="sig-paren">(</span><em>suppress=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#suppress_stdout"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.suppress_stdout" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="knowledgebases.ENRICHR">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">ENRICHR</code><a class="reference internal" href="_modules/knowledgebases.html#ENRICHR"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.ENRICHR" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">bioservices.services.REST</span></code></p>
<p>Queries some of the API endpoints of the EnrichR web service (<a class="reference external" href="https://maayanlab.cloud/Enrichr/help#api">https://maayanlab.cloud/Enrichr/help#api</a>).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>config</strong> (<em>dict</em>) – configuration parameters for EnrichR web service (as specified in config file).</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="knowledgebases.ENRICHR.addlist">
<code class="descname">addlist</code><span class="sig-paren">(</span><em>geneList</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#ENRICHR.addlist"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.ENRICHR.addlist" title="Permalink to this definition">¶</a></dt>
<dd><p>Queries EnrichR to annotate a given list of genes.
Returns a userListID, which can be used to retrieve the actual results in a second query.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>geneList</strong> – list of genes to annotate</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">json response containing a userListID.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dict of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.ENRICHR.export">
<code class="descname">export</code><span class="sig-paren">(</span><em>params</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#ENRICHR.export"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.ENRICHR.export" title="Permalink to this definition">¶</a></dt>
<dd><p>Download file of enrichment results.
Requires a userListId that was retrieved from a prior query.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>params</strong> (<em>list of str</em>) – list of parameters to use for that query (userListId: Identifier returned from addList endpoint, filename: Name of text file download, backgroundType: Gene set library for which to download results)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">text file containing enrichment results.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.ENRICHR.genemap">
<code class="descname">genemap</code><span class="sig-paren">(</span><em>params</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#ENRICHR.genemap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.ENRICHR.genemap" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds all terms, their descriptions, and optional categorizations, for a given gene identifier.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>params</strong> (<em>list of str</em>) – list of parameters to be used for the query (gene     Gene to use in search for terms, json (optional): Set “true” to return JSON rather plaintext, setup (optional): Set “true” to category information for the libraries)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">json object of all terms containing the specified gene and their descriptions.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dict of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.ENRICHR.enrich">
<code class="descname">enrich</code><span class="sig-paren">(</span><em>params</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#ENRICHR.enrich"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.ENRICHR.enrich" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all that are terms available in library (specified by backgroundType param) and enriched in the given set of genes (specified by userListId param).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>params</strong> (<em>list of str</em>) – list of parameters to be used for the query (userListId: Identifier returned from addList endpoint; backgroundType: Gene set library to enrich against)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">dataframe object of all enriched terms (unsorted, unfiltered.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dataframe</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.UMLS_AUTH">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">UMLS_AUTH</code><a class="reference internal" href="_modules/knowledgebases.html#UMLS_AUTH"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.UMLS_AUTH" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">bioservices.services.REST</span></code></p>
<p>Singleton class.
Python code encapsulates it in a way that is not shown in Sphinx, so have a look at the descriptions in the source code.</p>
<p>Authentication service to get access to the UMLS database UMLS database (which we need for retrieving CUI disease codes for querying DisGeNET).
You first have to get a ticket-granting ticket (tgt, valid for 8 hours) with the help of an API key.
With the tgt, you can then request a service ticket for every new query to the UMLS database.
The service ticket must then be used for the query.
The task of this class is to generate a valid tgt and subsequent service ticket.
Documentation on the authentication process: <a class="reference external" href="https://documentation.uts.nlm.nih.gov/rest/authentication.html">https://documentation.uts.nlm.nih.gov/rest/authentication.html</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>config</strong> (<em>dict</em>) – configuration parameters for UMLS web service as specified in config file.</li>
<li><strong>tgt_timestamp</strong> (<em>str</em>) – timestamp of the tgt. If it is older than 8 hours, we need to request a new tgt.</li>
<li><strong>tgt</strong> (<em>list of str</em>) – id of the ticket-granting ticket (valid for 8 hours). With this ticket, we can then query the actual UMLS web service.</li>
<li><strong>service</strong> (<em>str</em>) – uri for the service login</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="knowledgebases.UMLS_AUTH.instance">
<code class="descname">instance</code><em class="property"> = None</em><a class="headerlink" href="#knowledgebases.UMLS_AUTH.instance" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.UMLS">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">UMLS</code><a class="reference internal" href="_modules/knowledgebases.html#UMLS"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.UMLS" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">bioservices.services.REST</span></code></p>
<p>Retrieves UMLS CUI codes for labels, which can then be used for querying DisGeNET.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>config</strong> (<em>dict</em>) – configuration parameters for UMLS web service (as specified in configuration file).</li>
<li><strong>auth</strong> (<a class="reference internal" href="#knowledgebases.UMLS_AUTH" title="knowledgebases.UMLS_AUTH"><code class="xref py py-class docutils literal notranslate"><span class="pre">UMLS_AUTH</span></code></a>) – authentication component to generate a valid service ticket (required for every query).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="knowledgebases.UMLS.getCUIs">
<code class="descname">getCUIs</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#UMLS.getCUIs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.UMLS.getCUIs" title="Permalink to this definition">¶</a></dt>
<dd><p>Get CUIs for the given labels.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of identifiers for which to retrieve CUIs, e.g. disease names.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of CUIs.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.DISGENET">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">DISGENET</code><a class="reference internal" href="_modules/knowledgebases.html#DISGENET"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.DISGENET" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">bioservices.services.REST</span></code></p>
<p>Queries the DisGeNET web service for a given set of labels and retrieves association scores for all genes related to the query labels.
DisGeNET API documentation: <a class="reference external" href="https://www.disgenet.org/api/">https://www.disgenet.org/api/</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>umls</strong> (<a class="reference internal" href="#knowledgebases.UMLS" title="knowledgebases.UMLS"><code class="xref py py-class docutils literal notranslate"><span class="pre">UMLS</span></code></a> for transforming disease names to CUIs (required for query)) – list of gene names to be mapped</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="knowledgebases.DISGENET.getVersion">
<code class="descname">getVersion</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#DISGENET.getVersion"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.DISGENET.getVersion" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the current version of the DisGeNET API endpoint.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">web service version infos.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">json dict</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.DISGENET.query">
<code class="descname">query</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#DISGENET.query"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.DISGENET.query" title="Permalink to this definition">¶</a></dt>
<dd><p>Conducts the actual query to retrive gene-disease association scores for a given list of disease labels.
Transforms the disease labels into CUIs before with the UMLS web service.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of disease labels for which to retrieve gene-disease associations.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">DataFrame with gene-disease association scores.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.GCONVERT">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">GCONVERT</code><a class="reference internal" href="_modules/knowledgebases.html#GCONVERT"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.GCONVERT" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Queries the g:Convert web service to map a list of identifiers to a desired format.
g:Convert makes use of the Ensembl build.
g:Convert API documentation: https://biit.cs.ut.ee/gprofiler/page/apis</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>url</strong> (<em>str</em>) – API url as specified in the configuration file.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="knowledgebases.GCONVERT.query">
<code class="descname">query</code><span class="sig-paren">(</span><em>items</em>, <em>originalFormat</em>, <em>desiredFormat</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#GCONVERT.query"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.GCONVERT.query" title="Permalink to this definition">¶</a></dt>
<dd><p>Map a list of itendifiers to the desired format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>items</strong> (<em>list of str</em>) – list of identifiers to be mapped</li>
<li><strong>originalFormat</strong> (<em>str</em>) – current format of the identifiers</li>
<li><strong>desiredFormat</strong> (<em>str</em>) – desired identifier format</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">DataFrame containing the identifier mapping.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.PATHWAYCOMMONSWS">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">PATHWAYCOMMONSWS</code><a class="reference internal" href="_modules/knowledgebases.html#PATHWAYCOMMONSWS"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.PATHWAYCOMMONSWS" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">bioservices.services.REST</span></code></p>
<p>Queries the PathwayCommons web service.
Bioservices’ existing implementation to query PathwayCommons was not used because it contained outdated values for _valid_formats for pathway retrieval,
so we used the original code and adapted it to work correctly.</p>
<dl class="method">
<dt id="knowledgebases.PATHWAYCOMMONSWS.getVersion">
<code class="descname">getVersion</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#PATHWAYCOMMONSWS.getVersion"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.PATHWAYCOMMONSWS.getVersion" title="Permalink to this definition">¶</a></dt>
<dd><p>Map a list of genes to the desired format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>genes</strong> (<em>list of str</em>) – list of gene names to be mapped</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of mapped gene names.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="knowledgebases.PATHWAYCOMMONSWS.default_extension">
<code class="descname">default_extension</code><a class="headerlink" href="#knowledgebases.PATHWAYCOMMONSWS.default_extension" title="Permalink to this definition">¶</a></dt>
<dd><p>set extension of the requests (default is json). Can be ‘json’ or ‘xml’</p>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.PATHWAYCOMMONSWS.search">
<code class="descname">search</code><span class="sig-paren">(</span><em>q</em>, <em>page=0</em>, <em>datasource=None</em>, <em>organism=None</em>, <em>type=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#PATHWAYCOMMONSWS.search"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.PATHWAYCOMMONSWS.search" title="Permalink to this definition">¶</a></dt>
<dd><p>Text search in PathwayCommons using Lucene query syntax</p>
<p>Some of the parameters are BioPAX properties, others are composite
relationships.</p>
<p>All index fields are (case-sensitive): comment, ecnumber,
keyword, name, pathway, term, xrefdb, xrefid, dataSource, and organism.</p>
<p>The pathway field maps to all participants of pathways that contain
the keyword(s) in any of its text fields.</p>
<p>Finally, keyword is a transitive aggregate field that includes all
searchable keywords of that element and its child elements.</p>
<p>All searches can also be filtered by data source and organism.</p>
<p>It is also possible to restrict the domain class using the
‘type’ parameter.</p>
<p>This query can be used standalone or to retrieve starting points
for graph searches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>q</strong> (<em>str</em>) – requires a keyword , name, external identifier, or a
Lucene query string.</li>
<li><strong>page</strong> (<em>int</em>) – (N&gt;=0, default is 0), search result page number.</li>
<li><strong>datasource</strong> (<em>str</em>) – filter by data source (use names or URIs of
pathway data sources or of any existing Provenance object). If
multiple data source values are specified, a union of hits from
specified sources is returned. datasource=[reactome,pid] returns
hits associated with Reactome or PID.</li>
<li><strong>organism</strong> (<em>str</em>) – The organism can be specified either by
official name, e.g. “homo sapiens” or by NCBI taxonomy id,
e.g. “9606”. Similar to data sources, if multiple organisms
are declared a union of all hits from specified organisms
is returned. For example organism=[9606, 10016] returns results
for both human and mice.</li>
<li><strong>type</strong> (<em>str</em>) – BioPAX class filter</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.PATHWAYCOMMONSWS.get">
<code class="descname">get</code><span class="sig-paren">(</span><em>uri</em>, <em>frmt='BIOPAX'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#PATHWAYCOMMONSWS.get"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.PATHWAYCOMMONSWS.get" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieves full pathway information for a set of elements</p>
<p>elements can be for example pathway, interaction or physical
entity given the RDF IDs. Get commands only
retrieve the BioPAX elements that are directly mapped to the ID.
Use the <code class="xref py py-meth docutils literal notranslate"><span class="pre">traverse()</span></code> query to traverse BioPAX graph and
obtain child/owner elements.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>uri</strong> (<em>str</em>) – valid/existing BioPAX element’s URI (RDF ID; for
utility classes that were “normalized”, such as entity refereneces
and controlled vocabularies, it is usually a Identifiers.org URL.
Multiple IDs can be provided using list
uri=[<a class="reference external" href="http://identifiers.org/uniprot/Q06609">http://identifiers.org/uniprot/Q06609</a>,
<a class="reference external" href="http://identifiers.org/uniprot/Q549Z0">http://identifiers.org/uniprot/Q549Z0</a>’]
See also about MIRIAM and Identifiers.org.</li>
<li><strong>format</strong> (<em>str</em>) – output format (values)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a complete BioPAX representation for the record
pointed to by the given URI is returned. Other output
formats are produced by converting the BioPAX record on
demand and can be specified by the optional format
parameter. Please be advised that with some output formats
it might return “no result found” error if the conversion is
not applicable for the BioPAX result. For example,
BINARY_SIF output usually works if there are some
interactions, complexes, or pathways in the retrieved set
and not only physical entities.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.KnowledgeBaseFactory">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">KnowledgeBaseFactory</code><a class="reference internal" href="_modules/knowledgebases.html#KnowledgeBaseFactory"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.KnowledgeBaseFactory" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Singleton class.
Python code encapsulates it in a way that is not shown in Sphinx, so have a look at the descriptions in the source code.</p>
<p>Creates knowledge bases based on the provided name and creates all corresponding objects, e.g. web service endpoints.
Every knowledge base implementation must be registered here, otherwise it will not be accessible.</p>
<dl class="attribute">
<dt id="knowledgebases.KnowledgeBaseFactory.instance">
<code class="descname">instance</code><em class="property"> = None</em><a class="headerlink" href="#knowledgebases.KnowledgeBaseFactory.instance" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.KnowledgeBase">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">KnowledgeBase</code><span class="sig-paren">(</span><em>name</em>, <em>kb_config</em>, <em>webservice</em>, <em>geneInfo</em>, <em>pathwayInfo</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#KnowledgeBase"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.KnowledgeBase" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Super class for every knowledge base implementation.
If a new knowledge base is implemented, it must inherit from this class and implement methods <a class="reference internal" href="#knowledgebases.KnowledgeBase.getRelevantGenes" title="knowledgebases.KnowledgeBase.getRelevantGenes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">KnowledgeBase.getRelevantGenes()</span></code></a>, <a class="reference internal" href="#knowledgebases.KnowledgeBase.getGeneScores" title="knowledgebases.KnowledgeBase.getGeneScores"><code class="xref py py-meth docutils literal notranslate"><span class="pre">KnowledgeBase.getGeneScores()</span></code></a>, and <a class="reference internal" href="#knowledgebases.KnowledgeBase.getRelevantPathways" title="knowledgebases.KnowledgeBase.getRelevantPathways"><code class="xref py py-meth docutils literal notranslate"><span class="pre">KnowledgeBase.getRelevantPathways()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> (<em>str</em>) – name of the knowledge base</li>
<li><strong>config</strong> (<em>dict</em>) – configuration parameter of the knowledge base as specified in the config file.</li>
<li><strong>webservice</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bioservices.REST</span></code> or inheriting classes.) – web service querying object</li>
<li><strong>hasGeneInformation</strong> (<em>bool</em>) – true if the knowledge base provides gene association information, false otherwise</li>
<li><strong>hasPathwayInformation</strong> (<em>bool</em>) – true if the knowledge base also provides pathway information, false otherwise</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="knowledgebases.KnowledgeBase.getRelevantGenes">
<code class="descname">getRelevantGenes</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#KnowledgeBase.getRelevantGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.KnowledgeBase.getRelevantGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract.
Get all genes that are associated to a list of labels, e.g. disease names.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of labels for which to retrieve the genes.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of associated genes.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.KnowledgeBase.getGeneScores">
<code class="descname">getGeneScores</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#KnowledgeBase.getGeneScores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.KnowledgeBase.getGeneScores" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract.
Get all genes and their association scores for a given list of disease names.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of disease names for which to get gene-disease-association scores.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">DataFrame of genes and their association scores.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.KnowledgeBase.getRelevantPathways">
<code class="descname">getRelevantPathways</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#KnowledgeBase.getRelevantPathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.KnowledgeBase.getRelevantPathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Get all pathways related to a set of labels, e.g disease names.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of labels for which to find related pathways.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">dict of pathway names and pathway representations.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dict with <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code> as values</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.KnowledgeBase.getName">
<code class="descname">getName</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#KnowledgeBase.getName"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.KnowledgeBase.getName" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the name of the knowledge base.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">knowledge base name.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.KnowledgeBase.hasPathways">
<code class="descname">hasPathways</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#KnowledgeBase.hasPathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.KnowledgeBase.hasPathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns if knowledge base retrieves pathway information, i.e. if <a class="reference internal" href="#knowledgebases.KnowledgeBase.getRelevantPathways" title="knowledgebases.KnowledgeBase.getRelevantPathways"><code class="xref py py-meth docutils literal notranslate"><span class="pre">KnowledgeBase.getRelevantPathways()</span></code></a> is implemented..</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">true if knowledge base provides pathway information, false otherwise.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">bool</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.KnowledgeBase.hasGenes">
<code class="descname">hasGenes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#KnowledgeBase.hasGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.KnowledgeBase.hasGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns if knowledge base retrieves gene information, i.e. if <a class="reference internal" href="#knowledgebases.KnowledgeBase.getRelevantGenes" title="knowledgebases.KnowledgeBase.getRelevantGenes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">KnowledgeBase.getRelevantGenes()</span></code></a> <a class="reference internal" href="#knowledgebases.KnowledgeBase.getGeneScores" title="knowledgebases.KnowledgeBase.getGeneScores"><code class="xref py py-meth docutils literal notranslate"><span class="pre">KnowledgeBase.getGeneScores()</span></code></a> are implemented.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">true if knowledge base provides gene information, false otherwise.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">bool</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.Enrichr">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">Enrichr</code><a class="reference internal" href="_modules/knowledgebases.html#Enrichr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Enrichr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a></p>
<p>Special knowledge base not intended to be used by feature selection approaches.
Instead, it is used for evaluation purposes to annotate and enrich rankings.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> (<em>str</em>) – name of the knowledge base</li>
<li><strong>config</strong> (<em>dict</em>) – configuration parameter of the knowledge base as specified in the config file.</li>
<li><strong>webservice</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bioservices.REST</span></code> or inheriting classes.) – web service querying object</li>
<li><strong>hasGeneInformation</strong> (<em>bool</em>) – true if the knowledge base provides gene association information, false otherwise</li>
<li><strong>hasPathwayInformation</strong> (<em>bool</em>) – true if the knowledge base also provides pathway information, false otherwise</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="knowledgebases.Enrichr.downloadEnrichedTerms">
<code class="descname">downloadEnrichedTerms</code><span class="sig-paren">(</span><em>userIdList</em>, <em>filePrefix</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Enrichr.downloadEnrichedTerms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Enrichr.downloadEnrichedTerms" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads enriched terms from a former query into a file.
Filters these terms for those with an adjusted p-value &gt; 0.05, then sorts by combined score in descending order.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>userIdList</strong> (<em>str</em>) – userIdList to retrieve enrichment/annotation results from the original query.</li>
<li><strong>filePrefix</strong> (<em>str</em>) – prefix to use in filename.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Enrichr.getRelevantGenes">
<code class="descname">getRelevantGenes</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Enrichr.getRelevantGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Enrichr.getRelevantGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Is not implemented for EnrichR.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of gene names to be mapped</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code> as this knowledge base is not intended to be used for such analyses.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Enrichr.getGeneScores">
<code class="descname">getGeneScores</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Enrichr.getGeneScores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Enrichr.getGeneScores" title="Permalink to this definition">¶</a></dt>
<dd><p>Is not implemented for EnrichR.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of gene names to be mapped</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code> as this knowledge base is not intended to be used for such analyses.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Enrichr.getRelevantPathways">
<code class="descname">getRelevantPathways</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Enrichr.getRelevantPathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Enrichr.getRelevantPathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Is not implemented for EnrichR.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of labels for which to find related pathways.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code> as this knowledge base is not intended to be used for such analyses.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Enrichr.enrichGeneset">
<code class="descname">enrichGeneset</code><span class="sig-paren">(</span><em>geneList</em>, <em>filePrefix</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Enrichr.enrichGeneset"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Enrichr.enrichGeneset" title="Permalink to this definition">¶</a></dt>
<dd><p>Sends a list of identifies (here, genes) to EnrichR web service and stores all term enrichments in a file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>geneList</strong> (<em>list of str</em>) – list of gene names for which to retrieve enrichments.</li>
<li><strong>filePrefix</strong> (<em>str</em>) – prefix to use in file name (to store enrichments).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Enrichr.annotateGene">
<code class="descname">annotateGene</code><span class="sig-paren">(</span><em>gene</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Enrichr.annotateGene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Enrichr.annotateGene" title="Permalink to this definition">¶</a></dt>
<dd><p>Annotates a gene with terms.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>gene</strong> (<em>str</em>) – gene name.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of all annotations to the provided gene.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Enrichr.annotateGenes">
<code class="descname">annotateGenes</code><span class="sig-paren">(</span><em>geneList</em>, <em>filePrefix</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Enrichr.annotateGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Enrichr.annotateGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Annotates a list of genes with relevant terms.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>geneList</strong> (<em>list of str</em>) – list of gene names to annotate.</li>
<li><strong>filePrefix</strong> (<em>str</em>) – prefix to use when storing results in a file.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">dict of gene names and lists of their annotations.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dict</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.BioMART">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">BioMART</code><a class="reference internal" href="_modules/knowledgebases.html#BioMART"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.BioMART" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Maps a identifiers or data sets with identifiers to the desired format by using BiomaRt.
Wrapper class that internally invokes BiomaRt’s R code.
Very unstable, so currently not used.
However, it can be exchanged in <a class="reference internal" href="#benchutils.retrieveMappings" title="benchutils.retrieveMappings"><code class="xref py py-meth docutils literal notranslate"><span class="pre">benchutils.retrieveMappings()</span></code></a> function.</p>
<dl class="method">
<dt id="knowledgebases.BioMART.mapItems">
<code class="descname">mapItems</code><span class="sig-paren">(</span><em>itemList</em>, <em>originalFormat</em>, <em>desiredFormat</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#BioMART.mapItems"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.BioMART.mapItems" title="Permalink to this definition">¶</a></dt>
<dd><p>Map a list of identifiers to the desired format.
Internally invokes external R code that uses the BiomaRt package.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>itemList</strong> (<em>list of str</em>) – list of identifiers to be mapped</li>
<li><strong>originalFormat</strong> (<em>str</em>) – original identifier format.</li>
<li><strong>desiredFormat</strong> (<em>str</em>) – format to which to map identifiers.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">mapping data frame of identifiers (with original and desired format)</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.BioMART.getRelevantGenes">
<code class="descname">getRelevantGenes</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#BioMART.getRelevantGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.BioMART.getRelevantGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Is not implemented for BiomaRt.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of gene names to be mapped</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code> as this knowledge base is not intended to be used for such analyses.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.BioMART.getGeneScores">
<code class="descname">getGeneScores</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#BioMART.getGeneScores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.BioMART.getGeneScores" title="Permalink to this definition">¶</a></dt>
<dd><p>Is not implemented for BiomaRt.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of gene names to be mapped</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code> as this knowledge base is not intended to be used for such analyses.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.BioMART.getRelevantPathways">
<code class="descname">getRelevantPathways</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#BioMART.getRelevantPathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.BioMART.getRelevantPathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Is not implemented for BiomaRt.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of labels for which to find related pathways.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code> as this knowledge base is not intended to be used for such analyses.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.Gconvert">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">Gconvert</code><a class="reference internal" href="_modules/knowledgebases.html#Gconvert"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Gconvert" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a></p>
<p>Maps identifiers or data sets containing identifiers to the desired format by using the g:Convert web service.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> (<em>str</em>) – name of the knowledge base</li>
<li><strong>config</strong> (<em>dict</em>) – configuration parameter of the knowledge base as specified in the config file.</li>
<li><strong>webservice</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bioservices.REST</span></code> or inheriting classes) – web service querying object.</li>
<li><strong>hasGeneInformation</strong> (<em>bool</em>) – true if the knowledge base provides gene association information, false otherwise</li>
<li><strong>hasPathwayInformation</strong> (<em>bool</em>) – true if the knowledge base also provides pathway information, false otherwise</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="knowledgebases.Gconvert.mapItems">
<code class="descname">mapItems</code><span class="sig-paren">(</span><em>itemList</em>, <em>originalFormat</em>, <em>desiredFormat</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Gconvert.mapItems"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Gconvert.mapItems" title="Permalink to this definition">¶</a></dt>
<dd><p>Map a list of identifiers to the desired format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>itemList</strong> (<em>list of str</em>) – list of identifiers to be mapped.</li>
<li><strong>originalFormat</strong> (<em>str</em>) – current format of the identifiers.</li>
<li><strong>desiredFormat</strong> (<em>str</em>) – desired format to which to map identifiers.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">DataFrame table containing mappings of the identifiers from original to desired format.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Gconvert.getRelevantGenes">
<code class="descname">getRelevantGenes</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Gconvert.getRelevantGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Gconvert.getRelevantGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Is not implemented for g:Convert.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of gene names to be mapped</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code> as this knowledge base is not intended to be used for such analyses.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Gconvert.getGeneScores">
<code class="descname">getGeneScores</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Gconvert.getGeneScores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Gconvert.getGeneScores" title="Permalink to this definition">¶</a></dt>
<dd><p>Is not implemented for g:Convert.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of gene names to be mapped</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code> as this knowledge base is not intended to be used for such analyses.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Gconvert.getRelevantPathways">
<code class="descname">getRelevantPathways</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Gconvert.getRelevantPathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Gconvert.getRelevantPathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Is not implemented for g:Convert.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of labels for which to find related pathways.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code> as this knowledge base is not intended to be used for such analyses.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.OpenTargets">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">OpenTargets</code><a class="reference internal" href="_modules/knowledgebases.html#OpenTargets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.OpenTargets" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a></p>
<p>Knowledge base implementation of OpenTargets.
Uses the OpenTargetsClient Python implementation provided by OpenTargets to query the web service API.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> (<em>str</em>) – name of the knowledge base</li>
<li><strong>config</strong> (<em>dict</em>) – configuration parameter of the knowledge base as specified in the config file.</li>
<li><strong>webservice</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">opentargets.OpenTargetsClient</span></code>) – web service querying implementation.</li>
<li><strong>hasGeneInformation</strong> (<em>bool</em>) – true if the knowledge base provides gene association information, false otherwise</li>
<li><strong>hasPathwayInformation</strong> (<em>bool</em>) – true if the knowledge base also provides pathway information, false otherwise</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="knowledgebases.OpenTargets.getAssociations">
<code class="descname">getAssociations</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#OpenTargets.getAssociations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.OpenTargets.getAssociations" title="Permalink to this definition">¶</a></dt>
<dd><p>Get all relevant information for a given set of labels, sorted by their association scores in descending order.
Writes web service results into an intermediate file and maps the identifiers to have the correct format for further processing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of labels, e.g. disease names.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">DataFrame containing all related genes and their association scores.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.OpenTargets.getRelevantGenes">
<code class="descname">getRelevantGenes</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#OpenTargets.getRelevantGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.OpenTargets.getRelevantGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Get all genes that are somehow associated to the given labels, e.g. disease names.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of identifiers, e.g. disease names, for which to find associated genes.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of associated genes.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.OpenTargets.getGeneScores">
<code class="descname">getGeneScores</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#OpenTargets.getGeneScores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.OpenTargets.getGeneScores" title="Permalink to this definition">¶</a></dt>
<dd><p>Get all genes and their association scores that are related to the given labels, e.g. disease names.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of identifiers, e.g. disease names, for which to find associated genes.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">DataFrame of associated genes and their association scores, in descending order.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.OpenTargets.getRelevantPathways">
<code class="descname">getRelevantPathways</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#OpenTargets.getRelevantPathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.OpenTargets.getRelevantPathways" title="Permalink to this definition">¶</a></dt>
<dd><p>As OpenTargets currently does not provide pathway information, this feature is not implemented for OpenTargets.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of labels for which to find related pathways.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code> as this knowledge base is not intended to be used for such analyses.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.Kegg">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">Kegg</code><span class="sig-paren">(</span><em>pathwayparser</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Kegg"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Kegg" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a></p>
<p>Knowledge base implementation for KEGG.
Uses the KEGG web service implementation provided by bioservices.
Requires an instance of <a class="reference internal" href="#knowledgebases.KEGGPathwayParser" title="knowledgebases.KEGGPathwayParser"><code class="xref py py-class docutils literal notranslate"><span class="pre">KEGGPathwayParser</span></code></a> to be able to map retrieved pathways into the internal pathway format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> (<em>str</em>) – name of the knowledge base</li>
<li><strong>config</strong> (<em>dict</em>) – configuration parameter of the knowledge base as specified in the config file.</li>
<li><strong>webservice</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bioservices.KEGG</span></code>) – web service querying implementation.</li>
<li><strong>hasGeneInformation</strong> (<em>bool</em>) – true if the knowledge base provides gene association information, false otherwise</li>
<li><strong>hasPathwayInformation</strong> (<em>bool</em>) – true if the knowledge base also provides pathway information, false otherwise</li>
<li><strong>pathwayparser</strong> (<a class="reference internal" href="#knowledgebases.KEGGPathwayParser" title="knowledgebases.KEGGPathwayParser"><code class="xref py py-class docutils literal notranslate"><span class="pre">KEGGPathwayParser</span></code></a>) – pathway mapping class that transforms KEGG pathways in SIF format into the internally used pathway format.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="knowledgebases.Kegg.getPathwayNames">
<code class="descname">getPathwayNames</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Kegg.getPathwayNames"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Kegg.getPathwayNames" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve all pathway names related to the given labels, e.g. disease names.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list labels, e.g. disease names, for which to find pathways.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of pathway names.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Kegg.getRelevantGenes">
<code class="descname">getRelevantGenes</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Kegg.getRelevantGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Kegg.getRelevantGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Get all genes that are related to a set of labels, e.g. disease names.
For KEGG, this means we retrieve all genes that are contained in pathways associated to these labels.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of identifiers, e.g. disease names, for which to find associated genes.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of associated genes.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Kegg.getGeneScores">
<code class="descname">getGeneScores</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Kegg.getGeneScores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Kegg.getGeneScores" title="Permalink to this definition">¶</a></dt>
<dd><p>Get association scores for all genes that are related to the provided labels, e.g. disease names.
For KEGG, the association score for a gene is the sum of its degree percentile rank for every pathway, normalized by the overall number of pathways retrieved.
This favors hub genes/genes having many interactions with other genes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of identifiers, e.g. disease names, for which to find associated genes.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">DataFrame of associated genes and their association scores, in descending order.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Kegg.getRelevantPathways">
<code class="descname">getRelevantPathways</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Kegg.getRelevantPathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Kegg.getRelevantPathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Get all pathways related to a set of labels, e.g. disease names.
Uses the <a class="reference internal" href="#knowledgebases.KEGGPathwayParser" title="knowledgebases.KEGGPathwayParser"><code class="xref py py-class docutils literal notranslate"><span class="pre">KEGGPathwayParser</span></code></a> to map KEGG’s pathways from SIF to <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of gene names to be mapped</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">dict of pathway names and their internal representation as <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code>.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dict</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.Disgenet">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">Disgenet</code><a class="reference internal" href="_modules/knowledgebases.html#Disgenet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Disgenet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a></p>
<p>Knowledge base implementation for DisGeNET.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> (<em>str</em>) – name of the knowledge base</li>
<li><strong>config</strong> (<em>dict</em>) – configuration parameter of the knowledge base as specified in the config file.</li>
<li><strong>webservice</strong> (<a class="reference internal" href="#knowledgebases.DISGENET" title="knowledgebases.DISGENET"><code class="xref py py-class docutils literal notranslate"><span class="pre">DISGENET</span></code></a>) – web service querying implementation.</li>
<li><strong>hasGeneInformation</strong> (<em>bool</em>) – true if the knowledge base provides gene association information, false otherwise</li>
<li><strong>hasPathwayInformation</strong> (<em>bool</em>) – true if the knowledge base also provides pathway information, false otherwise</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="knowledgebases.Disgenet.getRelevantGenes">
<code class="descname">getRelevantGenes</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Disgenet.getRelevantGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Disgenet.getRelevantGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Get all genes that are related to a set of labels, e.g. disease names.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of identifiers, e.g. disease names, for which to find associated genes.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of associated genes.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Disgenet.getGeneScores">
<code class="descname">getGeneScores</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Disgenet.getGeneScores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Disgenet.getGeneScores" title="Permalink to this definition">¶</a></dt>
<dd><p>Get association scores for all genes that are related to the provided labels, e.g. disease names.
DisGeNET provides a couple of association scores to its genes (<a class="reference external" href="https://www.disgenet.org/dbinfo">https://www.disgenet.org/dbinfo</a>).
Which score to use can be defined by the user in the config file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of identifiers, e.g. disease names, for which to find associated genes.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">DataFrame of associated genes and their association scores, in descending order.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Disgenet.getRelevantPathways">
<code class="descname">getRelevantPathways</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Disgenet.getRelevantPathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Disgenet.getRelevantPathways" title="Permalink to this definition">¶</a></dt>
<dd><p>As DisGeNET currently does not provide pathway information, this feature is not implemented.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of labels for which to find related pathways.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code> as this knowledge base is not intended to be used for such analyses.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.Pathwaycommons">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">Pathwaycommons</code><a class="reference internal" href="_modules/knowledgebases.html#Pathwaycommons"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Pathwaycommons" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a></p>
<p>Knowledge base implementation for PathwayCommons.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> (<em>str</em>) – name of the knowledge base</li>
<li><strong>config</strong> (<em>dict</em>) – configuration parameter of the knowledge base as specified in the config file.</li>
<li><strong>webservice</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">opentargets.OpenTargetsClient</span></code>) – web service querying implementation.</li>
<li><strong>hasGeneInformation</strong> (<em>bool</em>) – true if the knowledge base provides gene association information, false otherwise</li>
<li><strong>hasPathwayInformation</strong> (<em>bool</em>) – true if the knowledge base also provides pathway information, false otherwise</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="knowledgebases.Pathwaycommons.getGeneScores">
<code class="descname">getGeneScores</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Pathwaycommons.getGeneScores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Pathwaycommons.getGeneScores" title="Permalink to this definition">¶</a></dt>
<dd><p>Get association scores for all genes that are related to the provided labels, e.g. disease names.
For PathwayCommons, the association score for a gene is the sum of its degree percentile rank for every pathway, normalized by the overall number of pathways retrieved.
This favors hub genes/genes having many interactions with other genes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of identifiers, e.g. disease names, for which to find associated genes.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">DataFrame of associated genes and their association scores, in descending order.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Pathwaycommons.getRelevantGenes">
<code class="descname">getRelevantGenes</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Pathwaycommons.getRelevantGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Pathwaycommons.getRelevantGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Get all genes that are related to a set of labels, e.g. disease names.
For PathwayCommons, this means we retrieve all genes that are contained in pathways associated to these labels.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of identifiers, e.g. disease names, for which to find associated genes.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of associated genes.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Pathwaycommons.readPathway">
<code class="descname">readPathway</code><span class="sig-paren">(</span><em>pathway</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Pathwaycommons.readPathway"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Pathwaycommons.readPathway" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads a pathway to create <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pathway</strong> (<em>str</em>) – pathway string to parse</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.Pathwaycommons.getRelevantPathways">
<code class="descname">getRelevantPathways</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#Pathwaycommons.getRelevantPathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.Pathwaycommons.getRelevantPathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Get all pathways related to a set of labels, e.g. disease names as <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list of str</em>) – list of gene names to be mapped</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">dict of pathway names and their internal representation as <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code>.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dict</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.PathwayParser">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">PathwayParser</code><a class="reference internal" href="_modules/knowledgebases.html#PathwayParser"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.PathwayParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Super class that maps a pathway from its original format (provided by a knowledge base) to the internally used <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code>.
When having to map pathways from a knowledge base, implement a new class that inherits from this one and implements <a class="reference internal" href="#knowledgebases.PathwayParser.parsePathway" title="knowledgebases.PathwayParser.parsePathway"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PathwayParser.parsePathway()</span></code></a>.</p>
<dl class="method">
<dt id="knowledgebases.PathwayParser.parsePathway">
<code class="descname">parsePathway</code><span class="sig-paren">(</span><em>pathway</em>, <em>pathwayID</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#PathwayParser.parsePathway"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.PathwayParser.parsePathway" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract method.
Parse a pathway to the internally used format of <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pathway</strong> (<em>str</em>) – pathway string to parse</li>
<li><strong>pathwayID</strong> (<em>str</em>) – name of the pathway</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">pathway in the internally used format..</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="knowledgebases.KEGGPathwayParser">
<em class="property">class </em><code class="descclassname">knowledgebases.</code><code class="descname">KEGGPathwayParser</code><a class="reference internal" href="_modules/knowledgebases.html#KEGGPathwayParser"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.KEGGPathwayParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#knowledgebases.PathwayParser" title="knowledgebases.PathwayParser"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.PathwayParser</span></code></a></p>
<p>Parse KEGG pathways, which are returned in KGML format.</p>
<dl class="method">
<dt id="knowledgebases.KEGGPathwayParser.readInteractions">
<code class="descname">readInteractions</code><span class="sig-paren">(</span><em>interactions</em>, <em>geneIds</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#KEGGPathwayParser.readInteractions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.KEGGPathwayParser.readInteractions" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses interactions for a set of genes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>interactions</strong> (<em>list</em>) – interactions to parse</li>
<li><strong>geneIds</strong> (<em>list of str</em>) – gene ids whose interactions to add</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="knowledgebases.KEGGPathwayParser.parsePathway">
<code class="descname">parsePathway</code><span class="sig-paren">(</span><em>kgml_pathway</em>, <em>pathwayID</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowledgebases.html#KEGGPathwayParser.parsePathway"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#knowledgebases.KEGGPathwayParser.parsePathway" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse KEGG pathway to the internally used format of <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pathway</strong> (<em>str</em>) – pathway string to parse</li>
<li><strong>pathwayID</strong> (<em>str</em>) – name of the pathway</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">pathway in the internally used format..</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.Network</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="evaluation-module">
<h2>evaluation module<a class="headerlink" href="#evaluation-module" title="Permalink to this headline">¶</a></h2>
<p>Contains all classes related to the evaluation part.
There are distinct classes for the following evaluation aspects:
* review of knowledge base coverage for the provided search terms (see <a class="reference internal" href="#evaluation.KnowledgeBaseEvaluator" title="evaluation.KnowledgeBaseEvaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.KnowledgeBaseEvaluator</span></code></a>)
* inspection of data set quality, e.g. via mds or density plots (see <a class="reference internal" href="#evaluation.DatasetEvaluator" title="evaluation.DatasetEvaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.DatasetEvaluator</span></code></a>)
* comparison and assessment of feature rankings, e.g. overlap (see <a class="reference internal" href="#evaluation.RankingsEvaluator" title="evaluation.RankingsEvaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.RankingsEvaluator</span></code></a>)
* annotation of feature rankings and enrichment via EnrichR (see <a class="reference internal" href="#evaluation.AnnotationEvaluator" title="evaluation.AnnotationEvaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.AnnotationEvaluator</span></code></a>)
* classification and subsequent visualization of standard metrics (see <a class="reference internal" href="#evaluation.ClassificationEvaluator" title="evaluation.ClassificationEvaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.ClassificationEvaluator</span></code></a>)
* cross-classification across a second data set and visualization of standard metrics (see <a class="reference internal" href="#evaluation.CrossEvaluator" title="evaluation.CrossEvaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.CrossEvaluator</span></code></a>)</p>
<p>Every one of these classes inherits from the abstract <a class="reference internal" href="#evaluation.Evaluator" title="evaluation.Evaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.Evaluator</span></code></a> and implements the evaluate() method.
<a class="reference internal" href="#evaluation.AttributeRemover" title="evaluation.AttributeRemover"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.AttributeRemover</span></code></a> is used in the overall benchmarking process to prepare the input data to contain only the selected features.
For a detailed look at the class architecture, have a look at ADD CLASS ARCHITECTURE LINK HERE.</p>
<span class="target" id="module-evaluation"></span><dl class="class">
<dt id="evaluation.AttributeRemover">
<em class="property">class </em><code class="descclassname">evaluation.</code><code class="descname">AttributeRemover</code><span class="sig-paren">(</span><em>dataDir</em>, <em>rankingsDir</em>, <em>topK</em>, <em>outputDir</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#AttributeRemover"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.AttributeRemover" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Prepares the input data set for subsequent classification by removing lowly-ranked features and only keeping the top k features.
Creates one “reduced” file for every ranking and from one to k (so if k is 50, we will end up with 50 files having one and up to 50 features.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>dataDir</strong> (<em>str</em>) – absolute path to the directory that contains the input data set whose features to reduce.</li>
<li><strong>rankingsDir</strong> (<em>str</em>) – absolute path to the directory that contains the rankings.</li>
<li><strong>topK</strong> (<em>str</em>) – maximum numbers of features to select.</li>
<li><strong>outputDir</strong> (<em>str</em>) – absolute path to the directory where the reduced files will be stored.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="evaluation.AttributeRemover.loadTopKRankings">
<code class="descname">loadTopKRankings</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#AttributeRemover.loadTopKRankings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.AttributeRemover.loadTopKRankings" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads all available rankings from files.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">Dictionary with selection methods as keys and  a ranked list of the (column) names of the top k features.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">dict</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.AttributeRemover.removeAttributesFromDataset">
<code class="descname">removeAttributesFromDataset</code><span class="sig-paren">(</span><em>method</em>, <em>ranking</em>, <em>dataset</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#AttributeRemover.removeAttributesFromDataset"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.AttributeRemover.removeAttributesFromDataset" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates reduced data sets from dataset for the given method’s ranking that only contain the top x features.
Creates multiple reduced data sets from topKmin to topKmax specified in the config.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>method</strong> (<em>str</em>) – selection method applied for the ranking.</li>
<li><strong>ranking</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of str) – (ranked) list of feature names from the top k features.</li>
<li><strong>dataset</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – original input data set</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.AttributeRemover.removeUnusedAttributes">
<code class="descname">removeUnusedAttributes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#AttributeRemover.removeUnusedAttributes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.AttributeRemover.removeUnusedAttributes" title="Permalink to this definition">¶</a></dt>
<dd><p>For every method and its corresponding ranking, create reduced files with only the top x features.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="evaluation.Evaluator">
<em class="property">class </em><code class="descclassname">evaluation.</code><code class="descname">Evaluator</code><span class="sig-paren">(</span><em>input</em>, <em>output</em>, <em>methodColors</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#Evaluator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.Evaluator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Abstract super class.
Every evaluation class has to inherit from this class and implement its <a class="reference internal" href="#evaluation.Evaluator.evaluate" title="evaluation.Evaluator.evaluate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">evaluate()</span></code></a> method.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to the directory where the input data is located.</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to the directory to which to save results.</li>
<li><strong>methodColors</strong> (<em>dict of str</em>) – dictionary containing a color string for every selection method.</li>
<li><strong>javaConfig</strong> (<em>str</em>) – configuration parameters for java code (as specified in the config file).</li>
<li><strong>rConfig</strong> (<em>str</em>) – configuration parameters for R code (as specified in the config file).</li>
<li><strong>evalConfig</strong> (<em>str</em>) – configuration parameters for evaluation, e.g. how many features to select (as specified in the config file).</li>
<li><strong>classificationConfig</strong> (<em>str</em>) – configuration parameters for classification, e.g. which classifiers to use (as specified in the config file).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="evaluation.Evaluator.evaluate">
<code class="descname">evaluate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#Evaluator.evaluate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.Evaluator.evaluate" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract.
Must be implemented by inheriting class as this method is invoked by <code class="xref py py-class docutils literal notranslate"><span class="pre">framework.Framework</span></code> to run the evaluation.</p>
</dd></dl>

<dl class="method">
<dt id="evaluation.Evaluator.loadRankings">
<code class="descname">loadRankings</code><span class="sig-paren">(</span><em>inputDir</em>, <em>maxRank</em>, <em>keepOrder</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#Evaluator.loadRankings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.Evaluator.loadRankings" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads all rankings from a specified input directory.
If only the top k features shall be in the ranking, set maxRank accordingly, set it to 0 if otherwise (so to load all features).
If feature order is important in the returned rankings, set keepOrder to true; if you are only interested in what features are among the top maxRank, set it to false.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>inputDir</strong> (<em>str</em>) – absolute path to directory where all rankings are located.</li>
<li><strong>maxRank</strong> (<em>int</em>) – maximum number of features to have in ranking.</li>
<li><strong>keepOrder</strong> (<em>bool</em>) – whether the order of the features in the ranking is important or not.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Dictionary of rankings per method, either as ordered list or set (depending on keepOrder attribute)</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dict</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.Evaluator.computeKendallsW">
<code class="descname">computeKendallsW</code><span class="sig-paren">(</span><em>rankings</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#Evaluator.computeKendallsW"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.Evaluator.computeKendallsW" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes Kendall’s W from two rankings.
Note: measure does not make much sense if the two rankings are highly disjunct, which can happen especially for traditional approaches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>rankings</strong> (<em>matrix</em>) – matrix containing two rankings for which to compute Kendall’s W.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Kendall’s W score.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">float</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="evaluation.ClassificationEvaluator">
<em class="property">class </em><code class="descclassname">evaluation.</code><code class="descname">ClassificationEvaluator</code><span class="sig-paren">(</span><em>inputDir</em>, <em>rankingsDir</em>, <em>intermediateDir</em>, <em>outputDir</em>, <em>methodColors</em>, <em>methodMarkers</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#ClassificationEvaluator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.ClassificationEvaluator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#evaluation.Evaluator" title="evaluation.Evaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.Evaluator</span></code></a></p>
<p>Evaluates selection methods via classification by using only the selected features and computing multiple standard metrics.
Uses <a class="reference internal" href="#evaluation.AttributeRemover" title="evaluation.AttributeRemover"><code class="xref py py-class docutils literal notranslate"><span class="pre">AttributeRemover</span></code></a> to create reduced datasets containing only the top k features, which are then used for subsequent classification.
Currently, classification and subsequent evaluation is wrapped here and is actually carried out by java jars using WEKA.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to the directory where the input data for classification is located.</li>
<li><strong>rankingsDir</strong> (<em>str</em>) – absolute path to the directory where the rankings are located.</li>
<li><strong>intermediateDir</strong> (<em>str</em>) – absolute path to the directory where the reduced datasets (containing only the top k features) are written to.</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to the directory to which to save results.</li>
<li><strong>methodColors</strong> (<em>dict of str</em>) – dictionary containing a color string for every selection method.</li>
<li><strong>javaConfig</strong> (<em>str</em>) – configuration parameters for java code (as specified in the config file).</li>
<li><strong>rConfig</strong> (<em>str</em>) – configuration parameters for R code (as specified in the config file).</li>
<li><strong>evalConfig</strong> (<em>str</em>) – configuration parameters for evaluation, e.g. how many features to select (as specified in the config file).</li>
<li><strong>classificationConfig</strong> (<em>str</em>) – configuration parameters for classification, e.g. which classifiers to use (as specified in the config file).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="evaluation.ClassificationEvaluator.drawLinePlot">
<code class="descname">drawLinePlot</code><span class="sig-paren">(</span><em>inputDir</em>, <em>outputDir</em>, <em>topK</em>, <em>metric</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#ClassificationEvaluator.drawLinePlot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.ClassificationEvaluator.drawLinePlot" title="Permalink to this definition">¶</a></dt>
<dd><p>Draws a line plot for a given metric, using all files containing evaluation results for that metric in inputDir.
In the end, the plot will have one line per feature selection approach for which classification results are available.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>inputDir</strong> (<em>str</em>) – absolute path to directory containing all input files (from which to draw the graph).</li>
<li><strong>outputDir</strong> (<em>str</em>) – absolute path to the output directory where the the graph will be saved.</li>
<li><strong>topK</strong> (<em>int</em>) – maximum x axis value</li>
<li><strong>metric</strong> (<em>str</em>) – metric name for which to draw the graph.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.ClassificationEvaluator.evaluate">
<code class="descname">evaluate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#ClassificationEvaluator.evaluate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.ClassificationEvaluator.evaluate" title="Permalink to this definition">¶</a></dt>
<dd><p>Triggers classification and evaluation in Java and creates corresponding plots for every metric that was selected in the config.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="evaluation.RankingsEvaluator">
<em class="property">class </em><code class="descclassname">evaluation.</code><code class="descname">RankingsEvaluator</code><span class="sig-paren">(</span><em>input</em>, <em>dataset</em>, <em>outputPath</em>, <em>methodColors</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#RankingsEvaluator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.RankingsEvaluator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#evaluation.Evaluator" title="evaluation.Evaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.Evaluator</span></code></a></p>
<p>Evaluates the rankings themselves by generating overlaps and comparing fold change differences.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to the directory where the input data is located.</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to the directory to which to save results.</li>
<li><strong>methodColors</strong> (<em>dict of str</em>) – dictionary containing a color string for every selection method.</li>
<li><strong>javaConfig</strong> (<em>str</em>) – configuration parameters for java code (as specified in the config file).</li>
<li><strong>rConfig</strong> (<em>str</em>) – configuration parameters for R code (as specified in the config file).</li>
<li><strong>evalConfig</strong> (<em>str</em>) – configuration parameters for evaluation, e.g. how many features to select (as specified in the config file).</li>
<li><strong>classificationConfig</strong> (<em>str</em>) – configuration parameters for classification, e.g. which classifiers to use (as specified in the config file).</li>
<li><strong>dataset</strong> (<em>str</em>) – absolute file path to the input data set (from which features were selected).</li>
<li><strong>metrics</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of str) – list of metrics to apply to ranking evaluation (as specified in the config file).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="evaluation.RankingsEvaluator.generateOverlaps">
<code class="descname">generateOverlaps</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#RankingsEvaluator.generateOverlaps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.RankingsEvaluator.generateOverlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates overlap plots for the available rankings (set during creating to self.input).
For up to two rankings, use Python’s matplotlib to create Venn diagrams.
For three rankings and above, create UpsetR (<a class="reference external" href="https://github.com/hms-dbmi/UpSetR">https://github.com/hms-dbmi/UpSetR</a>) diagrams via R.</p>
</dd></dl>

<dl class="method">
<dt id="evaluation.RankingsEvaluator.loadGeneRanks">
<code class="descname">loadGeneRanks</code><span class="sig-paren">(</span><em>inputDir</em>, <em>topK</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#RankingsEvaluator.loadGeneRanks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.RankingsEvaluator.loadGeneRanks" title="Permalink to this definition">¶</a></dt>
<dd><p>Used for computing Kendall’s W.
Loads rankings and creates a table (approach x features) containing individual ranks per feature per approach, e.g.
#approach   G1  G2  G3
#Ranker1    1   2   3
#Ranker 2   3   1   2</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>inputDir</strong> (<em>str</em>) – absolute path to the directory containing all ranking files.</li>
<li><strong>topK</strong> (<em>int</em>) – maximum number of features to use (=length of the rankings).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Ranking table containing every assigned rank for every feature per ranking approach.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.array</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.RankingsEvaluator.computePValue">
<code class="descname">computePValue</code><span class="sig-paren">(</span><em>W</em>, <em>m</em>, <em>n</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#RankingsEvaluator.computePValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.RankingsEvaluator.computePValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the p-value for a given Kendall’s W score via a simple permutation (1000 times) test.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>W</strong> (<em>float</em>) – Kendall’s W score.</li>
<li><strong>m</strong> (<em>int</em>) – number of approaches/rankings to compare.</li>
<li><strong>n</strong> (<em>int</em>) – number of features in each ranking.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">p-value of Kendall’s W score.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">float</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.RankingsEvaluator.computeKendallsWScores">
<code class="descname">computeKendallsWScores</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#RankingsEvaluator.computeKendallsWScores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.RankingsEvaluator.computeKendallsWScores" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes Kendall’s correlation coefficients (W) and its corresponding p-value for the top 50, 500, 5,000 and all (code: 0) ranked features of existing rankings.
Conducts a permutation test for all scores to receive p-value.
Writes output to a file containing the correlation coefficients and their corresponding p-value for different length of rankings.</p>
</dd></dl>

<dl class="method">
<dt id="evaluation.RankingsEvaluator.drawBoxPlot">
<code class="descname">drawBoxPlot</code><span class="sig-paren">(</span><em>data</em>, <em>labels</em>, <em>prefix</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#RankingsEvaluator.drawBoxPlot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.RankingsEvaluator.drawBoxPlot" title="Permalink to this definition">¶</a></dt>
<dd><p>Draws a box plot from the given data with the given labels on the x axis and the given prefix in the headlines.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of lists of floats) – Data to plot; a list containing lists of values.</li>
<li><strong>labels</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of str) – List of method names.</li>
<li><strong>prefix</strong> (<em>str</em>) – Prefix to use for file name and title.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.RankingsEvaluator.computeFoldChangeDiffs">
<code class="descname">computeFoldChangeDiffs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#RankingsEvaluator.computeFoldChangeDiffs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.RankingsEvaluator.computeFoldChangeDiffs" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes median and mean fold changes for all selected features per approach.
Writes fold changes to file and creates corresponding box plots.</p>
</dd></dl>

<dl class="method">
<dt id="evaluation.RankingsEvaluator.evaluate">
<code class="descname">evaluate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#RankingsEvaluator.evaluate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.RankingsEvaluator.evaluate" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs evaluations on feature rankings based on what is specified in the config file.
Currently, can compute feature overlaps, Kendall’s correlation coefficient (W), and box plots for mean and median fold changes of selected features.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="evaluation.CrossEvaluator">
<em class="property">class </em><code class="descclassname">evaluation.</code><code class="descname">CrossEvaluator</code><span class="sig-paren">(</span><em>input</em>, <em>rankingsDir</em>, <em>output</em>, <em>methodColors</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#CrossEvaluator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.CrossEvaluator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#evaluation.Evaluator" title="evaluation.Evaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.Evaluator</span></code></a></p>
<p>Runs the evaluation across a second data set.
Takes the top k ranked features, removes all other features from that second data set.
Runs a <a class="reference internal" href="#evaluation.ClassificationEvaluator" title="evaluation.ClassificationEvaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClassificationEvaluator</span></code></a> on that second data set with the selected features.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to the directory where the second data set for cross-validation is located.</li>
<li><strong>rankingsDir</strong> (<em>str</em>) – absolute path to the directory containing all rankings.</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to the directory to which to write all classification results.</li>
<li><strong>methodColors</strong> (<em>dict</em>) – Dictionary that assigns every (ranking) method a unique color (used for drawing subsequent plots).</li>
<li><strong>javaConfig</strong> (<em>str</em>) – configuration parameters for java code (as specified in the config file).</li>
<li><strong>rConfig</strong> (<em>str</em>) – configuration parameters for R code (as specified in the config file).</li>
<li><strong>evalConfig</strong> (<em>str</em>) – configuration parameters for evaluation, e.g. how many features to select (as specified in the config file).</li>
<li><strong>classificationConfig</strong> (<em>str</em>) – configuration parameters for classification, e.g. which classifiers to use (as specified in the config file).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="evaluation.CrossEvaluator.evaluate">
<code class="descname">evaluate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#CrossEvaluator.evaluate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.CrossEvaluator.evaluate" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs crossClassification = takes the features selected on the original data set and uses them to classify a second (cross-validation) data set.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="evaluation.AnnotationEvaluator">
<em class="property">class </em><code class="descclassname">evaluation.</code><code class="descname">AnnotationEvaluator</code><span class="sig-paren">(</span><em>input</em>, <em>output</em>, <em>methodColors</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#AnnotationEvaluator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.AnnotationEvaluator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#evaluation.Evaluator" title="evaluation.Evaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.Evaluator</span></code></a></p>
<p>Annotates and enriches feature rankings with EnrichR (<a class="reference external" href="https://maayanlab.cloud/Enrichr/">https://maayanlab.cloud/Enrichr/</a>).
What library to be used for annotation must be specified in the config file.
Can compute annotation and enrichment overlaps for different feature rankings.
Annotation = annotate features with terms/information
Enrichment = check what terms are enriched in a feature ranking and are related to multiple features.
Overlaps then can show a) if feature rankings represent the same underlying processes via annotation (maybe although having selected different features), or b) if the underlying processes are equally strongly represented by checking the enrichment (maybe altough having seleced different features).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to the directory where the second data set for cross-validation is located.</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to the directory to which to write all classification results.</li>
<li><strong>methodColors</strong> (<em>dict</em>) – dictionary that assigns every (ranking) method a unique color (used for drawing subsequent plots).</li>
<li><strong>metrics</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of str) – list of metrics to compute, to be configured in config.</li>
<li><strong>dataConfig</strong> (<em>dict</em>) – config parameters for the input dataset.</li>
<li><strong>javaConfig</strong> (<em>str</em>) – configuration parameters for java code (as specified in the config file).</li>
<li><strong>rConfig</strong> (<em>str</em>) – configuration parameters for R code (as specified in the config file).</li>
<li><strong>evalConfig</strong> (<em>str</em>) – configuration parameters for evaluation, e.g. how many features to select (as specified in the config file).</li>
<li><strong>classificationConfig</strong> (<em>str</em>) – configuration parameters for classification, e.g. which classifiers to use (as specified in the config file).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="evaluation.AnnotationEvaluator.countAnnotationPercentages">
<code class="descname">countAnnotationPercentages</code><span class="sig-paren">(</span><em>featureLists</em>, <em>inputDir</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#AnnotationEvaluator.countAnnotationPercentages"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.AnnotationEvaluator.countAnnotationPercentages" title="Permalink to this definition">¶</a></dt>
<dd><p>Count the number of features (=number of lines in annotation file) that have been annotated and compute percentages.
Write output to a file “annotationsPercentages.csv” in self.outputDir.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>featureLists</strong> (<em>dict</em>) – dictionary of lists of features per selection method.</li>
<li><strong>inputDir</strong> (<em>str</em>) – absolute path to directory containing annotation files.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.AnnotationEvaluator.loadAnnotationFiles">
<code class="descname">loadAnnotationFiles</code><span class="sig-paren">(</span><em>inputDir</em>, <em>inputFiles</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#AnnotationEvaluator.loadAnnotationFiles"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.AnnotationEvaluator.loadAnnotationFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads files with feature annotations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>inputDir</strong> (<em>str</em>) – absolute path to directory containing annotation files.</li>
<li><strong>inputFiles</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of str) – list of annotation file names to load.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">dictionary of annotation sets per selection method.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dict</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.AnnotationEvaluator.computeOverlap">
<code class="descname">computeOverlap</code><span class="sig-paren">(</span><em>inputDir</em>, <em>fileSuffix</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#AnnotationEvaluator.computeOverlap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.AnnotationEvaluator.computeOverlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates overlap plots for the available annotations/enrichments.
For up to two rankings, use Python’s matplotlib to create Venn diagrams.
For three rankings and above, create UpsetR (<a class="reference external" href="https://github.com/hms-dbmi/UpSetR">https://github.com/hms-dbmi/UpSetR</a>) diagrams via R.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>inputDir</strong> (<em>str</em>) – absolute path to directory containing files for which to compute overlap.</li>
<li><strong>fileSuffix</strong> (<em>str</em>) – suffix in filename to recognize the right files.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.AnnotationEvaluator.evaluate">
<code class="descname">evaluate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#AnnotationEvaluator.evaluate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.AnnotationEvaluator.evaluate" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs the annotation/enrichment evaluation on the rankings.
Depending on what was specified in the config file, annotate and/or enrich feature rankings and compute overlaps or percentages.
Overlaps then can show a) if feature rankings represent the same underlying processes via annotation (maybe although having selected different features), or b) if the underlying processes are equally strongly represented by checking the enrichment (maybe altough having seleced different features).</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="evaluation.DatasetEvaluator">
<em class="property">class </em><code class="descclassname">evaluation.</code><code class="descname">DatasetEvaluator</code><span class="sig-paren">(</span><em>input</em>, <em>output</em>, <em>separator</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#DatasetEvaluator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.DatasetEvaluator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#evaluation.Evaluator" title="evaluation.Evaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.Evaluator</span></code></a></p>
<p>Creates plots regarding data set quality, currently: MDS, density, and box plots.
Wrapper class because the actual evaluation and plot creation is done in an R script.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input</strong> (<em>str</em>) – absolute path to the directory where the input data set is located (for which to create the plots).</li>
<li><strong>output</strong> (<em>str</em>) – absolute path to the directory to which to save plots.</li>
<li><strong>separator</strong> (<em>str</em>) – separator character in data set to read it correctly.</li>
<li><strong>options</strong> (<em>list of str</em>) – what plots to create, a list of method names that must be specified in the config file.</li>
<li><strong>javaConfig</strong> (<em>str</em>) – configuration parameters for java code (as specified in the config file).</li>
<li><strong>rConfig</strong> (<em>str</em>) – configuration parameters for R code (as specified in the config file).</li>
<li><strong>evalConfig</strong> (<em>str</em>) – configuration parameters for evaluation, e.g. how many features to select (as specified in the config file).</li>
<li><strong>classificationConfig</strong> (<em>str</em>) – configuration parameters for classification, e.g. which classifiers to use (as specified in the config file).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="evaluation.DatasetEvaluator.evaluate">
<code class="descname">evaluate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#DatasetEvaluator.evaluate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.DatasetEvaluator.evaluate" title="Permalink to this definition">¶</a></dt>
<dd><p>Triggers the actual evaluation/plot generation in R.
If a second data set for cross-validation was provided, also run the corresponding R script on that data set.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="evaluation.KnowledgeBaseEvaluator">
<em class="property">class </em><code class="descclassname">evaluation.</code><code class="descname">KnowledgeBaseEvaluator</code><span class="sig-paren">(</span><em>output</em>, <em>knowledgebases</em>, <em>searchterms</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#KnowledgeBaseEvaluator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.KnowledgeBaseEvaluator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#evaluation.Evaluator" title="evaluation.Evaluator"><code class="xref py py-class docutils literal notranslate"><span class="pre">evaluation.Evaluator</span></code></a></p>
<p>Creates plots to evaluate knowledge base coverage.
Queries the knowledge bases with the given search terms and checks how many genes or pathways are found.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>output</strong> (<em>str</em>) – absolute path to the directory to which to save plots.</li>
<li><strong>knowledgebases</strong> (<em>list of str</em>) – a list of knowledgebases to test.</li>
<li><strong>searchterms</strong> (<em>list of str</em>) – list of search terms for which to check knowledge base coverage.</li>
<li><strong>javaConfig</strong> (<em>str</em>) – configuration parameters for java code (as specified in the config file).</li>
<li><strong>rConfig</strong> (<em>str</em>) – configuration parameters for R code (as specified in the config file).</li>
<li><strong>evalConfig</strong> (<em>str</em>) – configuration parameters for evaluation, e.g. how many features to select (as specified in the config file).</li>
<li><strong>classificationConfig</strong> (<em>str</em>) – configuration parameters for classification, e.g. which classifiers to use (as specified in the config file).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="evaluation.KnowledgeBaseEvaluator.drawCombinedPlot">
<code class="descname">drawCombinedPlot</code><span class="sig-paren">(</span><em>stats</em>, <em>colIndex</em>, <em>filename</em>, <em>title</em>, <em>ylabel1</em>, <em>ylabel2</em>, <em>colors</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#KnowledgeBaseEvaluator.drawCombinedPlot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.KnowledgeBaseEvaluator.drawCombinedPlot" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates combined plot of box and bar plot from a data set.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>stats</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – statistics to plot.</li>
<li><strong>colIndex</strong> (<em>int</em>) – column index to use as column.</li>
<li><strong>filename</strong> (<em>str</em>) – filename for the plot.</li>
<li><strong>title</strong> (<em>str</em>) – title for the plot.</li>
<li><strong>ylabel1</strong> (<em>str</em>) – label of y axis (left side/box plot).</li>
<li><strong>ylabel2</strong> (<em>str</em>) – label of y axis (right side/bar plot).</li>
<li><strong>colors</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of str) – List of colors to use for the different search terms.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.KnowledgeBaseEvaluator.createKnowledgeBases">
<code class="descname">createKnowledgeBases</code><span class="sig-paren">(</span><em>knowledgebaseList</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#KnowledgeBaseEvaluator.createKnowledgeBases"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.KnowledgeBaseEvaluator.createKnowledgeBases" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates knowledge base objects from a given list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>knowledgeBaseList</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of str.) – List of knowledge base names to create.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">List of knowledge base objects</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of <code class="xref py py-class docutils literal notranslate"><span class="pre">KnowledgeBase</span></code> or inheriting classes</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.KnowledgeBaseEvaluator.checkCoverage">
<code class="descname">checkCoverage</code><span class="sig-paren">(</span><em>kb</em>, <em>colors</em>, <em>useIDs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#KnowledgeBaseEvaluator.checkCoverage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.KnowledgeBaseEvaluator.checkCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the coverage for a given knowledge base and creates corresponding plots.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>kb</strong> (<a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a> or inheriting class) – knowledge base object for which to check coverage.</li>
<li><strong>colors</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of str) – List of colors to use for plots.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.KnowledgeBaseEvaluator.checkPathwayCoverage">
<code class="descname">checkPathwayCoverage</code><span class="sig-paren">(</span><em>kb</em>, <em>colors</em>, <em>useIDs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#KnowledgeBaseEvaluator.checkPathwayCoverage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.KnowledgeBaseEvaluator.checkPathwayCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the pathway coverage for a given knowledge base and creates corresponding plots.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>kb</strong> (<a class="reference internal" href="#knowledgebases.KnowledgeBase" title="knowledgebases.KnowledgeBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">knowledgebases.KnowledgeBase</span></code></a> or inheriting class) – knowledge base object for which to check pathway coverage.</li>
<li><strong>colors</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code> of str) – List of colors to use for plots.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="evaluation.KnowledgeBaseEvaluator.evaluate">
<code class="descname">evaluate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation.html#KnowledgeBaseEvaluator.evaluate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#evaluation.KnowledgeBaseEvaluator.evaluate" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluates every given knowledge base and checks how many genes and pathways (and how large they are) are in there for the given search terms.
Creates corresponding plots.</p>
</dd></dl>

</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="r-doc.html" class="btn btn-neutral float-right" title="R Code Documentation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="howtos.html" class="btn btn-neutral float-left" title="Extending Comprior - How-Tos" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Cindy Perscheid.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>